<head>
<META content="Microsoft FrontPage 4.0" name=GENERATOR>
<STYLE type=text/css>BODY {
	BACKGROUND-COLOR: #ffffff; COLOR: #000000; FONT-FAMILY: "Times New Roman"; FONT-SIZE: 12pt; TEXT-ALIGN: left
}
 UL.zzul {
	DISPLAY: block; LIST-STYLE: none}
SPAN.zzspace {
	LEFT: 11px; POSITION: relative
}
 A {
	COLOR: #0000ff
}
 A:visited {
	COLOR: #0000ff
}
 A:hover {
	COLOR: #ff0000
}
TD {
	COLOR: #000000; FONT-FAMILY: "Times New Roman"; FONT-SIZE: 12pt; TEXT-ALIGN: left
}
 TABLE {
	BORDER-BOTTOM: medium none; BORDER-LEFT: medium none; BORDER-RIGHT: medium none; BORDER-TOP: medium none; PADDING-BOTTOM: 3px; PADDING-LEFT: 3px; PADDING-RIGHT: 3px; PADDING-TOP: 3px; WIDTH: 100%
 }
 </STYLE>
 <SCRIPT type=text/javascript>
 <!--
 function exp_coll(ind)
 {
 s = document.getElementById("sp_" + ind);
 i = document.getElementById("im_" + ind);
 i2 = document.getElementById("im2_" + ind);
 if (s.style.display == 'none')
 {
  s.style.display = 'block';   
 }
 else if (s.style.display == 'block')
 {
   s.style.display = 'none';   
 }
}
 function coll_all(i)
 {  
 for(ind=1;ind<=i;ind++){
  x = document.getElementById("sp_" + ind);
  x.style.display = 'none';
  if (x.style.display == 'none')
  {    
 }
}
}
 function exp_all(i)
 {
  for(ind=1;ind<=i;ind++){
  x = document.getElementById("sp_" + ind);
  x.style.display = 'block';
  if (x.style.display == 'block')
  {    
 }
  }  
}
 -->
</SCRIPT>
</HEAD>
<BODY>
<center><a name="TMT Generated Test Plans"><h1>Test Plan</h1></a></center><hr><a name="xys"><table><tr><td>This is a TMT (Test Management Tool) generated Test Plan. To read a short note on TMT on the caBIG Best Practices SIG website, click <a href="http://cabig-bpsigwiki.bioinformatics.northwestern.edu/bpsigwiki/index.php/Manual_Testing_-_Test_Management_Tool">here</a>.<br> For more details or if you find any issue with this report, please send an email to <a href="mailto:tmt@persistent.co.in">tmt@persistent.co.in</a>.</td></tr></table></a><hr><br><br>
<A href="javascript:exp_all(4);">Expand All</A> &nbsp; <A href="javascript:coll_all(4);">Collapse All</A> 

<UL class=zzul>
  <LI>
<LI><A href="javascript:exp_coll(1);">+</A> <A 
	  href="javascript:exp_coll(1);"></A> &nbsp; <A  href="#Product Name: GeneConnect">GeneConnect</A> 
<UL class=zzul id=sp_1 style="DISPLAY: block">
<SCRIPT type=text/javascript>
 ind = 1; 
 x = document.getElementById("sp_" + ind);
 x.style.display = 'none'; 
 if (x.style.display == 'none'){ }
</SCRIPT>
<LI><A href="javascript:exp_coll(2);">+</A> <A 
	  href="javascript:exp_coll(2);"></A> &nbsp; <A  href="#Version Name: GeneConnectver 1.1">ver 1.1</A> 
<UL class=zzul id=sp_2 style="DISPLAY: block">
<SCRIPT type=text/javascript>
 ind = 2; 
 x = document.getElementById("sp_" + ind);
 x.style.display = 'none'; 
 if (x.style.display == 'none'){ }
</SCRIPT>
<LI><A href="javascript:exp_coll(3);">+</A> <A 
	  href="javascript:exp_coll(3);"></A> &nbsp; <A  href="#Component Name: GeneConnectver 1.1Metadata Calculation">Metadata Calculation</A> 
<UL class=zzul id=sp_3 style="DISPLAY: block">
<SCRIPT type=text/javascript>
 ind = 3; 
 x = document.getElementById("sp_" + ind);
 x.style.display = 'none'; 
 if (x.style.display == 'none'){ }
</SCRIPT>
<LI><A href="javascript:exp_coll(4);">+</A> <A 
	  href="javascript:exp_coll(4);"></A> &nbsp; <A  href="#Test Area Name: GeneConnectver 1.1Metadata CalculationMetadata Calculation">Metadata Calculation</A> 
<UL class=zzul id=sp_4 style="DISPLAY: block">
<SCRIPT type=text/javascript>
 ind = 4; 
 x = document.getElementById("sp_" + ind);
 x.style.display = 'none'; 
 if (x.style.display == 'none'){ }
</SCRIPT>
<LI><SPAN class=zzspace>&nbsp;&nbsp;
<A href="#Test Case ID: GeneConnectver 1.1Metadata CalculationMetadata CalculationMetadata Calculation">Metadata Calculation</A> </SPAN>
<LI><SPAN class=zzspace>&nbsp;&nbsp;
<A href="#Test Case ID: GeneConnectver 1.1Metadata CalculationMetadata CalculationMetadata Calculation for newly added DataSource">Metadata Calculation for newly added DataSource</A> </SPAN>
</LI></UL>

</LI></UL>

</LI></UL>
<br><br><a><table><tr><td width="50%"><table bgColor="#CCCCFF" border="1" cellPadding="2" width="100"><tr><td colspan="10"><font color="#660066"><b>Test case count: Component verses Priority</b></font></td></tr><tr><td  align="center" bgcolor="#ccccff"><font color="#660066"><b>Component Names</b></font></td><td align="center" bgcolor="#ccccff"> <font color="#660066"><b>P1</b></font></td><td align="center" bgcolor="#ccccff"> <font color="#660066"><b>P2</b></font></td><td align="center" bgcolor="#ccccff"> <font color="#660066"><b>P3</b></font></td><td align="center" bgcolor="#ccccff"> <font color="#660066"><b>P4</b></font></td><td align="center" bgcolor="#ccccff"> <font color="#660066"><b>Total</b></td></tr><tr><td bgcolor="#ffffff">ver 1.1</td><td  bgcolor="#ffffff">14</td><td  bgcolor="#ffffff">1</td><td  bgcolor="#ffffff">0</td><td  bgcolor="#ffffff">0</td><td bgcolor="#ffffff">15</td></tr><tr><td bgcolor="#ffffff">Metadata Calculation</td><td  bgcolor="#ffffff">0</td><td  bgcolor="#ffffff">2</td><td  bgcolor="#ffffff">0</td><td  bgcolor="#ffffff">0</td><td bgcolor="#ffffff">2</td></tr><tr><td bgcolor="#ccccff"><b>Column Total</b></td><td  bgcolor="#ccccff"><b>14<b></td><td  bgcolor="#ccccff"><b>3<b></td><td  bgcolor="#ccccff"><b>0<b></td><td  bgcolor="#ccccff"><b>0<b></td><td  bgcolor="#ccccff"><b>17</b></td></table></a></td><td width="50%">&nbsp;</td></tr></table><br><br><a name="Product Name: GeneConnect"><center><table bgColor="#CCCCFF" border="1" cellPadding="0" cellSpacing="0" width="700">  <tbody>    <tr>      <td align="middle" bgColor="#CCCCFF">        <p align="left"><font color="#660066"><b>Product Name: GeneConnect</b></font></p> </td> </tr> <tr> <td>&nbsp;</td> </tr> </tbody></table><br><table bgColor="#CCCCFF" border="1" cellPadding="0" cellSpacing="0" width="700"> <tbody> <tr> <td align="middle" bgColor="#CCCCFF"> <p align="left"><font color="#660066"><b>VITALS</b></font></p> </td> </tr> <tr> <td> <table bgColor="#CCCCFF" border="1" cellPadding="2" width="696"> <tbody> <tr bgColor="#ffffff"> <td vAlign="top" width="131"><b>Reference</b></td> <td vAlign="top" width="529">&nbsp;<a href="" target="_blank"></a></td> </tr> </tbody> </table> <tr> <td align="middle" bgColor="#CCCCFF"> <p align="left"><font color="#660066"><b>DESCRIPTION</b></font></p> </td> </tr> <tr> <td> <table bgColor="#CCCCFF" border="1" cellPadding="2" width="100%" height="55"> <tbody> <tr bgColor="#ffffff"> <td vAlign="top" height="1"></td> </tr> </tbody> </table> </td> </tr> </tbody></table>
<p align="right"><a href="#TMT Generated Test Plans">Top</a></p><a name="Version Name: GeneConnectver 1.1"><center><table bgColor="#CCCCFF" border="1" cellPadding="0" cellSpacing="0" width="700">  <tbody>    <tr>      <td align="middle" bgColor="#CCCCFF">        <p align="left"><font color="#660066"><b>Version Name: ver 1.1</b></font></p> </td> </tr> <tr> <td>&nbsp;</td> </tr> </tbody></table><br><table bgColor="#CCCCFF" border="1" cellPadding="0" cellSpacing="0" width="700"> <tbody> <tr> <td align="middle" bgColor="#CCCCFF"> <p align="left"><font color="#660066"><b>VITALS</b></font></p> </td> </tr> <tr> <td> <table bgColor="#CCCCFF" border="1" cellPadding="2" width="696"> <tbody> <tr bgColor="#ffffff"> <td vAlign="top" width="131"><b>Reference</b></td> <td vAlign="top" width="529">&nbsp;<a href="" target="_blank"></a></td> </tr> </tbody> </table> <tr> <td align="middle" bgColor="#CCCCFF"> <p align="left"><font color="#660066"><b>DESCRIPTION</b></font></p> </td> </tr> <tr> <td> <table bgColor="#CCCCFF" border="1" cellPadding="2" width="100%" height="55"> <tbody> <tr bgColor="#ffffff"> <td vAlign="top" height="1">First Version</td> </tr> </tbody> </table> </td> </tr> </tbody></table>
<p align="right"><a href="#TMT Generated Test Plans">Top</a></p><a name="Component Name: GeneConnectver 1.1Metadata Calculation"><center><table bgColor="#CCCCFF" border="1" cellPadding="0" cellSpacing="0" width="700">  <tbody>    <tr>      <td align="middle" bgColor="#CCCCFF">        <p align="left"><font color="#660066"><b>Component Name: Metadata Calculation</b></font></p> </td> </tr> <tr> <td>&nbsp;</td> </tr> </tbody></table><br><table bgColor="#CCCCFF" border="1" cellPadding="0" cellSpacing="0" width="700"> <tbody> <tr> <td align="middle" bgColor="#CCCCFF"> <p align="left"><font color="#660066"><b>VITALS</b></font></p> </td> </tr> <tr> <td> <table bgColor="#CCCCFF" border="1" cellPadding="2" width="696"> <tbody> <tr bgColor="#ffffff"> <td vAlign="top" width="131"><b>Reference</b></td> <td vAlign="top" width="529">&nbsp;<a href="" target="_blank"></a></td> </tr> </tbody> </table> <tr> <td align="middle" bgColor="#CCCCFF"> <p align="left"><font color="#660066"><b>DESCRIPTION</b></font></p> </td> </tr> <tr> <td> <table bgColor="#CCCCFF" border="1" cellPadding="2" width="100%" height="55"> <tbody> <tr bgColor="#ffffff"> <td vAlign="top" height="1"></td> </tr> </tbody> </table> </td> </tr> </tbody></table>
<p align="right"><a href="#TMT Generated Test Plans">Top</a></p><a name="Test Area Name: GeneConnectver 1.1Metadata CalculationMetadata Calculation"><center><table bgColor="#CCCCFF" border="1" cellPadding="0" cellSpacing="0" width="700">  <tbody>    <tr>      <td align="middle" bgColor="#CCCCFF">        <p align="left"><font color="#660066"><b>Test Area Name: Metadata Calculation</b></font></p> </td> </tr> <tr> <td>&nbsp;</td> </tr> </tbody></table><br><table bgColor="#CCCCFF" border="1" cellPadding="0" cellSpacing="0" width="700"> <tbody> <tr> <td align="middle" bgColor="#CCCCFF"> <p align="left"><font color="#660066"><b>VITALS</b></font></p> </td> </tr> <tr> <td> <table bgColor="#CCCCFF" border="1" cellPadding="2" width="696"> <tbody> <tr bgColor="#ffffff"> <td vAlign="top" width="131"><b>Reference</b></td> <td vAlign="top" width="529">&nbsp;<a href="" target="_blank"></a></td> </tr> </tbody> </table> <tr> <td align="middle" bgColor="#CCCCFF"> <p align="left"><font color="#660066"><b>DESCRIPTION</b></font></p> </td> </tr> <tr> <td> <table bgColor="#CCCCFF" border="1" cellPadding="2" width="100%" height="55"> <tbody> <tr bgColor="#ffffff"> <td vAlign="top" height="1"></td> </tr> </tbody> </table> </td> </tr> </tbody></table>
<p align="right"><a href="#TMT Generated Test Plans">Top</a></p><a name="Test Case ID: GeneConnectver 1.1Metadata CalculationMetadata CalculationMetadata Calculation"><center><table bgColor="#CCCCFF" border="1" cellPadding="0" cellSpacing="0" width="700">  <tbody>    <tr>      <td align="middle" bgColor="#CCCCFF">        <p align="left"><font color="#660066"><b>Test Case ID : 1893</b></font></p>      </td>    </tr>    <tr>      <td bgcolor="#FFFFFF"><b>Short Title : </b>Metadata Calculation</td>    </tr>    <tr>      <td bgcolor="#FFFFFF"><b>Title : </b>Testing Metadata Calculations</td>    </tr>  </tbody></table><br><table bgColor="#CCCCFF" border="1" cellPadding="0" cellSpacing="0" width="700">  <tbody>    <tr>      <td align="middle" bgColor="#CCCCFF">        <p align="left"><font color="#660066"><b>VITALS</b></font></p>      </td>    </tr>    <tr>      <td>        <table bgColor="#CCCCFF" border="1" cellPadding="2" width="100%">          <tbody>            <tr bgColor="#ffffff">              <td vAlign="top"><b>Product</b></td>              <td vAlign="top" colspan="3">GeneConnect</td>            <tr bgColor="#ffffff">              <td vAlign="top"><b>Component</b></td>              <td vAlign="top">Metadata Calculation<br>              </td>              <td vAlign="top"><b>Test Area</b></td>              <td vAlign="top">Metadata Calculation</td>            </tr>            <tr bgColor="#ffffff">              <td vAlign="top"><b>Reference</b></td>              <td vAlign="top">1</td>              <td vAlign="top"><b>Priority</b></td>              <td vAlign="top">P2</td>            </tr>          </tbody>        </table>    <tr>      <td align="middle" bgColor="#CCCCFF">        <p align="left"><font color="#660066"><b>DESCRIPTION</b></font></p>      </td>    </tr>    <tr>      <td>        <table bgColor="#CCCCFF" border="1" cellPadding="2" width="100%" height="113">          <tbody>            <tr bgColor="#ffffff"><td vAlign="top" height="32" colspan="4"><font class="itsmultlinefont"><pre>Purpose:
To check that the metadata tables are created and populated with the correct data.

Setup:
GeneConnect Server has been deployed on the machine (Deployment Target of the GeneConnect Build.xml has been run.) Please refer GeneConnectServer ReadMe for setting up the server.

Procedure:

1.Configure and Verify  the file GCGraph.txt in \config folder. (Refer GeneConnect Sever ReadMe for configuration details).
2.Configure and Verify  the file server.properties in \config folder.
3.Run the Deploy target of GeneConnect Build.xml or run calculateMetaData target if the server has been setup already and only metadata needs to be calculated.


Input/Data:
Configuration file representing GeneConnect graph (GCGraph.txt) located at \Config  folder.

Expected Output:

The following metadata tables must be loaded with correct data:
DATASOURCE
DATASOURCE_LINKS
BASETABLE_METADATA
COLUMN_METADATA
ONT
PATH
PATH_ONT
SUBPATH
TABLEJOIN_METADATA

Verification Logic:

This test case is written with reference to an example path, but can be applied to all paths/ONT's.
Consider the following path:
Path1: Ensembl Gene - Entrez Gene - RefSeqmRNA - Ensembl Transcript.

The Metadata tables can be verified as follows:
1.To verify that the table DATASOURCE is correctly loaded:
All the datasources as mentioned in the GCGraph.txt must be present in the table DATASOURCE and must have a unique DATASOURCE_ID.
Also corresponding CLASS and TABLE_NAME must be specified as GENE, MRNA or PROTEIN.

2.To verify that the table DATASOURCE_LINKS is correctly loaded:
(i)All the adjacent links (as per the GCGraph) are stored in the table DATASOURCE_LINKS. Thus , for the DataSource ,Ensembl Gene, all the adjacent links are :
Ensembl Gene - Entrez Gene
Ensembl Gene - UniGene
Ensembl Gene - Ensembl Transcript

(ii)All of these are represented in DATASOURCE_LINKS as follows:
For the 1st link Ensembl Gene - Entrez Gene, Ensembl Gene is the Source and Entrez Gene is the destination. The linktype is direct. Thus, the entry in the table would be SOURCE_DATASOURCE_ID as 0,TARGET_DATASOURCE_ID as 1 and LINK_TYPE_ID as 1.
Similar logic applies for all the links.


3. To verify that the table BASETABLE_METADATA is correctly loaded.
(i)Find the tables involved in path1 as follows
Table 1:ENSEMBLGENE_ENTREZGENE
Table 2:ENTREZGENE_REFSEQMRNA
Table 3:REFSEQMRNA_ENSEMBLTRANS

(ii)For Table 1:ENSEMBLGENE_ENTREZGENE, Ensembl Gene becomes the 'SOURCE data source' and Entrez Gene becomes the 'TARGET data source?. This is applied for all the tables.

(iii)For Table 1:ENSEMBLGENE_ENTREZGENE, find 'DATASOURCE_ID' from the table DATASOURCE for Ensembl Gene and Entrez Gene. These 'DATASOURCE_ID's must be present in the table BASETABLE_METADATA in the columns SOURCE_DATASOURCE_ID and TARGET_DATASOURCE_ID respectively corresponding to the table name 'ENSEMBLGENE_ENTREZGENE'.

Similar procedure must be followed for the tables ENTREZGENE_REFSEQMRNA and REFSEQMRNA_ENSEMBLTRANS.

4. To verify that the table COLUMN_METADATA is correctly loaded.

(i)For Table 1:ENSEMBLGENE_ENTREZGENE, find the following:
-TABLE_ID from the table BASETABLE_METADATA for the table ENSEMBLGENE_ENTREZGENE
-DATASOURCE_ID from DATASOURCE table for the two data sources in this table (Ensembl Gene and Entrez Gene).

(ii)These values must be present in the table COLUMN_METADATA in the columns TABLE_ID and DATASOURCE_ID respectively corresponding to the 3 columns (ESE_ENSEMBLGENEID, ESE_GENEID, ESE_LINKTYPE). 
Here, the TABLE_ID would be the same for all the 3 columns.
Note that ESE_LINKTYPE column would have no DATASOURCE_ID since it doesn?t represent a Data Source.

Similarly procedure must be followed for the tables ENTREZGENE_REFSEQMRNA and REFSEQMRNA_ENSEMBLTRANS.

5. To verify that the table TABLEJOIN_METADATA is correctly loaded.
(i)For the path1: Ensembl Gene - Entrez Gene - RefSeqmRNA - Ensembl Transcript, the 3 tables involved are: 
1. ENSEMBLGENE_ENTREZGENE
2. ENTREZGENE_REFSEQMRNA
3. REFSEQMRNA_ENSEMBLTRANS

(ii)Here the tables 1 and 2 have ENTREZGENE column in common. While relating table 1 and 2, table1 becomes Source table and table 2 becomes Target table. 
Find the following:
-TABLE_ID for Source and Target table from the table BASETABLE_METADATA
-Find the COLUMN_ID from table COLUMN_METADATA for the column which is common (Here Entrez Gene) in both the tables (1 and 2).

(iii)These values for the TABLE_ID for Source and Target table must be present in the table TABLEJOIN_METADATA as columns SOURCE_TABLE_ID and TARGET_TABLE_ID respectively.
The values for COLUMN_ID for table 1 and table 2 must be present in the table TABLEJOIN_METADATA as columns SOURCE_COLUMN_ID and TARGET_COLUMN_ID respectively.

Similar procedure must be followed for the tables ENTREZGENE_REFSEQMRNA and REFSEQMRNA_ENSEMBLTRANS.

6. To verify that the table PATH is correctly loaded: 
(i)For each of the Data Source involved in the path1, find the DATASOURCE_ID from the table DATASOURCE as follows:
Ensembl Gene - Entrez Gene - RefSeqmRNA - Ensembl Transcript.
Path is: 0_1_5_4

(ii)This path should be present in table PATH with 
SOURCE_DATASOURCE_ID as 0
TARGET_DATASOURCE_ID as 4
PATH (intermediate nodes) as 1_5

7. To verify that the table ONT and PATH_ONT is correctly loaded:
(i)Find the PATH_ID  for the path 0_1_5_4 from table PATH.

(ii)Find associated ONT_ID from PATH_ONT table for the path 0_1_5_4.
Lets assume the values 28, 32 as ONT_ID?s . 

(iii)For PATH_ID 28 and 32, from table ONT, the corresponding SOURCE_DS_ID must have 0, and the subsequent records must have the SOURCE_DS_ID as 1, 5, and 4 respectively indicating the data sources in the path.

(iv)The difference in the PATH_ID's 28 and 32 is depending on the link types involved in the same path. This difference is shown by the column LINKTYPE_ID, which has value 2 (inferred link) for link between SOURCE_DS_ID's 5 and 4 for PATH_ID 28 and value 4 (BLAST Identity link) for PATH_ID 32.

(The values for link types are referred from the table LINK_TYPE with columns LINK_TYPE_ID and LINK_TYPE_NAME.)


8. To verify that the table SUBPATH  is correctly loaded:
(i)All the sub paths of  a particular path can be referred in table SUBPATH. For PATH_ID 28, find the corresponding SUBPATH_ID in table SUBPATH. Here, for PATH 28, the SUBPATH_ID's are : 34,48,78,84,93,331,341,427. 

(ii)These ID's indicate the sub paths which can be verified as follows: 
Check the SUBPATH_ID value (34) in PATH table, in column PATH_ID.This entry shows the sub path 0-1, indicated by 0 in SOURCE_DATASOURCE_ID and 1 in TARGET_DATASOURCE_ID.

(iii)Similarly for PATH_ID : 48 in PATH table in column PATH_ID , shows the sub path 0-1-5 , indicated by 0 in SOURCE_DATASOURCE_ID, 5 in TARGET_DATASOURCE_ID and 1 in PATH (intermediate node).

All the other sub paths can be verified with similar logic.

</pre></font></td>            </tr>            <tr bgColor="#ffffff">              <td vAlign="top" height="21"><b>Mode of Execution</b></td>              <td vAlign="top" height="1">Manual                </td>              <td vAlign="top" height="21"><b>Verification Logic</b></td>              <td vAlign="top" height="1">BOTH<br>                </td>            </tr>            <tr bgColor="#ffffff">              <td vAlign="top" height="21"><b>Author</b></td>              <td vAlign="top" height="21">Shweta Chavan</td>              <td vAlign="top" height="21"><b>Date Entered</b></td>              <td vAlign="top" height="21">03/10/2006&nbsp;</td>            </tr>            <tr bgColor="#ffffff">              <td vAlign="top" height="21"></td>              <td vAlign="top" height="21"></td>              <td vAlign="top" height="21"><b>Date Last Update</b></td>              <td vAlign="top" height="21">04/12/2006&nbsp;</td>            </tr>          </tbody>        </table>      </td>    </tr>  </tbody></table>
<p align="right"><a href="#TMT Generated Test Plans">Top</a></p><a name="Test Case ID: GeneConnectver 1.1Metadata CalculationMetadata CalculationMetadata Calculation for newly added DataSource"><center><table bgColor="#CCCCFF" border="1" cellPadding="0" cellSpacing="0" width="700">  <tbody>    <tr>      <td align="middle" bgColor="#CCCCFF">        <p align="left"><font color="#660066"><b>Test Case ID : 1896</b></font></p>      </td>    </tr>    <tr>      <td bgcolor="#FFFFFF"><b>Short Title : </b>Metadata Calculation for newly added DataSource</td>    </tr>    <tr>      <td bgcolor="#FFFFFF"><b>Title : </b>Testing Metadata Calculations for newly added DataSource</td>    </tr>  </tbody></table><br><table bgColor="#CCCCFF" border="1" cellPadding="0" cellSpacing="0" width="700">  <tbody>    <tr>      <td align="middle" bgColor="#CCCCFF">        <p align="left"><font color="#660066"><b>VITALS</b></font></p>      </td>    </tr>    <tr>      <td>        <table bgColor="#CCCCFF" border="1" cellPadding="2" width="100%">          <tbody>            <tr bgColor="#ffffff">              <td vAlign="top"><b>Product</b></td>              <td vAlign="top" colspan="3">GeneConnect</td>            <tr bgColor="#ffffff">              <td vAlign="top"><b>Component</b></td>              <td vAlign="top">Metadata Calculation<br>              </td>              <td vAlign="top"><b>Test Area</b></td>              <td vAlign="top">Metadata Calculation</td>            </tr>            <tr bgColor="#ffffff">              <td vAlign="top"><b>Reference</b></td>              <td vAlign="top">1</td>              <td vAlign="top"><b>Priority</b></td>              <td vAlign="top">P2</td>            </tr>          </tbody>        </table>    <tr>      <td align="middle" bgColor="#CCCCFF">        <p align="left"><font color="#660066"><b>DESCRIPTION</b></font></p>      </td>    </tr>    <tr>      <td>        <table bgColor="#CCCCFF" border="1" cellPadding="2" width="100%" height="113">          <tbody>            <tr bgColor="#ffffff"><td vAlign="top" height="32" colspan="4"><font class="itsmultlinefont"><pre>Purpose:
To test whether the newly added Data Source is correctly represented in the relevant Metadata Tables.

Setup:
1.GeneConnect Server has been deployed on the machine (Deployment Target of the GeneConnect Build.xml has been run).

Procedure:

1.Check that the added Data Source has been added in the file GCGraph.txt in \Config folder.
2.Verify that the adjacency matrix is calculated accordingly.
Each entry in the matrix represents the link types between two Data Sources. For e.g if we are calculating the matrix value for Ensembl Protein to UniprotKB, then 
Link types = Alignment based + Inferred = 12 + 2 = 14.
Thus there should be value 14 for in the adjacency matrix for the directional link from Ensembl Protein to UniprotKB.
Similarly the adjaceny matrix values can be calculated for added data source depending on the links by which it is related to other data sources.
Legend:
Alignment based-&gt;12
Direct-&gt;1
Inferred-&gt;2
3.Verify  that the file is present in GCGraph.txt in \config folder.
4.Configure and Verify  the file server.properties in \config folder.
5.Modify the SQL scripts GCSchemaCreation.sql for defining the base tables for new data source and GCMetadataPopulation.sql  to define the metadata for those base tables.
Also modify sql scripts in \script folder for adding the column in GENE, MRNA or PROTEIN table.
6.Run the Deploy target of GeneConnect Build.xml.

Input/Data:
Configuration file representing GeneConnect graph (GCGraph.txt) located in \Config  folder.This file must also have the added Data Source.

Expected Output:
The following metadata tables must be loaded with correct data including the data for the added Data Source:
DATASOURCE
DATASOURCE_LINKS
BASETABLE_METADATA
COLUMN_METADATA
ONT
PATH
PATH_ONT
SUBPATH
TABLEJOIN_METADATA

Verification Logic:


1.Verify the entries for the newly added data source in the following tables:
(i) The added DataSource must be present in the table DATASOURCE with a unique DATASOURCE_ID. Other attributes like DATASOURCE_NAME, CLASS, ATTRIBUTE, ATTRIBUTE_TYPE, TABLE_NAME, and COLUMN_NAME should be specified. 

(ii).Depending on the Data Source ,corresponding CLASS and TABLE_NAME is specified as Gene, mRNA or Protein in the table DATASOURCE_ID.

2.Correspondingly verify the other tables by applying the verification logic described in the  test case 'Metadata Calculation' .



</pre></font></td>            </tr>            <tr bgColor="#ffffff">              <td vAlign="top" height="21"><b>Mode of Execution</b></td>              <td vAlign="top" height="1">Manual                </td>              <td vAlign="top" height="21"><b>Verification Logic</b></td>              <td vAlign="top" height="1">BOTH<br>                </td>            </tr>            <tr bgColor="#ffffff">              <td vAlign="top" height="21"><b>Author</b></td>              <td vAlign="top" height="21">Shweta Chavan</td>              <td vAlign="top" height="21"><b>Date Entered</b></td>              <td vAlign="top" height="21">04/10/2006&nbsp;</td>            </tr>            <tr bgColor="#ffffff">              <td vAlign="top" height="21"></td>              <td vAlign="top" height="21"></td>              <td vAlign="top" height="21"><b>Date Last Update</b></td>              <td vAlign="top" height="21">04/12/2006&nbsp;</td>            </tr>          </tbody>        </table>      </td>    </tr>  </tbody></table>
<p align="right"><a href="#TMT Generated Test Plans">Top</a></p></body></html>
