<head>
<META content="Microsoft FrontPage 4.0" name=GENERATOR>
<STYLE type=text/css>BODY {
	BACKGROUND-COLOR: #ffffff; COLOR: #000000; FONT-FAMILY: "Times New Roman"; FONT-SIZE: 12pt; TEXT-ALIGN: left
}
 UL.zzul {
	DISPLAY: block; LIST-STYLE: none}
SPAN.zzspace {
	LEFT: 11px; POSITION: relative
}
 A {
	COLOR: #0000ff
}
 A:visited {
	COLOR: #0000ff
}
 A:hover {
	COLOR: #ff0000
}
TD {
	COLOR: #000000; FONT-FAMILY: "Times New Roman"; FONT-SIZE: 12pt; TEXT-ALIGN: left
}
 TABLE {
	BORDER-BOTTOM: medium none; BORDER-LEFT: medium none; BORDER-RIGHT: medium none; BORDER-TOP: medium none; PADDING-BOTTOM: 3px; PADDING-LEFT: 3px; PADDING-RIGHT: 3px; PADDING-TOP: 3px; WIDTH: 100%
 }
 </STYLE>
 <SCRIPT type=text/javascript>
 <!--
 function exp_coll(ind)
 {
 s = document.getElementById("sp_" + ind);
 i = document.getElementById("im_" + ind);
 i2 = document.getElementById("im2_" + ind);
 if (s.style.display == 'none')
 {
  s.style.display = 'block';   
 }
 else if (s.style.display == 'block')
 {
   s.style.display = 'none';   
 }
}
 function coll_all(i)
 {  
 for(ind=1;ind<=i;ind++){
  x = document.getElementById("sp_" + ind);
  x.style.display = 'none';
  if (x.style.display == 'none')
  {    
 }
}
}
 function exp_all(i)
 {
  for(ind=1;ind<=i;ind++){
  x = document.getElementById("sp_" + ind);
  x.style.display = 'block';
  if (x.style.display == 'block')
  {    
 }
  }  
}
 -->
</SCRIPT>
</HEAD>
<BODY>
<center><a name="TMT Generated Test Plans"><h1>Test Plan</h1></a></center><hr><a name="xys"><table><tr><td>This is a TMT (Test Management Tool) generated Test Plan. To read a short note on TMT on the caBIG Best Practices SIG website, click <a href="http://cabig-bpsigwiki.bioinformatics.northwestern.edu/bpsigwiki/index.php/Manual_Testing_-_Test_Management_Tool">here</a>.<br> For more details or if you find any issue with this report, please send an email to <a href="mailto:tmt@persistent.co.in">tmt@persistent.co.in</a>.</td></tr></table></a><hr><br><br>
<A href="javascript:exp_all(4);">Expand All</A> &nbsp; <A href="javascript:coll_all(4);">Collapse All</A> 

<UL class=zzul>
  <LI>
<LI><A href="javascript:exp_coll(1);">+</A> <A 
	  href="javascript:exp_coll(1);"></A> &nbsp; <A  href="#Product Name: GeneConnect">GeneConnect</A> 
<UL class=zzul id=sp_1 style="DISPLAY: block">
<SCRIPT type=text/javascript>
 ind = 1; 
 x = document.getElementById("sp_" + ind);
 x.style.display = 'none'; 
 if (x.style.display == 'none'){ }
</SCRIPT>
<LI><A href="javascript:exp_coll(2);">+</A> <A 
	  href="javascript:exp_coll(2);"></A> &nbsp; <A  href="#Version Name: GeneConnectver 1.1">ver 1.1</A> 
<UL class=zzul id=sp_2 style="DISPLAY: block">
<SCRIPT type=text/javascript>
 ind = 2; 
 x = document.getElementById("sp_" + ind);
 x.style.display = 'none'; 
 if (x.style.display == 'none'){ }
</SCRIPT>
<LI><A href="javascript:exp_coll(3);">+</A> <A 
	  href="javascript:exp_coll(3);"></A> &nbsp; <A  href="#Component Name: GeneConnectver 1.1Calculation of all-to-all genomic identifier links">Calculation of all-to-all genomic identifier links</A> 
<UL class=zzul id=sp_3 style="DISPLAY: block">
<SCRIPT type=text/javascript>
 ind = 3; 
 x = document.getElementById("sp_" + ind);
 x.style.display = 'none'; 
 if (x.style.display == 'none'){ }
</SCRIPT>
<LI><A href="javascript:exp_coll(4);">+</A> <A 
	  href="javascript:exp_coll(4);"></A> &nbsp; <A  href="#Test Area Name: GeneConnectver 1.1Calculation of all-to-all genomic identifier linksCalculation of all-to-all genomic identifier links">Calculation of all-to-all genomic identifier links</A> 
<UL class=zzul id=sp_4 style="DISPLAY: block">
<SCRIPT type=text/javascript>
 ind = 4; 
 x = document.getElementById("sp_" + ind);
 x.style.display = 'none'; 
 if (x.style.display == 'none'){ }
</SCRIPT>
<LI><SPAN class=zzspace>&nbsp;&nbsp;
<A href="#Test Case ID: GeneConnectver 1.1Calculation of all-to-all genomic identifier linksCalculation of all-to-all genomic identifier linksSummary Table Calculation">Summary Table Calculation</A> </SPAN>
<LI><SPAN class=zzspace>&nbsp;&nbsp;
<A href="#Test Case ID: GeneConnectver 1.1Calculation of all-to-all genomic identifier linksCalculation of all-to-all genomic identifier linksSummary Table Calculation for a newly added Data S">Summary Table Calculation for a newly added Data S</A> </SPAN>
</LI></UL>

</LI></UL>

</LI></UL>
<br><br><a><table><tr><td width="50%"><table bgColor="#CCCCFF" border="1" cellPadding="2" width="100"><tr><td colspan="10"><font color="#660066"><b>Test case count: Component verses Priority</b></font></td></tr><tr><td  align="center" bgcolor="#ccccff"><font color="#660066"><b>Component Names</b></font></td><td align="center" bgcolor="#ccccff"> <font color="#660066"><b>P1</b></font></td><td align="center" bgcolor="#ccccff"> <font color="#660066"><b>P2</b></font></td><td align="center" bgcolor="#ccccff"> <font color="#660066"><b>P3</b></font></td><td align="center" bgcolor="#ccccff"> <font color="#660066"><b>P4</b></font></td><td align="center" bgcolor="#ccccff"> <font color="#660066"><b>Total</b></td></tr><tr><td bgcolor="#ffffff">ver 1.1</td><td  bgcolor="#ffffff">14</td><td  bgcolor="#ffffff">1</td><td  bgcolor="#ffffff">0</td><td  bgcolor="#ffffff">0</td><td bgcolor="#ffffff">15</td></tr><tr><td bgcolor="#ffffff">Calculation of all-to-all genomic identifier links</td><td  bgcolor="#ffffff">0</td><td  bgcolor="#ffffff">2</td><td  bgcolor="#ffffff">0</td><td  bgcolor="#ffffff">0</td><td bgcolor="#ffffff">2</td></tr><tr><td bgcolor="#ccccff"><b>Column Total</b></td><td  bgcolor="#ccccff"><b>14<b></td><td  bgcolor="#ccccff"><b>3<b></td><td  bgcolor="#ccccff"><b>0<b></td><td  bgcolor="#ccccff"><b>0<b></td><td  bgcolor="#ccccff"><b>17</b></td></table></a></td><td width="50%">&nbsp;</td></tr></table><br><br><a name="Product Name: GeneConnect"><center><table bgColor="#CCCCFF" border="1" cellPadding="0" cellSpacing="0" width="700">  <tbody>    <tr>      <td align="middle" bgColor="#CCCCFF">        <p align="left"><font color="#660066"><b>Product Name: GeneConnect</b></font></p> </td> </tr> <tr> <td>&nbsp;</td> </tr> </tbody></table><br><table bgColor="#CCCCFF" border="1" cellPadding="0" cellSpacing="0" width="700"> <tbody> <tr> <td align="middle" bgColor="#CCCCFF"> <p align="left"><font color="#660066"><b>VITALS</b></font></p> </td> </tr> <tr> <td> <table bgColor="#CCCCFF" border="1" cellPadding="2" width="696"> <tbody> <tr bgColor="#ffffff"> <td vAlign="top" width="131"><b>Reference</b></td> <td vAlign="top" width="529">&nbsp;<a href="" target="_blank"></a></td> </tr> </tbody> </table> <tr> <td align="middle" bgColor="#CCCCFF"> <p align="left"><font color="#660066"><b>DESCRIPTION</b></font></p> </td> </tr> <tr> <td> <table bgColor="#CCCCFF" border="1" cellPadding="2" width="100%" height="55"> <tbody> <tr bgColor="#ffffff"> <td vAlign="top" height="1"></td> </tr> </tbody> </table> </td> </tr> </tbody></table>
<p align="right"><a href="#TMT Generated Test Plans">Top</a></p><a name="Version Name: GeneConnectver 1.1"><center><table bgColor="#CCCCFF" border="1" cellPadding="0" cellSpacing="0" width="700">  <tbody>    <tr>      <td align="middle" bgColor="#CCCCFF">        <p align="left"><font color="#660066"><b>Version Name: ver 1.1</b></font></p> </td> </tr> <tr> <td>&nbsp;</td> </tr> </tbody></table><br><table bgColor="#CCCCFF" border="1" cellPadding="0" cellSpacing="0" width="700"> <tbody> <tr> <td align="middle" bgColor="#CCCCFF"> <p align="left"><font color="#660066"><b>VITALS</b></font></p> </td> </tr> <tr> <td> <table bgColor="#CCCCFF" border="1" cellPadding="2" width="696"> <tbody> <tr bgColor="#ffffff"> <td vAlign="top" width="131"><b>Reference</b></td> <td vAlign="top" width="529">&nbsp;<a href="" target="_blank"></a></td> </tr> </tbody> </table> <tr> <td align="middle" bgColor="#CCCCFF"> <p align="left"><font color="#660066"><b>DESCRIPTION</b></font></p> </td> </tr> <tr> <td> <table bgColor="#CCCCFF" border="1" cellPadding="2" width="100%" height="55"> <tbody> <tr bgColor="#ffffff"> <td vAlign="top" height="1">First Version</td> </tr> </tbody> </table> </td> </tr> </tbody></table>
<p align="right"><a href="#TMT Generated Test Plans">Top</a></p><a name="Component Name: GeneConnectver 1.1Calculation of all-to-all genomic identifier links"><center><table bgColor="#CCCCFF" border="1" cellPadding="0" cellSpacing="0" width="700">  <tbody>    <tr>      <td align="middle" bgColor="#CCCCFF">        <p align="left"><font color="#660066"><b>Component Name: Calculation of all-to-all genomic identifier links</b></font></p> </td> </tr> <tr> <td>&nbsp;</td> </tr> </tbody></table><br><table bgColor="#CCCCFF" border="1" cellPadding="0" cellSpacing="0" width="700"> <tbody> <tr> <td align="middle" bgColor="#CCCCFF"> <p align="left"><font color="#660066"><b>VITALS</b></font></p> </td> </tr> <tr> <td> <table bgColor="#CCCCFF" border="1" cellPadding="2" width="696"> <tbody> <tr bgColor="#ffffff"> <td vAlign="top" width="131"><b>Reference</b></td> <td vAlign="top" width="529">&nbsp;<a href="" target="_blank"></a></td> </tr> </tbody> </table> <tr> <td align="middle" bgColor="#CCCCFF"> <p align="left"><font color="#660066"><b>DESCRIPTION</b></font></p> </td> </tr> <tr> <td> <table bgColor="#CCCCFF" border="1" cellPadding="2" width="100%" height="55"> <tbody> <tr bgColor="#ffffff"> <td vAlign="top" height="1"></td> </tr> </tbody> </table> </td> </tr> </tbody></table>
<p align="right"><a href="#TMT Generated Test Plans">Top</a></p><a name="Test Area Name: GeneConnectver 1.1Calculation of all-to-all genomic identifier linksCalculation of all-to-all genomic identifier links"><center><table bgColor="#CCCCFF" border="1" cellPadding="0" cellSpacing="0" width="700">  <tbody>    <tr>      <td align="middle" bgColor="#CCCCFF">        <p align="left"><font color="#660066"><b>Test Area Name: Calculation of all-to-all genomic identifier links</b></font></p> </td> </tr> <tr> <td>&nbsp;</td> </tr> </tbody></table><br><table bgColor="#CCCCFF" border="1" cellPadding="0" cellSpacing="0" width="700"> <tbody> <tr> <td align="middle" bgColor="#CCCCFF"> <p align="left"><font color="#660066"><b>VITALS</b></font></p> </td> </tr> <tr> <td> <table bgColor="#CCCCFF" border="1" cellPadding="2" width="696"> <tbody> <tr bgColor="#ffffff"> <td vAlign="top" width="131"><b>Reference</b></td> <td vAlign="top" width="529">&nbsp;<a href="" target="_blank"></a></td> </tr> </tbody> </table> <tr> <td align="middle" bgColor="#CCCCFF"> <p align="left"><font color="#660066"><b>DESCRIPTION</b></font></p> </td> </tr> <tr> <td> <table bgColor="#CCCCFF" border="1" cellPadding="2" width="100%" height="55"> <tbody> <tr bgColor="#ffffff"> <td vAlign="top" height="1"></td> </tr> </tbody> </table> </td> </tr> </tbody></table>
<p align="right"><a href="#TMT Generated Test Plans">Top</a></p><a name="Test Case ID: GeneConnectver 1.1Calculation of all-to-all genomic identifier linksCalculation of all-to-all genomic identifier linksSummary Table Calculation"><center><table bgColor="#CCCCFF" border="1" cellPadding="0" cellSpacing="0" width="700">  <tbody>    <tr>      <td align="middle" bgColor="#CCCCFF">        <p align="left"><font color="#660066"><b>Test Case ID : 1892</b></font></p>      </td>    </tr>    <tr>      <td bgcolor="#FFFFFF"><b>Short Title : </b>Summary Table Calculation</td>    </tr>    <tr>      <td bgcolor="#FFFFFF"><b>Title : </b>Testing Summary Table Calculations</td>    </tr>  </tbody></table><br><table bgColor="#CCCCFF" border="1" cellPadding="0" cellSpacing="0" width="700">  <tbody>    <tr>      <td align="middle" bgColor="#CCCCFF">        <p align="left"><font color="#660066"><b>VITALS</b></font></p>      </td>    </tr>    <tr>      <td>        <table bgColor="#CCCCFF" border="1" cellPadding="2" width="100%">          <tbody>            <tr bgColor="#ffffff">              <td vAlign="top"><b>Product</b></td>              <td vAlign="top" colspan="3">GeneConnect</td>            <tr bgColor="#ffffff">              <td vAlign="top"><b>Component</b></td>              <td vAlign="top">Calculation of all-to-all genomic identifier links<br>              </td>              <td vAlign="top"><b>Test Area</b></td>              <td vAlign="top">Calculation of all-to-all genomic identifier links</td>            </tr>            <tr bgColor="#ffffff">              <td vAlign="top"><b>Reference</b></td>              <td vAlign="top">1</td>              <td vAlign="top"><b>Priority</b></td>              <td vAlign="top">P2</td>            </tr>          </tbody>        </table>    <tr>      <td align="middle" bgColor="#CCCCFF">        <p align="left"><font color="#660066"><b>DESCRIPTION</b></font></p>      </td>    </tr>    <tr>      <td>        <table bgColor="#CCCCFF" border="1" cellPadding="2" width="100%" height="113">          <tbody>            <tr bgColor="#ffffff"><td vAlign="top" height="32" colspan="4"><font class="itsmultlinefont"><pre>Purpose:
To check whether the Summary tables are correctly populated.

Setup:
1. GC Server has been deployed.
2. MetaData tables have been populated.
3. All the Base tables have been populated or should be populated as a part of this server run.

Procedure:
1. Configure and Verify  the file server.properties in \config folder.
2. Run the GC Server using Run target of Build.xml.

Summary Table Verification Procedure:

1.To verify that the tables GENE, MRNA and PROTEIN are correctly loaded:

(i)Consider the table GENE, this table must have Ensembl Gene ID, Entrez Gene ID and Unigene ID columns. All the ID's in a particular row must belong to a single locus.
Similarly, in the table MRNA (with columns Ensembl Transcript ID, GenBank mRNA ID ,RefSeqmRNA ID) and in table PROTEIN (with columns Ensembl Peptide ID , RefSeq Protien ID, UniprotKB ID and GenBank Protein ID) in a single row must belong to a single locus.

(ii)The correctness of the data loaded can be checked by manually browsing through the Online database (explained in Verification Logic). 

(iii)Each of the tables GENE, MRNA and PROTEIN must have GENE_ID,MRNA_ID and PROTEIN_ID as a unique identifier for a particular row.

2.To verify that the tables GENE_MRNA,MRNA_PROTEIN and PROTEIN_GENE are correctly populated:

(i)In the table GENE_MRNA, the GENE_ID from the GENE table and the MRNA_ID from the MRNA table are related and each such relation/association is shown by a particular row.
Thus if GENE_ID 1 from GENE table is related to MRNA_ID 1 and 2 in MRNA table, then GENE_MRNA table must have the following two rows:
GENE_ID	MRNA_ID
1	1
1	2

(ii)This indicates that all the Genomic Identifiers in the 1st row of GENE table and that of the 1st row of MRNA table belong to the same locus.
Since GENE_ID 1 is also related to MRNA_ID 2, all the Genomic Identifiers in the 1st row of GENE table and that of the 2nd row of MRNA table belong to the same locus.

Similar logic applies to MRNA_PROTEIN and PROTEIN_GENE also.

3.To verify that the table  GENOMIC_IDENTIFIER_SET: is correctly populated:
It has the columns :SET_ID, GENE_ID, MRNA_ID, PROTEIN_ID, SOLUTION_ID
The SET_ID is the unique identifier for all the rows in this table and indicates the association between GENE_ID, MRNA_ID, PROTEIN_ID from the tables GENE, MRNA and PROTEIN respectively.

For e.g. if the table GENOMIC_IDENTIFIER_SET has following data:
SET_ID	GENE_ID	MRNA_ID	PROTEIN_ID  SOLUTION_ID
1	1	1	1	    null
2	1	2	1	    null

For SET_ID 2, 
It shows that GENE_ID 1, MRNA_ID 2 and PROTEIN ID 1 are associated.
This can be verified as follows:
(i)Check in table GENE_MRNA, whether GENE_ID 1 and MRNA_ID 2 are related,i.e. present in the same row.
(ii)Check in table MRNA_PROTEIN, whether MRNA_ID 2 and PROTEIN_ID 1 are related,i.e. present in the same row.
(iii)Check in table PROTEIN_GENE, whether PROTEIN_ID 1 and GENE_ID 1 are related,i.e. present in the same row.

If all the three (GENE_ID,MRNA_ID and PROTEIN_ID) are found to be related then only the data mentioned in the GENOMIC_IDENTIFIER_SET table is correct.

4. To verify that the table SET_ONT is correctly populated:
(i)If a particular SET_ID involves the datasources as follows:
Ensembl Gene ?Entrez Gene-RefSeqmRNA-RefSeqProtein
For this set of DataSources the possible ONT?s would be:
Ensembl Gene ?Entrez Gene-RefSeqmRNA-RefSeqProtein
RefSeqProtein- RefSeqmRNA- Entrez Gene- Ensembl Gene
RefSeqmRNA- Entrez Gene- Ensembl Gene
etc.

(ii)Thus in table SET_ONT, for that SET_ID, all the PATH_ID  (PATH_ID from table ONT ) corresponding to all the paths mentioned above must be present.

Input/Data:
1.All the base tables and MetaData tables must be poplulated.

Expected Output:
All the following tables are populated with correct data:
GENE
MRNA
PROTEIN 
GENE_MRNA 
MRNA_PROTEIN 
PROTEIN_GENE
GENOMIC_IDENTIFIER_SET
SET_ONT

Verification Logic:

1. For a particular SET_ID from GENOMIC_IDENTIFIER_SET , find out the GENE_ID,MRNA_ID and the PROTEIN_ID found to be associated .

2.For the GENE_ID (e.g.: 1) find the actual genomic identifiers in the table GENE corresponding to each of the Data Sources viz., Ensembl Gene ID, Entrez Gene ID and Unigene ID.

3.Go to the page http://www.ensembl.org/index.html. For the Ensembl Gene ID , ( as mentioned in the table GENE) find the links to Entrez Gene and Unigene . Check whether the genomic identifiers mentioned here and those in the GENE table are same for Entrez Gene and Unigene.

Similar logic must be applied for MRNA and PROTEIN also.
This verifies whether the data in the tables GENE,MRNA and PROTEIN are correctly loaded.

4.For the MRNA_ID (e.g.: 2) find the actual genomic identifiers in the table MRNA corresponding to each of the Data Sources viz., Ensembl Transcript, GenBank accession and RefSeqmrna.

5.Go to the page http://www.ensembl.org/index.html. For the Ensembl Gene ID , find the links to Ensembl Transcript, GenBank accession and RefSeqmrna. Check whether the genomic identifiers mentioned here and those in the MRNA table are same for Ensembl Transcript, GenBank accession and RefSeqmrna.

Similar logic must be applied for verification of :
Entrez Gene ID to :Ensembl Transcript, GenBank accession and RefSeqmrna.
Unigene ID to :Ensembl Transcript, GenBank accession and RefSeqmrna.
This verifies that all the genomic identifiers in the GENE table are associated with that of MRNA table for a particular SET_ID.

Thus, this verifies that the table GENE_MRNA is correctly loaded.
The tables MRNA_PROTEIN and PROTEIN_GENE must be verified with a similar logic.


















</pre></font></td>            </tr>            <tr bgColor="#ffffff">              <td vAlign="top" height="21"><b>Mode of Execution</b></td>              <td vAlign="top" height="1">Manual                </td>              <td vAlign="top" height="21"><b>Verification Logic</b></td>              <td vAlign="top" height="1">BOTH<br>                </td>            </tr>            <tr bgColor="#ffffff">              <td vAlign="top" height="21"><b>Author</b></td>              <td vAlign="top" height="21">Shweta Chavan</td>              <td vAlign="top" height="21"><b>Date Entered</b></td>              <td vAlign="top" height="21">03/10/2006&nbsp;</td>            </tr>            <tr bgColor="#ffffff">              <td vAlign="top" height="21"></td>              <td vAlign="top" height="21"></td>              <td vAlign="top" height="21"><b>Date Last Update</b></td>              <td vAlign="top" height="21">04/10/2006&nbsp;</td>            </tr>          </tbody>        </table>      </td>    </tr>  </tbody></table>
<p align="right"><a href="#TMT Generated Test Plans">Top</a></p><a name="Test Case ID: GeneConnectver 1.1Calculation of all-to-all genomic identifier linksCalculation of all-to-all genomic identifier linksSummary Table Calculation for a newly added Data S"><center><table bgColor="#CCCCFF" border="1" cellPadding="0" cellSpacing="0" width="700">  <tbody>    <tr>      <td align="middle" bgColor="#CCCCFF">        <p align="left"><font color="#660066"><b>Test Case ID : 1895</b></font></p>      </td>    </tr>    <tr>      <td bgcolor="#FFFFFF"><b>Short Title : </b>Summary Table Calculation for a newly added Data S</td>    </tr>    <tr>      <td bgcolor="#FFFFFF"><b>Title : </b>Testing Summary Table Calculations for a newly added Data Source</td>    </tr>  </tbody></table><br><table bgColor="#CCCCFF" border="1" cellPadding="0" cellSpacing="0" width="700">  <tbody>    <tr>      <td align="middle" bgColor="#CCCCFF">        <p align="left"><font color="#660066"><b>VITALS</b></font></p>      </td>    </tr>    <tr>      <td>        <table bgColor="#CCCCFF" border="1" cellPadding="2" width="100%">          <tbody>            <tr bgColor="#ffffff">              <td vAlign="top"><b>Product</b></td>              <td vAlign="top" colspan="3">GeneConnect</td>            <tr bgColor="#ffffff">              <td vAlign="top"><b>Component</b></td>              <td vAlign="top">Calculation of all-to-all genomic identifier links<br>              </td>              <td vAlign="top"><b>Test Area</b></td>              <td vAlign="top">Calculation of all-to-all genomic identifier links</td>            </tr>            <tr bgColor="#ffffff">              <td vAlign="top"><b>Reference</b></td>              <td vAlign="top">1</td>              <td vAlign="top"><b>Priority</b></td>              <td vAlign="top">P2</td>            </tr>          </tbody>        </table>    <tr>      <td align="middle" bgColor="#CCCCFF">        <p align="left"><font color="#660066"><b>DESCRIPTION</b></font></p>      </td>    </tr>    <tr>      <td>        <table bgColor="#CCCCFF" border="1" cellPadding="2" width="100%" height="113">          <tbody>            <tr bgColor="#ffffff"><td vAlign="top" height="32" colspan="4"><font class="itsmultlinefont"><pre>Purpose:
To test whether the newly added Data Source is correctly represented in the relevant Summary Tables.

Setup:
1. GC Server has been deployed.
2. MetaData tables have been populated.
3. Base tables for newly added data source have been populated or should be populated as a part of this server run.

Procedure:
1.Configure and Verify  the file server.properties in \config folder.
2. Run the GC Server using Run target of Build.xml.

Input/Data:
1. All the base tables and MetaData tables must be poplulated .

Expected Output:
1.All the following tables are populated with correct data:
GENE
MRNA
PROTEIN 
GENE_MRNA 
MRNA_PROTEIN 
PROTEIN_GENE
GENOMIC_IDENTIFIER_SET
SET_ONT
2. Depending upon the Class of the DataSource , MRNA or GENE or PROTEIN table must have data for the newly added  DataSource. 

Verification Logic:

1. If the DataSource added is a mRNA DataSource (e.g NEW_MRNA), then the table MRNA must have a column for the Data Source NEW_MRNA . The identifier for  NEW_MRNA in a particular row must represent the identifier for the same locus as indicated by the other Genomic Identifiers in that row.

2. For a particular SET_ID from GENOMIC_IDENTIFIER_SET , find out the MRNA_ID/PROTEIN_ID/GENE_ID (Here, MNRA_ID) found to be associated .

3. For the MRNA_ID find the actual genomic identifiers in the table MRNA corresponding to each of the Data Sources viz., Ensembl Transcript, GenBank accession , RefSeqmrna and NEW_MRNA. Check if the Genomic Identifier for NEW_MRNA is present.

4. Verify from the website if all the above four genomic identifiers belong to the same locus. This can be done by browsing the website of one DataSource (e.g. Ensembl) and finding the links to other three DataSources.





</pre></font></td>            </tr>            <tr bgColor="#ffffff">              <td vAlign="top" height="21"><b>Mode of Execution</b></td>              <td vAlign="top" height="1">Manual                </td>              <td vAlign="top" height="21"><b>Verification Logic</b></td>              <td vAlign="top" height="1">BOTH<br>                </td>            </tr>            <tr bgColor="#ffffff">              <td vAlign="top" height="21"><b>Author</b></td>              <td vAlign="top" height="21">Shweta Chavan</td>              <td vAlign="top" height="21"><b>Date Entered</b></td>              <td vAlign="top" height="21">04/10/2006&nbsp;</td>            </tr>            <tr bgColor="#ffffff">              <td vAlign="top" height="21"></td>              <td vAlign="top" height="21"></td>              <td vAlign="top" height="21"><b>Date Last Update</b></td>              <td vAlign="top" height="21">04/10/2006&nbsp;</td>            </tr>          </tbody>        </table>      </td>    </tr>  </tbody></table>
<p align="right"><a href="#TMT Generated Test Plans">Top</a></p></body></html>
