<META HTTP-EQUIV="Content-Type" CONTENT="text/html; charset=us-ascii">
<p align=\       ><b><font face=\         size=\    >GFI MailSecurity's HTML threat engine found HTML scripts in this email and has disabled them.</font></b></p>

<STYLE type=text/css>BODY {
	BACKGROUND-COLOR: #ffffff; COLOR: #000000; FONT-FAMILY: "Times New Roman"; FONT-SIZE: 12pt; TEXT-ALIGN: left
}
 UL.zzul {
	DISPLAY: block; LIST-STYLE: none}
SPAN.zzspace {
	LEFT: 11px; POSITION: relative
}
 A {
	COLOR: #0000ff
}
 A:visited {
	COLOR: #0000ff
}
 A:hover {
	COLOR: #ff0000
}
TD {
	COLOR: #000000; FONT-FAMILY: "Times New Roman"; FONT-SIZE: 12pt; TEXT-ALIGN: left
}
 TABLE {
	BORDER-BOTTOM: medium none; BORDER-LEFT: medium none; BORDER-RIGHT: medium none; BORDER-TOP: medium none; PADDING-BOTTOM: 3px; PADDING-LEFT: 3px; PADDING-RIGHT: 3px; PADDING-TOP: 3px; WIDTH: 100%
 }
 </STYLE>
 


<center><a name="TMT Generated Test Plans"><h1>Test Plan</h1></a></center><hr><a name="xys"><table><tr><td>This is a TMT (Test Management Tool) generated Test Plan. To read a short note on TMT on the caBIG Best Practices SIG website, click <a href="http://cabig-bpsigwiki.bioinformatics.northwestern.edu/bpsigwiki/index.php/Manual_Testing_-_Test_Management_Tool">here</a>.<br> For more details or if you find any issue with this report, please send an email to <a href="mailto:tmt@persistent.co.in">tmt@persistent.co.in</a>.</td></tr></table></a><hr><br><br>
<A href="">Expand All</A> &nbsp; <A href="">Collapse All</A> 

<UL class=zzul>
  <LI>
<LI><A href="">+</A> <A 
	  href=""></A> &nbsp; <A  href="#Product Name: GeneConnect">GeneConnect</A> 
<UL class=zzul id=sp_1 style="DISPLAY: block">

<LI><A href="">+</A> <A 
	  href=""></A> &nbsp; <A  href="#Version Name: GeneConnectver 1.1">ver 1.1</A> 
<UL class=zzul id=sp_2 style="DISPLAY: block">

<LI><A href="">+</A> <A 
	  href=""></A> &nbsp; <A  href="#Component Name: GeneConnectver 1.1XMLRPC System">XMLRPC System</A> 
<UL class=zzul id=sp_3 style="DISPLAY: block">

<LI><A href="">+</A> <A 
	  href=""></A> &nbsp; <A  href="#Test Area Name: GeneConnectver 1.1XMLRPC SystemSystem Tests">System Tests</A> 
<UL class=zzul id=sp_4 style="DISPLAY: block">

<LI><SPAN class=zzspace>&nbsp;&nbsp;
<A href="#Test Case ID: GeneConnectver 1.1XMLRPC SystemSystem TestsBlastp  - Refseq Protein vs. Ensembl peptide">Blastp  - Refseq Protein vs. Ensembl peptide</A> </SPAN>
<LI><SPAN class=zzspace>&nbsp;&nbsp;
<A href="#Test Case ID: GeneConnectver 1.1XMLRPC SystemSystem TestsBlastn - Refseq mRNA vs. Ensembl cDNA">Blastn - Refseq mRNA vs. Ensembl cDNA</A> </SPAN>
</LI></UL>

</LI></UL>

</LI></UL>
<br><br><a><table><tr><td width="50%"><table bgColor="#CCCCFF" border="1" cellPadding="2" width="100"><tr><td colspan="10"><font color="#660066"><b>Test case count: Component verses Priority</b></font></td></tr><tr><td  align="center" bgcolor="#ccccff"><font color="#660066"><b>Component Names</b></font></td><td align="center" bgcolor="#ccccff"> <font color="#660066"><b>P1</b></font></td><td align="center" bgcolor="#ccccff"> <font color="#660066"><b>P2</b></font></td><td align="center" bgcolor="#ccccff"> <font color="#660066"><b>P3</b></font></td><td align="center" bgcolor="#ccccff"> <font color="#660066"><b>P4</b></font></td><td align="center" bgcolor="#ccccff"> <font color="#660066"><b>Total</b></td></tr><tr><td bgcolor="#ffffff">ver 1.1</td><td  bgcolor="#ffffff">15</td><td  bgcolor="#ffffff">0</td><td  bgcolor="#ffffff">0</td><td  bgcolor="#ffffff">0</td><td bgcolor="#ffffff">15</td></tr><tr><td bgcolor="#ffffff">XMLRPC System</td><td  bgcolor="#ffffff">2</td><td  bgcolor="#ffffff">0</td><td  bgcolor="#ffffff">0</td><td  bgcolor="#ffffff">0</td><td bgcolor="#ffffff">2</td></tr><tr><td bgcolor="#ccccff"><b>Column Total</b></td><td  bgcolor="#ccccff"><b>17<b></td><td  bgcolor="#ccccff"><b>0<b></td><td  bgcolor="#ccccff"><b>0<b></td><td  bgcolor="#ccccff"><b>0<b></td><td  bgcolor="#ccccff"><b>17</b></td></table></a></td><td width="50%">&nbsp;</td></tr></table><br><br><a name="Product Name: GeneConnect"><center><table bgColor="#CCCCFF" border="1" cellPadding="0" cellSpacing="0" width="700">  <tbody>    <tr>      <td align="middle" bgColor="#CCCCFF">        <p align="left"><font color="#660066"><b>Product Name: GeneConnect</b></font></p> </td> </tr> <tr> <td>&nbsp;</td> </tr> </tbody></table><br><table bgColor="#CCCCFF" border="1" cellPadding="0" cellSpacing="0" width="700"> <tbody> <tr> <td align="middle" bgColor="#CCCCFF"> <p align="left"><font color="#660066"><b>VITALS</b></font></p> </td> </tr> <tr> <td> <table bgColor="#CCCCFF" border="1" cellPadding="2" width="696"> <tbody> <tr bgColor="#ffffff"> <td vAlign="top" width="131"><b>Reference</b></td> <td vAlign="top" width="529">&nbsp;<a href="" target="_blank"></a></td> </tr> </tbody> </table> <tr> <td align="middle" bgColor="#CCCCFF"> <p align="left"><font color="#660066"><b>DESCRIPTION</b></font></p> </td> </tr> <tr> <td> <table bgColor="#CCCCFF" border="1" cellPadding="2" width="100%" height="55"> <tbody> <tr bgColor="#ffffff"> <td vAlign="top" height="1"></td> </tr> </tbody> </table> </td> </tr> </tbody></table>
<p align="right"><a href="#TMT Generated Test Plans">Top</a></p><a name="Version Name: GeneConnectver 1.1"><center><table bgColor="#CCCCFF" border="1" cellPadding="0" cellSpacing="0" width="700">  <tbody>    <tr>      <td align="middle" bgColor="#CCCCFF">        <p align="left"><font color="#660066"><b>Version Name: ver 1.1</b></font></p> </td> </tr> <tr> <td>&nbsp;</td> </tr> </tbody></table><br><table bgColor="#CCCCFF" border="1" cellPadding="0" cellSpacing="0" width="700"> <tbody> <tr> <td align="middle" bgColor="#CCCCFF"> <p align="left"><font color="#660066"><b>VITALS</b></font></p> </td> </tr> <tr> <td> <table bgColor="#CCCCFF" border="1" cellPadding="2" width="696"> <tbody> <tr bgColor="#ffffff"> <td vAlign="top" width="131"><b>Reference</b></td> <td vAlign="top" width="529">&nbsp;<a href="" target="_blank"></a></td> </tr> </tbody> </table> <tr> <td align="middle" bgColor="#CCCCFF"> <p align="left"><font color="#660066"><b>DESCRIPTION</b></font></p> </td> </tr> <tr> <td> <table bgColor="#CCCCFF" border="1" cellPadding="2" width="100%" height="55"> <tbody> <tr bgColor="#ffffff"> <td vAlign="top" height="1">First Version</td> </tr> </tbody> </table> </td> </tr> </tbody></table>
<p align="right"><a href="#TMT Generated Test Plans">Top</a></p><a name="Component Name: GeneConnectver 1.1XMLRPC System"><center><table bgColor="#CCCCFF" border="1" cellPadding="0" cellSpacing="0" width="700">  <tbody>    <tr>      <td align="middle" bgColor="#CCCCFF">        <p align="left"><font color="#660066"><b>Component Name: XMLRPC System</b></font></p> </td> </tr> <tr> <td>&nbsp;</td> </tr> </tbody></table><br><table bgColor="#CCCCFF" border="1" cellPadding="0" cellSpacing="0" width="700"> <tbody> <tr> <td align="middle" bgColor="#CCCCFF"> <p align="left"><font color="#660066"><b>VITALS</b></font></p> </td> </tr> <tr> <td> <table bgColor="#CCCCFF" border="1" cellPadding="2" width="696"> <tbody> <tr bgColor="#ffffff"> <td vAlign="top" width="131"><b>Reference</b></td> <td vAlign="top" width="529">&nbsp;<a href="" target="_blank"></a></td> </tr> </tbody> </table> <tr> <td align="middle" bgColor="#CCCCFF"> <p align="left"><font color="#660066"><b>DESCRIPTION</b></font></p> </td> </tr> <tr> <td> <table bgColor="#CCCCFF" border="1" cellPadding="2" width="100%" height="55"> <tbody> <tr bgColor="#ffffff"> <td vAlign="top" height="1">The XMLRPC system used by GeneConnect</td> </tr> </tbody> </table> </td> </tr> </tbody></table>
<p align="right"><a href="#TMT Generated Test Plans">Top</a></p><a name="Test Area Name: GeneConnectver 1.1XMLRPC SystemSystem Tests"><center><table bgColor="#CCCCFF" border="1" cellPadding="0" cellSpacing="0" width="700">  <tbody>    <tr>      <td align="middle" bgColor="#CCCCFF">        <p align="left"><font color="#660066"><b>Test Area Name: System Tests</b></font></p> </td> </tr> <tr> <td>&nbsp;</td> </tr> </tbody></table><br><table bgColor="#CCCCFF" border="1" cellPadding="0" cellSpacing="0" width="700"> <tbody> <tr> <td align="middle" bgColor="#CCCCFF"> <p align="left"><font color="#660066"><b>VITALS</b></font></p> </td> </tr> <tr> <td> <table bgColor="#CCCCFF" border="1" cellPadding="2" width="696"> <tbody> <tr bgColor="#ffffff"> <td vAlign="top" width="131"><b>Reference</b></td> <td vAlign="top" width="529">&nbsp;<a href="" target="_blank"></a></td> </tr> </tbody> </table> <tr> <td align="middle" bgColor="#CCCCFF"> <p align="left"><font color="#660066"><b>DESCRIPTION</b></font></p> </td> </tr> <tr> <td> <table bgColor="#CCCCFF" border="1" cellPadding="2" width="100%" height="55"> <tbody> <tr bgColor="#ffffff"> <td vAlign="top" height="1"></td> </tr> </tbody> </table> </td> </tr> </tbody></table>
<p align="right"><a href="#TMT Generated Test Plans">Top</a></p><a name="Test Case ID: GeneConnectver 1.1XMLRPC SystemSystem TestsBlastp  - Refseq Protein vs. Ensembl peptide"><center><table bgColor="#CCCCFF" border="1" cellPadding="0" cellSpacing="0" width="700">  <tbody>    <tr>      <td align="middle" bgColor="#CCCCFF">        <p align="left"><font color="#660066"><b>Test Case ID : 1661</b></font></p>      </td>    </tr>    <tr>      <td bgcolor="#FFFFFF"><b>Short Title : </b>Blastp  - Refseq Protein vs. Ensembl peptide</td>    </tr>    <tr>      <td bgcolor="#FFFFFF"><b>Title : </b>Blast against a protein sequence to verify the functionality of the XMLRPC system</td>    </tr>  </tbody></table><br><table bgColor="#CCCCFF" border="1" cellPadding="0" cellSpacing="0" width="700">  <tbody>    <tr>      <td align="middle" bgColor="#CCCCFF">        <p align="left"><font color="#660066"><b>VITALS</b></font></p>      </td>    </tr>    <tr>      <td>        <table bgColor="#CCCCFF" border="1" cellPadding="2" width="100%">          <tbody>            <tr bgColor="#ffffff">              <td vAlign="top"><b>Product</b></td>              <td vAlign="top" colspan="3">GeneConnect</td>            <tr bgColor="#ffffff">              <td vAlign="top"><b>Component</b></td>              <td vAlign="top">XMLRPC System<br>              </td>              <td vAlign="top"><b>Test Area</b></td>              <td vAlign="top">System Tests</td>            </tr>            <tr bgColor="#ffffff">              <td vAlign="top"><b>Reference</b></td>              <td vAlign="top">1</td>              <td vAlign="top"><b>Priority</b></td>              <td vAlign="top">P1</td>            </tr>          </tbody>        </table>    <tr>      <td align="middle" bgColor="#CCCCFF">        <p align="left"><font color="#660066"><b>DESCRIPTION</b></font></p>      </td>    </tr>    <tr>      <td>        <table bgColor="#CCCCFF" border="1" cellPadding="2" width="100%" height="113">          <tbody>            <tr bgColor="#ffffff"><td vAlign="top" height="32" colspan="4"><font class="itsmultlinefont"><pre>Purpose:
To ensure that the XMLRPC system works with blast against a protein sequence.

Setup:
1. Ensure that the following are set up/installed for Perl and to obtain sequence alignement results:

        Perl (CPAN)
		bioperl
		bioperl-run
		libwww
		URI
		DBM::Deep
		Fontier-RPC

	Sequence Alignment
		needle
		blastn
		blastp

2. The following custom modules should be present in Perl:

		Cluster::Job
		Cluster::JobQueue
		Cluster::Machine
		Cluster::MachineQueue
		Cluster::QueueManager
		GeneConnect::Hit
		GeneConnect::Params
		SplitSeqFile.pm

3. The target database (Ensembl's Homo_sapiens.NCBI36.apr.pep.fa) should be set up as a BLAST database on the compute node (compute-0-16).  The path to the database (/state/partition1/blastDB/) should be specified in the environment variable BLASTDB.

4. Copy the following files to /home/clususer/geneconnect/XML_RPC_System/test
		SplitSeqFile.pm
		split_fasta_file.pl
		XML_RPC_Server.pl
		XML_RPC_Server_Monitor.pl
		XML_RPC_testclient.pl
		machines.txt
		fastarefprotein10.fasta (containing query sequences)
5. Copy the following files to /home/clususer/geneconnect/XML_RPC_System/test/Cluster
		Job.pm
		JobQueue.pm
		Machine.pm
		MachineQueue.pm
		QueueManager.pm
6. Copy the following files to /home/clususer/geneconnect/XML_RPC_System/test/GeneConnect
		Hit.pm
		Params.pm
7. Copy gc_seq_align_pipeline.pl to the compute node (compute-0-16), in /state/partition1/geneconnect
8. Set the configuration parameters in XML_RPC_testclient.pl to the following:

	SRCFILE =&gt; 'fastarefprotein10.fasta',
	DESTINATIONDB =&gt; 'Homo_sapiens.NCBI36.apr.pep.fa',
	PATHS =&gt; {
	 	    db =&gt; '/state/partition1/blastDB/',
		base =&gt; '/home/clususer/geneconnect/XML_RPC_system/test/',
	      source =&gt; '/home/clususer/geneconnect/XML_RPC_system/test/',
	      output =&gt; '/home/clususer/geneconnect/XML_RPC_system/test/',
		  },
	      BLAST =&gt; {
			  blastType =&gt; 'blastp',
			  EValue =&gt; 1e-25,
			  ScoringMatrix =&gt; 'BLOSUM62',
			  GapOpenCost =&gt; undef,
			  GapExtensionCost =&gt; undef,
			  Dropoff =&gt; undef,
			  WordSize =&gt; undef,
			  seqType =&gt; 'protein',
			},
			BLAST_FILTER =&gt; {								     PercentSequenceAlignment =&gt; 0.80,
				},
			NEEDLE =&gt; {
				GapOpenCost =&gt; undef,
		                GapExtend =&gt; undef,
		                Matrix =&gt; undef,
		              },
			NEEDLE_FILTER =&gt; {
		                 	PercentSequenceAlignment =&gt; 0.90,
			                 	SimilarityScore =&gt; 0.90,
			                 	IdentityScore =&gt; 0.90,
		                 		NumberMisMatch =&gt; 2,
		                 	},
			CONFIG =&gt; {
				numSeqsPerJob =&gt; 3,
		                saveAlignOutput =&gt; 1,
		                controlFileHeader =&gt; 'test header',
		                controlFileColumnSeparator =&gt; '###',
		              },

9. The machines.txt file should contain the name of the compute node (compute-0-16).

10. In XML_RPC_Server.pl, verify the following :
		a. the master node name (im-vishnu) is provided to initialize_queues(), 'db_queue'
		b. in seq_align_pipeline(), the path to split_fasta_file.pl (on the master node) is correct
		c. in do_alignments(), the path to gc_seq_align_pipeline.pl (on the compute nodes) is correct.

Procedure:
1. Start the XML_RPC_Server.
2. (optional) In a separate terminal, start the XML_RPC_Server_Monitor.
3. In a separate terminal, execute XML_RPC_testclient.pl.

Input/Data:
Homo_sapiens.NCBI36.apr.pep.fa (target database)
fastarefprotein10.fasta (containing query sequences)

Expected Output:

1. A working directory will be created in the base path (/home/clususer/geneconnect/XML_RPC_system/test/).  The working directory is named YYYYMMDD_(query sequence filename)_(target database), where YYYYMMDD corresponds to the current year, month, and day and the query sequence and target database names are as specified in the configuration parameters.  The working directory for this test should be named 'YYYYMMDD_fastarefprotein10.fasta_Homo_sapiens.NCBI36.apr.pep.fa', where YYYYMMDD is the current date.
	2. The following files will be created in the working directory, where '?' is 1, 2, 3, or 4.
		fastarefprotein10.fasta_part?.fa
		fastarefprotein10.fasta_part?.fa_control.txt
		fastarefprotein10.fasta_part?.fa_pipeline_log.txt
		fastarefprotein10.fasta_part?.fa_pipeline_output.txt





</pre></font></td>            </tr>            <tr bgColor="#ffffff">              <td vAlign="top" height="21"><b>Mode of Execution</b></td>              <td vAlign="top" height="1">Manual                </td>              <td vAlign="top" height="21"><b>Verification Logic</b></td>              <td vAlign="top" height="1">DATABASE<br>                </td>            </tr>            <tr bgColor="#ffffff">              <td vAlign="top" height="21"><b>Author</b></td>              <td vAlign="top" height="21">Nisha Thomas</td>              <td vAlign="top" height="21"><b>Date Entered</b></td>              <td vAlign="top" height="21">17/08/2006&nbsp;</td>            </tr>            <tr bgColor="#ffffff">              <td vAlign="top" height="21"></td>              <td vAlign="top" height="21"></td>              <td vAlign="top" height="21"><b>Date Last Update</b></td>              <td vAlign="top" height="21">19/08/2006&nbsp;</td>            </tr>          </tbody>        </table>      </td>    </tr>  </tbody></table>
<p align="right"><a href="#TMT Generated Test Plans">Top</a></p><a name="Test Case ID: GeneConnectver 1.1XMLRPC SystemSystem TestsBlastn - Refseq mRNA vs. Ensembl cDNA"><center><table bgColor="#CCCCFF" border="1" cellPadding="0" cellSpacing="0" width="700">  <tbody>    <tr>      <td align="middle" bgColor="#CCCCFF">        <p align="left"><font color="#660066"><b>Test Case ID : 1662</b></font></p>      </td>    </tr>    <tr>      <td bgcolor="#FFFFFF"><b>Short Title : </b>Blastn - Refseq mRNA vs. Ensembl cDNA</td>    </tr>    <tr>      <td bgcolor="#FFFFFF"><b>Title : </b>Blast against a nucleotide sequence to verify the functionality of the XMLRPC system</td>    </tr>  </tbody></table><br><table bgColor="#CCCCFF" border="1" cellPadding="0" cellSpacing="0" width="700">  <tbody>    <tr>      <td align="middle" bgColor="#CCCCFF">        <p align="left"><font color="#660066"><b>VITALS</b></font></p>      </td>    </tr>    <tr>      <td>        <table bgColor="#CCCCFF" border="1" cellPadding="2" width="100%">          <tbody>            <tr bgColor="#ffffff">              <td vAlign="top"><b>Product</b></td>              <td vAlign="top" colspan="3">GeneConnect</td>            <tr bgColor="#ffffff">              <td vAlign="top"><b>Component</b></td>              <td vAlign="top">XMLRPC System<br>              </td>              <td vAlign="top"><b>Test Area</b></td>              <td vAlign="top">System Tests</td>            </tr>            <tr bgColor="#ffffff">              <td vAlign="top"><b>Reference</b></td>              <td vAlign="top">1</td>              <td vAlign="top"><b>Priority</b></td>              <td vAlign="top">P1</td>            </tr>          </tbody>        </table>    <tr>      <td align="middle" bgColor="#CCCCFF">        <p align="left"><font color="#660066"><b>DESCRIPTION</b></font></p>      </td>    </tr>    <tr>      <td>        <table bgColor="#CCCCFF" border="1" cellPadding="2" width="100%" height="113">          <tbody>            <tr bgColor="#ffffff"><td vAlign="top" height="32" colspan="4"><font class="itsmultlinefont"><pre>Purpose:
To ensure that the XMLRPC system works with blast against a nucleotide sequence.

Setup:
1. Ensure that the following are set up/installed for Perl and to obtain sequence alignement results:

        Perl (CPAN)
		bioperl
		bioperl-run
		libwww
		URI
		DBM::Deep
		Fontier-RPC

	Sequence Alignment
		needle
		blastn
		blastp

2. The following custom modules should be present in Perl:

		Cluster::Job
		Cluster::JobQueue
		Cluster::Machine
		Cluster::MachineQueue
		Cluster::QueueManager
		GeneConnect::Hit
		GeneConnect::Params
		SplitSeqFile.pm

3. The target database (Ensembl's Homo_sapiens.NCBI36.apr.pep.fa) should be set up as a BLAST database on the compute node (compute-0-16).  The path to the database (/state/partition1/blastDB/) should be specified in the environment variable BLASTDB.

4. Copy the following files to /home/clususer/geneconnect/XML_RPC_System/test
		SplitSeqFile.pm
		split_fasta_file.pl
		XML_RPC_Server.pl
		XML_RPC_Server_Monitor.pl
		XML_RPC_testclient.pl
		machines.txt
		fastarefmrna10.fasta (containing query sequences)
5. Copy the following files to /home/clususer/geneconnect/XML_RPC_System/test/Cluster
		Job.pm
		JobQueue.pm
		Machine.pm
		MachineQueue.pm
		QueueManager.pm
6. Copy the following files to /home/clususer/geneconnect/XML_RPC_System/test/GeneConnect
		Hit.pm
		Params.pm
7. Copy gc_seq_align_pipeline.pl to the compute node (compute-0-16), in /state/partition1/geneconnect
8. Set the configuration parameters in XML_RPC_testclient.pl to the following:

	SRCFILE =&gt; 'fastarefmrna10.fasta',
	DESTINATIONDB =&gt; 'Homo_sapiens.NCBI36.apr.pep.fa',
	PATHS =&gt; {
	 	    db =&gt; '/state/partition1/blastDB/',
		base =&gt; '/home/clususer/geneconnect/XML_RPC_system/test/',
	      source =&gt; '/home/clususer/geneconnect/XML_RPC_system/test/',
	      output =&gt; '/home/clususer/geneconnect/XML_RPC_system/test/',
		  },
	      BLAST =&gt; {
			  blastType =&gt; 'blastn',
			  EValue =&gt; 1e-25,
			  ScoringMatrix =&gt; 'BLOSUM62',
			  GapOpenCost =&gt; undef,
			  GapExtensionCost =&gt; undef,
			  Dropoff =&gt; undef,
			  WordSize =&gt; undef,
			  seqType =&gt; 'nucleotide',
			},
			BLAST_FILTER =&gt; {								     PercentSequenceAlignment =&gt; 0.80,
				},
			NEEDLE =&gt; {
				GapOpenCost =&gt; undef,
		                GapExtend =&gt; undef,
		                Matrix =&gt; undef,
		              },
			NEEDLE_FILTER =&gt; {
		                 	PercentSequenceAlignment =&gt; 0.90,
			                 	SimilarityScore =&gt; 0.90,
			                 	IdentityScore =&gt; 0.90,
		                 		NumberMisMatch =&gt; 2,
		                 	},
			CONFIG =&gt; {
				numSeqsPerJob =&gt; 3,
		                saveAlignOutput =&gt; 1,
		                controlFileHeader =&gt; 'test header',
		                controlFileColumnSeparator =&gt; '###',
		              },

9. The machines.txt file should contain the name of the compute node (compute-0-16).

10. In XML_RPC_Server.pl, verify the following :
		a. the master node name (im-vishnu) is provided to initialize_queues(), 'db_queue'
		b. in seq_align_pipeline(), the path to split_fasta_file.pl (on the master node) is correct
		c. in do_alignments(), the path to gc_seq_align_pipeline.pl (on the compute nodes) is correct.

Procedure:
1. Start the XML_RPC_Server.
2. (optional) In a separate terminal, start the XML_RPC_Server_Monitor.
3. In a separate terminal, execute XML_RPC_testclient.pl.

Input/Data:
Homo_sapiens.NCBI36.apr.pep.fa (target database)
fastarefmrna10.fasta (containing query sequences)

Expected Output:

1. A working directory will be created in the base path (/home/clususer/geneconnect/XML_RPC_system/test/).  The working directory is named YYYYMMDD_(query sequence filename)_(target database), where YYYYMMDD corresponds to the current year, month, and day and the query sequence and target database names are as specified in the configuration parameters.  The working directory for this test should be named 'YYYYMMDD_fastarefmrna10.fasta_Homo_sapiens.NCBI36.apr.pep.fa', where YYYYMMDD is the current date.
	2. The following files will be created in the working directory, where '?' is 1, 2, 3, or 4.
		fastarefmrna10.fasta_part?.fa
		fastarefmrna10.fasta_part?.fa_control.txt
		fastarefmrna10.fasta_part?.fa_pipeline_log.txt
		fastarefmrna10.fasta_part?.fa_pipeline_output.txt







</pre></font></td>            </tr>            <tr bgColor="#ffffff">              <td vAlign="top" height="21"><b>Mode of Execution</b></td>              <td vAlign="top" height="1">Manual                </td>              <td vAlign="top" height="21"><b>Verification Logic</b></td>              <td vAlign="top" height="1">DATABASE<br>                </td>            </tr>            <tr bgColor="#ffffff">              <td vAlign="top" height="21"><b>Author</b></td>              <td vAlign="top" height="21">Nisha Thomas</td>              <td vAlign="top" height="21"><b>Date Entered</b></td>              <td vAlign="top" height="21">17/08/2006&nbsp;</td>            </tr>            <tr bgColor="#ffffff">              <td vAlign="top" height="21"></td>              <td vAlign="top" height="21"></td>              <td vAlign="top" height="21"><b>Date Last Update</b></td>              <td vAlign="top" height="21">19/08/2006&nbsp;</td>            </tr>          </tbody>        </table>      </td>    </tr>  </tbody></table>
<p align="right"><a href="#TMT Generated Test Plans">Top</a></p></body></html>
