<head>
<META content="Microsoft FrontPage 4.0" name=GENERATOR>
<STYLE type=text/css>BODY {
	BACKGROUND-COLOR: #ffffff; COLOR: #000000; FONT-FAMILY: "Times New Roman"; FONT-SIZE: 12pt; TEXT-ALIGN: left
}
 UL.zzul {
	DISPLAY: block; LIST-STYLE: none}
SPAN.zzspace {
	LEFT: 11px; POSITION: relative
}
 A {
	COLOR: #0000ff
}
 A:visited {
	COLOR: #0000ff
}
 A:hover {
	COLOR: #ff0000
}
TD {
	COLOR: #000000; FONT-FAMILY: "Times New Roman"; FONT-SIZE: 12pt; TEXT-ALIGN: left
}
 TABLE {
	BORDER-BOTTOM: medium none; BORDER-LEFT: medium none; BORDER-RIGHT: medium none; BORDER-TOP: medium none; PADDING-BOTTOM: 3px; PADDING-LEFT: 3px; PADDING-RIGHT: 3px; PADDING-TOP: 3px; WIDTH: 100%
 }
 </STYLE>
 <SCRIPT type=text/javascript>
 <!--
 function exp_coll(ind)
 {
 s = document.getElementById("sp_" + ind);
 i = document.getElementById("im_" + ind);
 i2 = document.getElementById("im2_" + ind);
 if (s.style.display == 'none')
 {
  s.style.display = 'block';   
 }
 else if (s.style.display == 'block')
 {
   s.style.display = 'none';   
 }
}
 function coll_all(i)
 {  
 for(ind=1;ind<=i;ind++){
  x = document.getElementById("sp_" + ind);
  x.style.display = 'none';
  if (x.style.display == 'none')
  {    
 }
}
}
 function exp_all(i)
 {
  for(ind=1;ind<=i;ind++){
  x = document.getElementById("sp_" + ind);
  x.style.display = 'block';
  if (x.style.display == 'block')
  {    
 }
  }  
}
 -->
</SCRIPT>
</HEAD>
<BODY>
<center><a name="TMT Generated Test Plans"><h1>Test Plan</h1></a></center><hr><a name="xys"><table><tr><td>This is a TMT (Test Management Tool) generated Test Plan. To read a short note on TMT on the caBIG Best Practices SIG website, click <a href="http://cabig-bpsigwiki.bioinformatics.northwestern.edu/bpsigwiki/index.php/Manual_Testing_-_Test_Management_Tool">here</a>.<br> For more details or if you find any issue with this report, please send an email to <a href="mailto:tmt@persistent.co.in">tmt@persistent.co.in</a>.</td></tr></table></a><hr><br><br>
<A href="javascript:exp_all(10);">Expand All</A> &nbsp; <A href="javascript:coll_all(10);">Collapse All</A> 

<UL class=zzul>
  <LI>
<LI><A href="javascript:exp_coll(1);">+</A> <A 
	  href="javascript:exp_coll(1);"></A> &nbsp; <A  href="#Product Name: GeneConnect">GeneConnect</A> 
<UL class=zzul id=sp_1 style="DISPLAY: block">
<SCRIPT type=text/javascript>
 ind = 1; 
 x = document.getElementById("sp_" + ind);
 x.style.display = 'none'; 
 if (x.style.display == 'none'){ }
</SCRIPT>
<LI><A href="javascript:exp_coll(2);">+</A> <A 
	  href="javascript:exp_coll(2);"></A> &nbsp; <A  href="#Version Name: GeneConnectver 1.1">ver 1.1</A> 
<UL class=zzul id=sp_2 style="DISPLAY: block">
<SCRIPT type=text/javascript>
 ind = 2; 
 x = document.getElementById("sp_" + ind);
 x.style.display = 'none'; 
 if (x.style.display == 'none'){ }
</SCRIPT>
<LI><A href="javascript:exp_coll(3);">+</A> <A 
	  href="javascript:exp_coll(3);"></A> &nbsp; <A  href="#Component Name: GeneConnectver 1.1Parsers">Parsers</A> 
<UL class=zzul id=sp_3 style="DISPLAY: block">
<SCRIPT type=text/javascript>
 ind = 3; 
 x = document.getElementById("sp_" + ind);
 x.style.display = 'none'; 
 if (x.style.display == 'none'){ }
</SCRIPT>
<LI><A href="javascript:exp_coll(4);">+</A> <A 
	  href="javascript:exp_coll(4);"></A> &nbsp; <A  href="#Test Area Name: GeneConnectver 1.1ParsersBlast Destination Parser">Blast Destination Parser</A> 
<UL class=zzul id=sp_4 style="DISPLAY: block">
<SCRIPT type=text/javascript>
 ind = 4; 
 x = document.getElementById("sp_" + ind);
 x.style.display = 'none'; 
 if (x.style.display == 'none'){ }
</SCRIPT>
<LI><SPAN class=zzspace>&nbsp;&nbsp;
<A href="#Test Case ID: GeneConnectver 1.1ParsersBlast Destination ParserEnsembl Transcript">Ensembl Transcript</A> </SPAN>
<LI><SPAN class=zzspace>&nbsp;&nbsp;
<A href="#Test Case ID: GeneConnectver 1.1ParsersBlast Destination ParserEnsembl Protein">Ensembl Protein</A> </SPAN>
<LI><SPAN class=zzspace>&nbsp;&nbsp;
<A href="#Test Case ID: GeneConnectver 1.1ParsersBlast Destination ParserRefseq mrna">Refseq mrna</A> </SPAN>
<LI><SPAN class=zzspace>&nbsp;&nbsp;
<A href="#Test Case ID: GeneConnectver 1.1ParsersBlast Destination ParserRefseq Protein">Refseq Protein</A> </SPAN>
<LI><SPAN class=zzspace>&nbsp;&nbsp;
<A href="#Test Case ID: GeneConnectver 1.1ParsersBlast Destination ParserUniprot">Uniprot</A> </SPAN>
</LI></UL>

<LI><A href="javascript:exp_coll(5);">+</A> <A 
	  href="javascript:exp_coll(5);"></A> &nbsp; <A  href="#Test Area Name: GeneConnectver 1.1ParsersBlast Source Parser">Blast Source Parser</A> 
<UL class=zzul id=sp_5 style="DISPLAY: block">
<SCRIPT type=text/javascript>
 ind = 5; 
 x = document.getElementById("sp_" + ind);
 x.style.display = 'none'; 
 if (x.style.display == 'none'){ }
</SCRIPT>
<LI><SPAN class=zzspace>&nbsp;&nbsp;
<A href="#Test Case ID: GeneConnectver 1.1ParsersBlast Source ParserGenBankmRNA_01">GenBankmRNA_01</A> </SPAN>
<LI><SPAN class=zzspace>&nbsp;&nbsp;
<A href="#Test Case ID: GeneConnectver 1.1ParsersBlast Source ParserGenBankProtein_01">GenBankProtein_01</A> </SPAN>
</LI></UL>

<LI><A href="javascript:exp_coll(6);">+</A> <A 
	  href="javascript:exp_coll(6);"></A> &nbsp; <A  href="#Test Area Name: GeneConnectver 1.1ParsersEnsembl External Parser">Ensembl External Parser</A> 
<UL class=zzul id=sp_6 style="DISPLAY: block">
<SCRIPT type=text/javascript>
 ind = 6; 
 x = document.getElementById("sp_" + ind);
 x.style.display = 'none'; 
 if (x.style.display == 'none'){ }
</SCRIPT>
<LI><SPAN class=zzspace>&nbsp;&nbsp;
<A href="#Test Case ID: GeneConnectver 1.1ParsersEnsembl External ParserEnsembl Parser">Ensembl Parser</A> </SPAN>
</LI></UL>

<LI><A href="javascript:exp_coll(7);">+</A> <A 
	  href="javascript:exp_coll(7);"></A> &nbsp; <A  href="#Test Area Name: GeneConnectver 1.1ParsersEntrez Gene parser">Entrez Gene parser</A> 
<UL class=zzul id=sp_7 style="DISPLAY: block">
<SCRIPT type=text/javascript>
 ind = 7; 
 x = document.getElementById("sp_" + ind);
 x.style.display = 'none'; 
 if (x.style.display == 'none'){ }
</SCRIPT>
<LI><SPAN class=zzspace>&nbsp;&nbsp;
<A href="#Test Case ID: GeneConnectver 1.1ParsersEntrez Gene parserEntrez-Unigene_01">Entrez-Unigene_01</A> </SPAN>
</LI></UL>

<LI><A href="javascript:exp_coll(8);">+</A> <A 
	  href="javascript:exp_coll(8);"></A> &nbsp; <A  href="#Test Area Name: GeneConnectver 1.1ParsersEntrez_Refseq External Parser">Entrez_Refseq External Parser</A> 
<UL class=zzul id=sp_8 style="DISPLAY: block">
<SCRIPT type=text/javascript>
 ind = 8; 
 x = document.getElementById("sp_" + ind);
 x.style.display = 'none'; 
 if (x.style.display == 'none'){ }
</SCRIPT>
<LI><SPAN class=zzspace>&nbsp;&nbsp;
<A href="#Test Case ID: GeneConnectver 1.1ParsersEntrez_Refseq External ParserEntrezGene_RefSeqmRNA_01">EntrezGene_RefSeqmRNA_01</A> </SPAN>
<LI><SPAN class=zzspace>&nbsp;&nbsp;
<A href="#Test Case ID: GeneConnectver 1.1ParsersEntrez_Refseq External ParserRefseqmRNA_ RefseqProtein_01">RefseqmRNA_ RefseqProtein_01</A> </SPAN>
</LI></UL>

<LI><A href="javascript:exp_coll(9);">+</A> <A 
	  href="javascript:exp_coll(9);"></A> &nbsp; <A  href="#Test Area Name: GeneConnectver 1.1ParsersUnigene Parser">Unigene Parser</A> 
<UL class=zzul id=sp_9 style="DISPLAY: block">
<SCRIPT type=text/javascript>
 ind = 9; 
 x = document.getElementById("sp_" + ind);
 x.style.display = 'none'; 
 if (x.style.display == 'none'){ }
</SCRIPT>
<LI><SPAN class=zzspace>&nbsp;&nbsp;
<A href="#Test Case ID: GeneConnectver 1.1ParsersUnigene ParserUnigene Parser">Unigene Parser</A> </SPAN>
</LI></UL>

<LI><A href="javascript:exp_coll(10);">+</A> <A 
	  href="javascript:exp_coll(10);"></A> &nbsp; <A  href="#Test Area Name: GeneConnectver 1.1ParsersUniprot External Parser">Uniprot External Parser</A> 
<UL class=zzul id=sp_10 style="DISPLAY: block">
<SCRIPT type=text/javascript>
 ind = 10; 
 x = document.getElementById("sp_" + ind);
 x.style.display = 'none'; 
 if (x.style.display == 'none'){ }
</SCRIPT>
<LI><SPAN class=zzspace>&nbsp;&nbsp;
<A href="#Test Case ID: GeneConnectver 1.1ParsersUniprot External ParserUniprot to Refseq">Uniprot to Refseq</A> </SPAN>
<LI><SPAN class=zzspace>&nbsp;&nbsp;
<A href="#Test Case ID: GeneConnectver 1.1ParsersUniprot External ParserUniprot to Genbank">Uniprot to Genbank</A> </SPAN>
</LI></UL>

</LI></UL>

</LI></UL>
<br><br><a><table><tr><td width="50%"><table bgColor="#CCCCFF" border="1" cellPadding="2" width="100"><tr><td colspan="10"><font color="#660066"><b>Test case count: Component verses Priority</b></font></td></tr><tr><td  align="center" bgcolor="#ccccff"><font color="#660066"><b>Component Names</b></font></td><td align="center" bgcolor="#ccccff"> <font color="#660066"><b>P1</b></font></td><td align="center" bgcolor="#ccccff"> <font color="#660066"><b>P2</b></font></td><td align="center" bgcolor="#ccccff"> <font color="#660066"><b>P3</b></font></td><td align="center" bgcolor="#ccccff"> <font color="#660066"><b>P4</b></font></td><td align="center" bgcolor="#ccccff"> <font color="#660066"><b>Total</b></td></tr><tr><td bgcolor="#ffffff">ver 1.1</td><td  bgcolor="#ffffff">15</td><td  bgcolor="#ffffff">0</td><td  bgcolor="#ffffff">0</td><td  bgcolor="#ffffff">0</td><td bgcolor="#ffffff">15</td></tr><tr><td bgcolor="#ffffff">Parsers</td><td  bgcolor="#ffffff">2</td><td  bgcolor="#ffffff">9</td><td  bgcolor="#ffffff">3</td><td  bgcolor="#ffffff">0</td><td bgcolor="#ffffff">14</td></tr><tr><td bgcolor="#ccccff"><b>Column Total</b></td><td  bgcolor="#ccccff"><b>17<b></td><td  bgcolor="#ccccff"><b>9<b></td><td  bgcolor="#ccccff"><b>3<b></td><td  bgcolor="#ccccff"><b>0<b></td><td  bgcolor="#ccccff"><b>29</b></td></table></a></td><td width="50%">&nbsp;</td></tr></table><br><br><a name="Product Name: GeneConnect"><center><table bgColor="#CCCCFF" border="1" cellPadding="0" cellSpacing="0" width="700">  <tbody>    <tr>      <td align="middle" bgColor="#CCCCFF">        <p align="left"><font color="#660066"><b>Product Name: GeneConnect</b></font></p> </td> </tr> <tr> <td>&nbsp;</td> </tr> </tbody></table><br><table bgColor="#CCCCFF" border="1" cellPadding="0" cellSpacing="0" width="700"> <tbody> <tr> <td align="middle" bgColor="#CCCCFF"> <p align="left"><font color="#660066"><b>VITALS</b></font></p> </td> </tr> <tr> <td> <table bgColor="#CCCCFF" border="1" cellPadding="2" width="696"> <tbody> <tr bgColor="#ffffff"> <td vAlign="top" width="131"><b>Reference</b></td> <td vAlign="top" width="529">&nbsp;<a href="" target="_blank"></a></td> </tr> </tbody> </table> <tr> <td align="middle" bgColor="#CCCCFF"> <p align="left"><font color="#660066"><b>DESCRIPTION</b></font></p> </td> </tr> <tr> <td> <table bgColor="#CCCCFF" border="1" cellPadding="2" width="100%" height="55"> <tbody> <tr bgColor="#ffffff"> <td vAlign="top" height="1"></td> </tr> </tbody> </table> </td> </tr> </tbody></table>
<p align="right"><a href="#TMT Generated Test Plans">Top</a></p><a name="Version Name: GeneConnectver 1.1"><center><table bgColor="#CCCCFF" border="1" cellPadding="0" cellSpacing="0" width="700">  <tbody>    <tr>      <td align="middle" bgColor="#CCCCFF">        <p align="left"><font color="#660066"><b>Version Name: ver 1.1</b></font></p> </td> </tr> <tr> <td>&nbsp;</td> </tr> </tbody></table><br><table bgColor="#CCCCFF" border="1" cellPadding="0" cellSpacing="0" width="700"> <tbody> <tr> <td align="middle" bgColor="#CCCCFF"> <p align="left"><font color="#660066"><b>VITALS</b></font></p> </td> </tr> <tr> <td> <table bgColor="#CCCCFF" border="1" cellPadding="2" width="696"> <tbody> <tr bgColor="#ffffff"> <td vAlign="top" width="131"><b>Reference</b></td> <td vAlign="top" width="529">&nbsp;<a href="" target="_blank"></a></td> </tr> </tbody> </table> <tr> <td align="middle" bgColor="#CCCCFF"> <p align="left"><font color="#660066"><b>DESCRIPTION</b></font></p> </td> </tr> <tr> <td> <table bgColor="#CCCCFF" border="1" cellPadding="2" width="100%" height="55"> <tbody> <tr bgColor="#ffffff"> <td vAlign="top" height="1">First Version</td> </tr> </tbody> </table> </td> </tr> </tbody></table>
<p align="right"><a href="#TMT Generated Test Plans">Top</a></p><a name="Component Name: GeneConnectver 1.1Parsers"><center><table bgColor="#CCCCFF" border="1" cellPadding="0" cellSpacing="0" width="700">  <tbody>    <tr>      <td align="middle" bgColor="#CCCCFF">        <p align="left"><font color="#660066"><b>Component Name: Parsers</b></font></p> </td> </tr> <tr> <td>&nbsp;</td> </tr> </tbody></table><br><table bgColor="#CCCCFF" border="1" cellPadding="0" cellSpacing="0" width="700"> <tbody> <tr> <td align="middle" bgColor="#CCCCFF"> <p align="left"><font color="#660066"><b>VITALS</b></font></p> </td> </tr> <tr> <td> <table bgColor="#CCCCFF" border="1" cellPadding="2" width="696"> <tbody> <tr bgColor="#ffffff"> <td vAlign="top" width="131"><b>Reference</b></td> <td vAlign="top" width="529">&nbsp;<a href="" target="_blank"></a></td> </tr> </tbody> </table> <tr> <td align="middle" bgColor="#CCCCFF"> <p align="left"><font color="#660066"><b>DESCRIPTION</b></font></p> </td> </tr> <tr> <td> <table bgColor="#CCCCFF" border="1" cellPadding="2" width="100%" height="55"> <tbody> <tr bgColor="#ffffff"> <td vAlign="top" height="1"></td> </tr> </tbody> </table> </td> </tr> </tbody></table>
<p align="right"><a href="#TMT Generated Test Plans">Top</a></p><a name="Test Area Name: GeneConnectver 1.1ParsersBlast Destination Parser"><center><table bgColor="#CCCCFF" border="1" cellPadding="0" cellSpacing="0" width="700">  <tbody>    <tr>      <td align="middle" bgColor="#CCCCFF">        <p align="left"><font color="#660066"><b>Test Area Name: Blast Destination Parser</b></font></p> </td> </tr> <tr> <td>&nbsp;</td> </tr> </tbody></table><br><table bgColor="#CCCCFF" border="1" cellPadding="0" cellSpacing="0" width="700"> <tbody> <tr> <td align="middle" bgColor="#CCCCFF"> <p align="left"><font color="#660066"><b>VITALS</b></font></p> </td> </tr> <tr> <td> <table bgColor="#CCCCFF" border="1" cellPadding="2" width="696"> <tbody> <tr bgColor="#ffffff"> <td vAlign="top" width="131"><b>Reference</b></td> <td vAlign="top" width="529">&nbsp;<a href="" target="_blank"></a></td> </tr> </tbody> </table> <tr> <td align="middle" bgColor="#CCCCFF"> <p align="left"><font color="#660066"><b>DESCRIPTION</b></font></p> </td> </tr> <tr> <td> <table bgColor="#CCCCFF" border="1" cellPadding="2" width="100%" height="55"> <tbody> <tr bgColor="#ffffff"> <td vAlign="top" height="1"></td> </tr> </tbody> </table> </td> </tr> </tbody></table>
<p align="right"><a href="#TMT Generated Test Plans">Top</a></p><a name="Test Case ID: GeneConnectver 1.1ParsersBlast Destination ParserEnsembl Transcript"><center><table bgColor="#CCCCFF" border="1" cellPadding="0" cellSpacing="0" width="700">  <tbody>    <tr>      <td align="middle" bgColor="#CCCCFF">        <p align="left"><font color="#660066"><b>Test Case ID : 1650</b></font></p>      </td>    </tr>    <tr>      <td bgcolor="#FFFFFF"><b>Short Title : </b>Ensembl Transcript</td>    </tr>    <tr>      <td bgcolor="#FFFFFF"><b>Title : </b>Check whether the Ensembl Transcript files are formatted for Blast</td>    </tr>  </tbody></table><br><table bgColor="#CCCCFF" border="1" cellPadding="0" cellSpacing="0" width="700">  <tbody>    <tr>      <td align="middle" bgColor="#CCCCFF">        <p align="left"><font color="#660066"><b>VITALS</b></font></p>      </td>    </tr>    <tr>      <td>        <table bgColor="#CCCCFF" border="1" cellPadding="2" width="100%">          <tbody>            <tr bgColor="#ffffff">              <td vAlign="top"><b>Product</b></td>              <td vAlign="top" colspan="3">GeneConnect</td>            <tr bgColor="#ffffff">              <td vAlign="top"><b>Component</b></td>              <td vAlign="top">Parsers<br>              </td>              <td vAlign="top"><b>Test Area</b></td>              <td vAlign="top">Blast Destination Parser</td>            </tr>            <tr bgColor="#ffffff">              <td vAlign="top"><b>Reference</b></td>              <td vAlign="top">1</td>              <td vAlign="top"><b>Priority</b></td>              <td vAlign="top">P2</td>            </tr>          </tbody>        </table>    <tr>      <td align="middle" bgColor="#CCCCFF">        <p align="left"><font color="#660066"><b>DESCRIPTION</b></font></p>      </td>    </tr>    <tr>      <td>        <table bgColor="#CCCCFF" border="1" cellPadding="2" width="100%" height="113">          <tbody>            <tr bgColor="#ffffff"><td vAlign="top" height="32" colspan="4"><font class="itsmultlinefont"><pre>Purpose:
To ensure that the Ensembl Transcript fasta files are formatted by format-db which are later going to be used as destination database for Sequence Alignment.

Workflow details:
1.The GeneConnect Server downloads the data from the website, the Parsers parse the datafiles to create link files and  the uploader loads the data from the link files in to the respective base tables in the GeneConnect database. This test case is for testing  of the parser: FormatdbEnsemblTrans.pl 
2.This Parser is involved in the GeneConnect First Run ? Preparation run for preparing Blast Destination database.

Setup:
1. Verify following entries in UpdateLinux.sh:
b. The JVM parameter -DentityExpansionLimit = 5000000 has been specified in the java command.
c. The JVM parameter -Xmx512M has been specified in the java command. 
d. Check if JAVA_HOME, ORACLE_HOME, PATH AND CLASS_PATH are set properly.
e. Set the parameters of the java command:
java  -Xmx512M  -DentityExpansionLimit=5000000 -Djava.security.manager  -Djava.security.policy=java.policy com.dataminer.server.jobmanager.JobManager

2. Verify the path of the input data file (EnsemblTranscriptHumanTestFile.fa, which resides in the main folder from which GeneConnect server is run) in the config file (/Config/CommandFile.xml).

3. Check the server.properties file (in the 'Config' folder) and verify the database driver name and the dbURL and other database arguments including the TNS service name for the Oracle database, which can be any one of the following:
a) oracle.jdbc.driver.OracleDriver : for Driver name
b) jdbc:oracle:oci8:@&lt;database name&gt; : for database URL 

4. Check the external FormatdbEnsemblTrans.pl Perl Parser set up as follows:
Ensure that the external parser's .sh file is present in the 'ParserScripts' folder.

5. Add the following entry in Config/CommandFile.xml. This denotes that 
a) Test file is present in '/GeneConnectServer' directory of local machine. 
b) FormatdbEnsemblTrans.sh will call the external perl parser

&lt;DataSource name="ENSEMBL_TRANSCRIPT" type="" siteurl="opt1" 
	 	username="anonymous" password=" "&gt;
	 	&lt;ExternalParser writeToDB="false" commandfile="ParserScripts/FormatdbEnsemblTrans.sh"/&gt;
		&lt;BaseDirectory dir="/GeneConnectServer" islocal="true"&gt;
	 		&lt;File name="EnsemblTranscriptHumanTestFile.fa"/&gt;
	 	&lt;/BaseDirectory&gt;		
&lt;/DataSource&gt; 

6. Verify that Perl, Bioperl and Blast (standalone) is installed and   paths of Perl and Bioperl are specified in the UpdateLinux.sh file.

7. Run the script UpdateLinux.sh. 

Input/Data:
The test input file provided with the deliverable i.e.EnsemblTranscriptHumanTestFile.fa  [The input files are taken organism wise i.e. formating is done for a single organism file at a time.]

Expected Output:
1.Check if the following three files are formed:
EnsemblTranscriptHumanTestFile.fa.nin
EnsemblTranscriptHumanTestFile.fa.nsq EnsemblTranscriptHumanTestFile.fa.nhr 

2. If the above files are not formed, check formatdb.log files which provides the reason for failure of formatting.(e.g. Sequence#1 missing, Empty database etc.).
</pre></font></td>            </tr>            <tr bgColor="#ffffff">              <td vAlign="top" height="21"><b>Mode of Execution</b></td>              <td vAlign="top" height="1">Manual                </td>              <td vAlign="top" height="21"><b>Verification Logic</b></td>              <td vAlign="top" height="1">UI<br>                </td>            </tr>            <tr bgColor="#ffffff">              <td vAlign="top" height="21"><b>Author</b></td>              <td vAlign="top" height="21">Shweta Chavan</td>              <td vAlign="top" height="21"><b>Date Entered</b></td>              <td vAlign="top" height="21">01/08/2006&nbsp;</td>            </tr>            <tr bgColor="#ffffff">              <td vAlign="top" height="21"></td>              <td vAlign="top" height="21"></td>              <td vAlign="top" height="21"><b>Date Last Update</b></td>              <td vAlign="top" height="21">08/09/2006&nbsp;</td>            </tr>          </tbody>        </table>      </td>    </tr>  </tbody></table>
<p align="right"><a href="#TMT Generated Test Plans">Top</a></p><a name="Test Case ID: GeneConnectver 1.1ParsersBlast Destination ParserEnsembl Protein"><center><table bgColor="#CCCCFF" border="1" cellPadding="0" cellSpacing="0" width="700">  <tbody>    <tr>      <td align="middle" bgColor="#CCCCFF">        <p align="left"><font color="#660066"><b>Test Case ID : 1651</b></font></p>      </td>    </tr>    <tr>      <td bgcolor="#FFFFFF"><b>Short Title : </b>Ensembl Protein</td>    </tr>    <tr>      <td bgcolor="#FFFFFF"><b>Title : </b>Check whether the Ensembl Protein fasta files are formatted for Blast</td>    </tr>  </tbody></table><br><table bgColor="#CCCCFF" border="1" cellPadding="0" cellSpacing="0" width="700">  <tbody>    <tr>      <td align="middle" bgColor="#CCCCFF">        <p align="left"><font color="#660066"><b>VITALS</b></font></p>      </td>    </tr>    <tr>      <td>        <table bgColor="#CCCCFF" border="1" cellPadding="2" width="100%">          <tbody>            <tr bgColor="#ffffff">              <td vAlign="top"><b>Product</b></td>              <td vAlign="top" colspan="3">GeneConnect</td>            <tr bgColor="#ffffff">              <td vAlign="top"><b>Component</b></td>              <td vAlign="top">Parsers<br>              </td>              <td vAlign="top"><b>Test Area</b></td>              <td vAlign="top">Blast Destination Parser</td>            </tr>            <tr bgColor="#ffffff">              <td vAlign="top"><b>Reference</b></td>              <td vAlign="top">1</td>              <td vAlign="top"><b>Priority</b></td>              <td vAlign="top">P2</td>            </tr>          </tbody>        </table>    <tr>      <td align="middle" bgColor="#CCCCFF">        <p align="left"><font color="#660066"><b>DESCRIPTION</b></font></p>      </td>    </tr>    <tr>      <td>        <table bgColor="#CCCCFF" border="1" cellPadding="2" width="100%" height="113">          <tbody>            <tr bgColor="#ffffff"><td vAlign="top" height="32" colspan="4"><font class="itsmultlinefont"><pre>Purpose:
To ensure that the Ensembl Protein fasta files are formatted by format-db which are later going to be used as destination database for Sequence Alignment.

Workflow details:
1.The GeneConnect Server downloads the data from the website, the Parsers parse the datafiles to create link files and  the uploader loads the data from the link files in to the respective base tables in the GeneConnect database. This test case is for testing  of the parser: FormatdbEnsemblProt.pl 
2.This Parser is involved in the GeneConnect First Run ? Preparation run for preparing Blast Destination database.

Setup:
1. Verify following entries in UpdateLinux.sh:
b. The JVM parameter -DentityExpansionLimit = 5000000 has been specified in the java command.
c. The JVM parameter -Xmx512M has been specified in the java command. 
d. Check if JAVA_HOME, ORACLE_HOME, PATH AND CLASS_PATH are set properly.
e. Set the parameters of the java command:
java  -Xmx512M  -DentityExpansionLimit=5000000 -Djava.security.manager  -Djava.security.policy=java.policy com.dataminer.server.jobmanager.JobManager

2. Verify the path of the input data file (EnsemblProteinHumanTestFile.fa, which resides in the main folder from which GeneConnect server is run) in the config file (/Config/CommandFile.xml) . 
Also verify that the properties file, properties_blastdestnensmblprot.txt is present in '/GeneConnectServer/ParserScripts' folder.It must have following parameters:
IsDBTypeProtein=T
DBName=Human_Ensemblprotein (as per the organism name and database name) 

3. Check the server.properties file (in the 'Config' folder) and verify the database driver name and the dbURL and other database arguments including the TNS service name for the Oracle database, which can be any one of the following:
a) oracle.jdbc.driver.OracleDriver : for Driver name
b) jdbc:oracle:oci8:@&lt;database name&gt; : for database URL

4. Check the external FormatdbEnsemblProt.pl Perl Parser set up as follows:
Ensure that the external parser's .sh file is present in the 'ParserScripts' folder.

5. Add the following entry in Config/CommandFile.xml.This denotes that 
a) Test file is present in '/GeneConnectServer' directory of local machine. 
b) FormatdbSingle.sh will call the external perl parser

&lt;DataSource name="ENSEMBL_PROTEIN" type="" siteurl="opt1" 
	 	username="anonymous" password=" "&gt;
	 	&lt;ExternalParser writeToDB="false" commandfile="ParserScripts/FormatdbEnsemblProt.sh"/&gt;
		&lt;BaseDirectory dir="/GeneConnectServer" islocal="true"&gt;
	 		&lt;File name="EnsemblProteinHumanTestFile.fa"/&gt;
	 	&lt;/BaseDirectory&gt;		
&lt;/DataSource&gt; 

6. Verify that Perl, Bioperl and Blast (standalone) is installed and   paths of Perl and Bioperl are specified in the UpdateLinux.sh file.

7. Run the script UpdateLinux.sh. 

Input/Data:
The test input file provided with the deliverable: EnsemblProteinHumanTestFile.fa [The input files are taken organism wise i.e. formating is done for a single organism file at a time.]

Expected Output:
1.Check if the following three files are formed:
EnsemblProteinHumanTestFile.fa.pin
EnsemblProteinHumanTestFile.fa.psq
EnsemblProteinHumanTestFile.fa.phr  

2.If the above files are not formed, check formatdb.log files which provides the reason for failure of formatting.(e.g. Sequence#1 missing, Empty database etc.).

</pre></font></td>            </tr>            <tr bgColor="#ffffff">              <td vAlign="top" height="21"><b>Mode of Execution</b></td>              <td vAlign="top" height="1">Manual                </td>              <td vAlign="top" height="21"><b>Verification Logic</b></td>              <td vAlign="top" height="1">UI<br>                </td>            </tr>            <tr bgColor="#ffffff">              <td vAlign="top" height="21"><b>Author</b></td>              <td vAlign="top" height="21">Shweta Chavan</td>              <td vAlign="top" height="21"><b>Date Entered</b></td>              <td vAlign="top" height="21">01/08/2006&nbsp;</td>            </tr>            <tr bgColor="#ffffff">              <td vAlign="top" height="21"></td>              <td vAlign="top" height="21"></td>              <td vAlign="top" height="21"><b>Date Last Update</b></td>              <td vAlign="top" height="21">08/09/2006&nbsp;</td>            </tr>          </tbody>        </table>      </td>    </tr>  </tbody></table>
<p align="right"><a href="#TMT Generated Test Plans">Top</a></p><a name="Test Case ID: GeneConnectver 1.1ParsersBlast Destination ParserRefseq mrna"><center><table bgColor="#CCCCFF" border="1" cellPadding="0" cellSpacing="0" width="700">  <tbody>    <tr>      <td align="middle" bgColor="#CCCCFF">        <p align="left"><font color="#660066"><b>Test Case ID : 1653</b></font></p>      </td>    </tr>    <tr>      <td bgcolor="#FFFFFF"><b>Short Title : </b>Refseq mrna</td>    </tr>    <tr>      <td bgcolor="#FFFFFF"><b>Title : </b>Check whether the Refseq mrna files are formatted for Blast</td>    </tr>  </tbody></table><br><table bgColor="#CCCCFF" border="1" cellPadding="0" cellSpacing="0" width="700">  <tbody>    <tr>      <td align="middle" bgColor="#CCCCFF">        <p align="left"><font color="#660066"><b>VITALS</b></font></p>      </td>    </tr>    <tr>      <td>        <table bgColor="#CCCCFF" border="1" cellPadding="2" width="100%">          <tbody>            <tr bgColor="#ffffff">              <td vAlign="top"><b>Product</b></td>              <td vAlign="top" colspan="3">GeneConnect</td>            <tr bgColor="#ffffff">              <td vAlign="top"><b>Component</b></td>              <td vAlign="top">Parsers<br>              </td>              <td vAlign="top"><b>Test Area</b></td>              <td vAlign="top">Blast Destination Parser</td>            </tr>            <tr bgColor="#ffffff">              <td vAlign="top"><b>Reference</b></td>              <td vAlign="top">1</td>              <td vAlign="top"><b>Priority</b></td>              <td vAlign="top">P2</td>            </tr>          </tbody>        </table>    <tr>      <td align="middle" bgColor="#CCCCFF">        <p align="left"><font color="#660066"><b>DESCRIPTION</b></font></p>      </td>    </tr>    <tr>      <td>        <table bgColor="#CCCCFF" border="1" cellPadding="2" width="100%" height="113">          <tbody>            <tr bgColor="#ffffff"><td vAlign="top" height="32" colspan="4"><font class="itsmultlinefont"><pre>Purpose:
To ensure that the Refseq mrna fasta files are formatted by format-db which are later going to be used as destination database for Sequence Alignment.

Workflow details:
1.The GeneConnect Server downloads the data from the website, the Parsers parse the datafiles to create link files and  the uploader loads the data from the link files in to the respective base tables in the GeneConnect database. This test case is for testing  of the parser: FormatdbRefSeqmRNA.pl 
2.This Parser is involved in the GeneConnect First Run ? Preparation run for preparing Blast Destination database.

Setup:
1. Verify following entries in UpdateLinux.sh:
b. The JVM parameter -DentityExpansionLimit = 5000000 has been specified in the java command.
c. The JVM parameter -Xmx512M has been specified in the java command. 
d. Check if JAVA_HOME, ORACLE_HOME, PATH AND CLASS_PATH are set properly.
e. Set the parameters of the java command:
java  -Xmx512M  -DentityExpansionLimit=5000000 -Djava.security.manager  -Djava.security.policy=java.policy com.dataminer.server.jobmanager.JobManager

2. Verify the path of the input data file (RefseqmrnaHumanTestFile.fna, which resides in the main folder from which GeneConnect server is run) in the config file (/Config/CommandFile.xml).
Also verify that the properties file, properties_blastdestnrefmrna.txt is present in '/GeneConnectServer/ParserScripts' folder.It must have following parameters:
IsDBTypeProtein=F
DBName=Human_Refseqmrna(as per the organism name and database name)
 
3. Check the server.properties file (in the 'Config' folder) and verify the database driver name and the dbURL and other database arguments including the TNS service name for the Oracle database, which can be any one of the following:
a) oracle.jdbc.driver.OracleDriver : for Driver name
b) jdbc:oracle:oci8:@&lt;database name&gt; : for database URL

4. Check the external FormatdbRefSeqmRNA.pl Perl Parser set up as follows:
Ensure that the external parser's .sh file is present in the 'ParserScripts' folder.

5. Add the following entry in Config/CommandFile.xml. This denotes that 
a) Test file is present in '/GeneConnectServer' directory of local machine. 
b) FormatdbRefSeqmRNA.sh will call the external perl parser

&lt;DataSource name="REFSEQ_MRNA" type="" siteurl="opt1" 
	 	username="anonymous" password=" "&gt;
	 	&lt;ExternalParser writeToDB="false" commandfile="ParserScripts/FormatdbRefSeqmRNA.sh"/&gt;
		&lt;BaseDirectory dir="/GeneConnectServer" islocal="true"&gt;
	 		&lt;File name="RefseqmrnaHumanTestFile.fna"/&gt;
	 	&lt;/BaseDirectory&gt;		
&lt;/DataSource&gt;

6. Verify that Perl, Bioperl and Blast (standalone) is installed and   paths of Perl and Bioperl are specified in the UpdateLinux.sh file. 

7. Run the script UpdateLinux.sh. 

Input/Data:
The test input file provided with the deliverable i.e. RefseqmrnaHumanTestFile.fna [The input files are taken organism wise i.e. formating is done for a single organism file at a time.]

Expected Output:
1.Check if the following three files are formed : RefseqmrnaHumanTestFile.fna.nin
RefseqmrnaHumanTestFile.fna.nsq
RefseqmrnaHumanTestFile.fna.nhr

2. If the above files are not formed, check formatdb.log files which provides the reason for failure of formatting.(e.g. Sequence#1 missing, Empty database etc.).






</pre></font></td>            </tr>            <tr bgColor="#ffffff">              <td vAlign="top" height="21"><b>Mode of Execution</b></td>              <td vAlign="top" height="1">Manual                </td>              <td vAlign="top" height="21"><b>Verification Logic</b></td>              <td vAlign="top" height="1">UI<br>                </td>            </tr>            <tr bgColor="#ffffff">              <td vAlign="top" height="21"><b>Author</b></td>              <td vAlign="top" height="21">Shweta Chavan</td>              <td vAlign="top" height="21"><b>Date Entered</b></td>              <td vAlign="top" height="21">01/08/2006&nbsp;</td>            </tr>            <tr bgColor="#ffffff">              <td vAlign="top" height="21"></td>              <td vAlign="top" height="21"></td>              <td vAlign="top" height="21"><b>Date Last Update</b></td>              <td vAlign="top" height="21">08/09/2006&nbsp;</td>            </tr>          </tbody>        </table>      </td>    </tr>  </tbody></table>
<p align="right"><a href="#TMT Generated Test Plans">Top</a></p><a name="Test Case ID: GeneConnectver 1.1ParsersBlast Destination ParserRefseq Protein"><center><table bgColor="#CCCCFF" border="1" cellPadding="0" cellSpacing="0" width="700">  <tbody>    <tr>      <td align="middle" bgColor="#CCCCFF">        <p align="left"><font color="#660066"><b>Test Case ID : 1655</b></font></p>      </td>    </tr>    <tr>      <td bgcolor="#FFFFFF"><b>Short Title : </b>Refseq Protein</td>    </tr>    <tr>      <td bgcolor="#FFFFFF"><b>Title : </b>Check whether the Refseq Protein files are formatted for Blast</td>    </tr>  </tbody></table><br><table bgColor="#CCCCFF" border="1" cellPadding="0" cellSpacing="0" width="700">  <tbody>    <tr>      <td align="middle" bgColor="#CCCCFF">        <p align="left"><font color="#660066"><b>VITALS</b></font></p>      </td>    </tr>    <tr>      <td>        <table bgColor="#CCCCFF" border="1" cellPadding="2" width="100%">          <tbody>            <tr bgColor="#ffffff">              <td vAlign="top"><b>Product</b></td>              <td vAlign="top" colspan="3">GeneConnect</td>            <tr bgColor="#ffffff">              <td vAlign="top"><b>Component</b></td>              <td vAlign="top">Parsers<br>              </td>              <td vAlign="top"><b>Test Area</b></td>              <td vAlign="top">Blast Destination Parser</td>            </tr>            <tr bgColor="#ffffff">              <td vAlign="top"><b>Reference</b></td>              <td vAlign="top">1</td>              <td vAlign="top"><b>Priority</b></td>              <td vAlign="top">P2</td>            </tr>          </tbody>        </table>    <tr>      <td align="middle" bgColor="#CCCCFF">        <p align="left"><font color="#660066"><b>DESCRIPTION</b></font></p>      </td>    </tr>    <tr>      <td>        <table bgColor="#CCCCFF" border="1" cellPadding="2" width="100%" height="113">          <tbody>            <tr bgColor="#ffffff"><td vAlign="top" height="32" colspan="4"><font class="itsmultlinefont"><pre>Purpose:
To ensure that the Refseq Protein fasta files are formatted by format-db which are later going to be used as destination database for Sequence Alignment.

Workflow details:
1.The GeneConnect Server downloads the data from the website, the Parsers parse the datafiles to create link files and  the uploader loads the data from the link files in to the respective base tables in the GeneConnect database. This test case is for testing  of the parser: FormatdbRefSeqProt.pl 
2.This Parser is involved in the GeneConnect First Run ? Preparation run for preparing Blast Destination database.

Setup:
1. Verify following entries in UpdateLinux.sh:
b. The JVM parameter -DentityExpansionLimit = 5000000 has been specified in the java command.
c. The JVM parameter -Xmx512M has been specified in the java command. 
d. Check if JAVA_HOME, ORACLE_HOME, PATH AND CLASS_PATH are set properly.
e. Set the parameters of the java command:
java  -Xmx512M  -DentityExpansionLimit=5000000 -Djava.security.manager  -Djava.security.policy=java.policy com.dataminer.server.jobmanager.JobManager

2. Verify the path of the input data file (RefseqproteinHumanTestFile.faa, which resides in the main folder from which GeneConnect server is run) in the config file (/Config/CommandFile.xml).
Also verify that the properties file, properties_blastdestnrefprot.txt is present in '/GeneConnectServer/ParserScripts' folder.It must have following parameters:
IsDBTypeProtein=T
DBName=Human_Refseqprotein(as per the organism name and database name) 

3. Check the server.properties file (in the 'Config' folder) and verify the database driver name and the dbURL and other database arguments including the TNS service name for the Oracle database, which can be any one of the following:
a) oracle.jdbc.driver.OracleDriver : for Driver name
b) jdbc:oracle:oci8:@&lt;database name&gt; : for database URL

4. Check the external FormatdbRefSeqProt.pl Perl Parser set up as follows:
Ensure that the external parser's .sh file is present in the 'ParserScripts' folder.

5. Add the following entry in Config/CommandFile.xml. This denotes that 
a) Test file is present in '/GeneConnectServer' directory of local machine. 
b) FormatdbRefSeqProt.sh will call the external perl parser

&lt;DataSource name="REFSEQ_PROTEIN" type="" siteurl="opt1" 
	 	username="anonymous" password=" "&gt;
	 	&lt;ExternalParser writeToDB="false" commandfile="ParserScripts/FormatdbRefSeqProt.sh"/&gt;
		&lt;BaseDirectory dir="/GeneConnectServer" islocal="true"&gt;
	 		&lt;File name="RefseqproteinHumanTestFile.faa"/&gt;
	 	&lt;/BaseDirectory&gt;		
&lt;/DataSource&gt; 

6. Verify that Perl, Bioperl and Blast (standalone) is installed and   paths of Perl and Bioperl are specified in the UpdateLinux.sh file.

7. Run the script UpdateLinux.sh. 

Input/Data:
The test input file provided with the deliverable i.e.RefseqproteinHumanTestFile.faa [The input files are taken organism wise i.e. formating is done for a single organism file at a time.]

Expected Output:
1.Check if the following files are formed: RefseqproteinHumanTestFile.faa.pin
RefseqproteinHumanTestFile.faa.psq
RefseqproteinHumanTestFile.faa.phr 

2. If the above files are not formed, check formatdb.log files which provides the reason for failure of formatting.(e.g. Sequence#1 missing, Empty database etc.).

</pre></font></td>            </tr>            <tr bgColor="#ffffff">              <td vAlign="top" height="21"><b>Mode of Execution</b></td>              <td vAlign="top" height="1">Manual                </td>              <td vAlign="top" height="21"><b>Verification Logic</b></td>              <td vAlign="top" height="1">UI<br>                </td>            </tr>            <tr bgColor="#ffffff">              <td vAlign="top" height="21"><b>Author</b></td>              <td vAlign="top" height="21">Shweta Chavan</td>              <td vAlign="top" height="21"><b>Date Entered</b></td>              <td vAlign="top" height="21">01/08/2006&nbsp;</td>            </tr>            <tr bgColor="#ffffff">              <td vAlign="top" height="21"></td>              <td vAlign="top" height="21"></td>              <td vAlign="top" height="21"><b>Date Last Update</b></td>              <td vAlign="top" height="21">08/09/2006&nbsp;</td>            </tr>          </tbody>        </table>      </td>    </tr>  </tbody></table>
<p align="right"><a href="#TMT Generated Test Plans">Top</a></p><a name="Test Case ID: GeneConnectver 1.1ParsersBlast Destination ParserUniprot"><center><table bgColor="#CCCCFF" border="1" cellPadding="0" cellSpacing="0" width="700">  <tbody>    <tr>      <td align="middle" bgColor="#CCCCFF">        <p align="left"><font color="#660066"><b>Test Case ID : 1657</b></font></p>      </td>    </tr>    <tr>      <td bgcolor="#FFFFFF"><b>Short Title : </b>Uniprot</td>    </tr>    <tr>      <td bgcolor="#FFFFFF"><b>Title : </b>Check whether the Uniprot files are formatted for Blast</td>    </tr>  </tbody></table><br><table bgColor="#CCCCFF" border="1" cellPadding="0" cellSpacing="0" width="700">  <tbody>    <tr>      <td align="middle" bgColor="#CCCCFF">        <p align="left"><font color="#660066"><b>VITALS</b></font></p>      </td>    </tr>    <tr>      <td>        <table bgColor="#CCCCFF" border="1" cellPadding="2" width="100%">          <tbody>            <tr bgColor="#ffffff">              <td vAlign="top"><b>Product</b></td>              <td vAlign="top" colspan="3">GeneConnect</td>            <tr bgColor="#ffffff">              <td vAlign="top"><b>Component</b></td>              <td vAlign="top">Parsers<br>              </td>              <td vAlign="top"><b>Test Area</b></td>              <td vAlign="top">Blast Destination Parser</td>            </tr>            <tr bgColor="#ffffff">              <td vAlign="top"><b>Reference</b></td>              <td vAlign="top">1</td>              <td vAlign="top"><b>Priority</b></td>              <td vAlign="top">P2</td>            </tr>          </tbody>        </table>    <tr>      <td align="middle" bgColor="#CCCCFF">        <p align="left"><font color="#660066"><b>DESCRIPTION</b></font></p>      </td>    </tr>    <tr>      <td>        <table bgColor="#CCCCFF" border="1" cellPadding="2" width="100%" height="113">          <tbody>            <tr bgColor="#ffffff"><td vAlign="top" height="32" colspan="4"><font class="itsmultlinefont"><pre>Purpose:
To ensure that the Trembl and Swissprot fasta files are parsed to get organism specific fasta files(e.g. UniProt_HUMAN.fasta) and that these organism wise fasta files are formatted by format-db so that they are ready to act as Blast destinations.

Workflow details:
1.The GeneConnect Server downloads the data from the website, the Parsers parse the datafiles to create link files and  the uploader loads the data from the link files in to the respective base tables in the GeneConnect database. This test case is for testing  of the parser: FormatdbUniprot.pl 
2.This Parser is involved in the GeneConnect First Run ? Preparation run for preparing Blast Destination database.

Setup:
1. Verify following entries in UpdateLinux.sh:
b. The JVM parameter -DentityExpansionLimit = 5000000 has been specified in the java command.
c. The JVM parameter -Xmx512M has been specified in the java command. 
d. Check if JAVA_HOME, ORACLE_HOME, PATH AND CLASS_PATH are set properly.
e. Set the parameters of the java command:
java  -Xmx512M  -DentityExpansionLimit=5000000 -Djava.security.manager  -Djava.security.policy=java.policy com.dataminer.server.jobmanager.JobManager

2. Verify the path of the input data files (uniprot_sprot_TestFile.fasta, uniprot_trembl_TestFile.fasta, which resides in the main folder from which GeneConnect server is run) in the config file (/Config/CommandFile.xml). 
Also verify that the properties file, properties_blastdestnuniprot.txt is present in '/GeneConnectServer/ParserScripts' folder.It must have following parameters:
IsDBTypeProtein=T
Organism=HUMAN,MOUSE,RAT

3. Check the server.properties file (in the 'Config' folder) and verify the database driver name and the dbURL and other database arguments including the TNS service name for the Oracle database, which can be any one of the following:
a) oracle.jdbc.driver.OracleDriver : for Driver name
b) jdbc:oracle:oci8:@&lt;database name&gt; : for database URL

4. Check the external FormatdbUniprot.pl Perl Parser set up as follows:
Ensure that the external parser's .sh file is present in the 'ParserScripts' folder.

5. Add the following entry in Config/CommandFile.xml.This denotes that 
a) Test file is present in '/GeneConnectServer' directory of local machine. 
b) FormatdbUniprot.sh will call the external perl parser

&lt;DataSource name="UNIPROT" type="" siteurl="opt1" 
	 	username="anonymous" password=" "&gt;
	 	&lt;ExternalParser writeToDB="false" commandfile="ParserScripts/FormatdbUniprot.sh "/&gt;
		&lt;BaseDirectory dir="/GeneConnectServer" islocal="true"&gt;
	 		&lt;File name="uniprot_sprot_TestFile.fasta"/&gt;
	 	&lt;/BaseDirectory&gt;		
&lt;/DataSource&gt; 

6. Verify that Perl, Bioperl and Blast (standalone) is installed and   paths of Perl and Bioperl are specified in the UpdateLinux.sh file.

7. Run the script UpdateLinux.sh. 

Input/Data:
The test input files are provided with the deliverable.[There are two input files provided here, uniprot_sprot_TestFile.fasta and uniprot_trembl_TestFile.fasta]

Expected Output:
1.Check whether three intermediate files for each input file are formed.
a)uniprot_sprot_TestFile_HUMAN.fasta
b)uniprot_sprot_TestFile_MOUSE.fasta
c)uniprot_sprot_TestFile_RAT.fasta
and
a)uniprot_trembl_TestFile_HUMAN.fasta
b)uniprot_trembl_TestFile_MOUSE.fasta
c)uniprot_trembl_TestFile_RAT.fasta

2.Check whether the following three files with extensions , '.pin' , '.psq', '.phr' are formed for each of the above mentioned files. e.g.for the file uniprot_sprot_TestFile_HUMAN.fasta:
a)uniprot_sprot_TestFile_HUMAN.fasta.pin
b)uniprot_sprot_TestFile_HUMAN.fasta.psq
c)uniprot_sprot_TestFile_HUMAN.fasta.phr 
are formed.

3. If the above files are not formed, check the log files formed.For e.g. formatdb.log file, which provides the reason for failure of formatting.(e.g. Sequence#1 missing, Empty database etc.).




</pre></font></td>            </tr>            <tr bgColor="#ffffff">              <td vAlign="top" height="21"><b>Mode of Execution</b></td>              <td vAlign="top" height="1">Manual                </td>              <td vAlign="top" height="21"><b>Verification Logic</b></td>              <td vAlign="top" height="1">UI<br>                </td>            </tr>            <tr bgColor="#ffffff">              <td vAlign="top" height="21"><b>Author</b></td>              <td vAlign="top" height="21">Shweta Chavan</td>              <td vAlign="top" height="21"><b>Date Entered</b></td>              <td vAlign="top" height="21">02/08/2006&nbsp;</td>            </tr>            <tr bgColor="#ffffff">              <td vAlign="top" height="21"></td>              <td vAlign="top" height="21"></td>              <td vAlign="top" height="21"><b>Date Last Update</b></td>              <td vAlign="top" height="21">08/09/2006&nbsp;</td>            </tr>          </tbody>        </table>      </td>    </tr>  </tbody></table>
<p align="right"><a href="#TMT Generated Test Plans">Top</a></p><a name="Test Area Name: GeneConnectver 1.1ParsersBlast Source Parser"><center><table bgColor="#CCCCFF" border="1" cellPadding="0" cellSpacing="0" width="700">  <tbody>    <tr>      <td align="middle" bgColor="#CCCCFF">        <p align="left"><font color="#660066"><b>Test Area Name: Blast Source Parser</b></font></p> </td> </tr> <tr> <td>&nbsp;</td> </tr> </tbody></table><br><table bgColor="#CCCCFF" border="1" cellPadding="0" cellSpacing="0" width="700"> <tbody> <tr> <td align="middle" bgColor="#CCCCFF"> <p align="left"><font color="#660066"><b>VITALS</b></font></p> </td> </tr> <tr> <td> <table bgColor="#CCCCFF" border="1" cellPadding="2" width="696"> <tbody> <tr bgColor="#ffffff"> <td vAlign="top" width="131"><b>Reference</b></td> <td vAlign="top" width="529">&nbsp;<a href="" target="_blank"></a></td> </tr> </tbody> </table> <tr> <td align="middle" bgColor="#CCCCFF"> <p align="left"><font color="#660066"><b>DESCRIPTION</b></font></p> </td> </tr> <tr> <td> <table bgColor="#CCCCFF" border="1" cellPadding="2" width="100%" height="55"> <tbody> <tr bgColor="#ffffff"> <td vAlign="top" height="1"></td> </tr> </tbody> </table> </td> </tr> </tbody></table>
<p align="right"><a href="#TMT Generated Test Plans">Top</a></p><a name="Test Case ID: GeneConnectver 1.1ParsersBlast Source ParserGenBankmRNA_01"><center><table bgColor="#CCCCFF" border="1" cellPadding="0" cellSpacing="0" width="700">  <tbody>    <tr>      <td align="middle" bgColor="#CCCCFF">        <p align="left"><font color="#660066"><b>Test Case ID : 1652</b></font></p>      </td>    </tr>    <tr>      <td bgcolor="#FFFFFF"><b>Short Title : </b>GenBankmRNA_01</td>    </tr>    <tr>      <td bgcolor="#FFFFFF"><b>Title : </b>GenBank mRNA FASTA</td>    </tr>  </tbody></table><br><table bgColor="#CCCCFF" border="1" cellPadding="0" cellSpacing="0" width="700">  <tbody>    <tr>      <td align="middle" bgColor="#CCCCFF">        <p align="left"><font color="#660066"><b>VITALS</b></font></p>      </td>    </tr>    <tr>      <td>        <table bgColor="#CCCCFF" border="1" cellPadding="2" width="100%">          <tbody>            <tr bgColor="#ffffff">              <td vAlign="top"><b>Product</b></td>              <td vAlign="top" colspan="3">GeneConnect</td>            <tr bgColor="#ffffff">              <td vAlign="top"><b>Component</b></td>              <td vAlign="top">Parsers<br>              </td>              <td vAlign="top"><b>Test Area</b></td>              <td vAlign="top">Blast Source Parser</td>            </tr>            <tr bgColor="#ffffff">              <td vAlign="top"><b>Reference</b></td>              <td vAlign="top">1</td>              <td vAlign="top"><b>Priority</b></td>              <td vAlign="top">P1</td>            </tr>          </tbody>        </table>    <tr>      <td align="middle" bgColor="#CCCCFF">        <p align="left"><font color="#660066"><b>DESCRIPTION</b></font></p>      </td>    </tr>    <tr>      <td>        <table bgColor="#CCCCFF" border="1" cellPadding="2" width="100%" height="113">          <tbody>            <tr bgColor="#ffffff"><td vAlign="top" height="32" colspan="4"><font class="itsmultlinefont"><pre>Purpose:
To ensure that the GenBank files(genbank format flat files) are parsed to get organism wise GenBank mRNA fasta files so that be the source files for Sequence Alignment.

Workflow details:
1.The GeneConnect Server downloads the data from the website, the Parsers parse the datafiles to create link files and  the uploader loads the data from the link files in to the respective base tables in the GeneConnect database. This test case is for testing  of the parser: GenbankmRNAParser.pl 
2.This Parser is involved in the GeneConnect First Run ? Preparation run for preparing organism wise fasta files for BLAST.

Setup:
1. Verify following entries in UpdateLinux.bat:
b. The JVM parameter -DentityExpansionLimit = 5000000 has been specified in the java command.
c. The JVM parameter -Xmx512M has been specified in the java command. 
d. Check if JAVA_HOME, ORACLE_HOME, PATH AND CLASS_PATH are set properly.
e. Set the parameters of the java command:
java  -Xmx512M  -DentityExpansionLimit=5000000 -Djava.security.manager  -Djava.security.policy=java.policy com.dataminer.server.jobmanager.JobManager

2. Verify the path of the input data file (gbankmrnaTestFile.seq, which resides in the main folder from which GeneConnect server is run) in the config file (/Config/CommandFile.xml). 

3. Check the server.properties file (in the 'Config' folder) and verify the database driver name and the dbURL and other database arguments including the TNS service name for the Oracle database, which can be any one of the following:
a) oracle.jdbc.driver.OracleDriver : for Driver name
b) jdbc:oracle:oci8:@&lt;database name&gt; : for database URL

4. Check the external GenbankmRNAParser.pl Perl Parser set up as follows:
a) Ensure that the external parser's .sh file (genbankmrna.sh) is present in the 'ParserScripts' folder.

5. Add the following entry in Config/CommandFile.xml. This denotes that 
a) Test file is present in '/GeneConnectServer' directory of local machine. 
b) GenbankmRNAParser.sh will call the external perl parser

&lt;DataSource name="GENBANK mRNA" type="" siteurl="opt1" 
	 	username="anonymous" password=" "&gt;
	 	&lt;ExternalParser writeToDB="false" commandfile="ParserScripts/GenbankmRNAParser.sh"/&gt;
		&lt;BaseDirectory dir="/GeneConnectServer" islocal="true"&gt;
	 		&lt;File name="gbankmrnaTestFile.seq"/&gt;
	 	&lt;/BaseDirectory&gt;		
&lt;/DataSource&gt; 

6. Verify that Perl, Bioperl and Blast (standalone) is installed and   paths of Perl and Bioperl are specified in the UpdateLinux.sh file.

7. Run the script UpdateLinux.sh. 

Input/Data:
The test input file provided with the deliverable.i.e. gbankmrnaTestFile.seq [The input files are taken organism wise i.e. formating is done for a single organism file at a time.]

Expected Output:
1. Check if the following files are formed.
genbank_mrna_Homo_sapiens.fasta
genbank_mrna_Mus_musculus.fasta
genbank_mrna_Rattus_norvegicus.fasta  

</pre></font></td>            </tr>            <tr bgColor="#ffffff">              <td vAlign="top" height="21"><b>Mode of Execution</b></td>              <td vAlign="top" height="1">Manual                </td>              <td vAlign="top" height="21"><b>Verification Logic</b></td>              <td vAlign="top" height="1">DATABASE<br>                </td>            </tr>            <tr bgColor="#ffffff">              <td vAlign="top" height="21"><b>Author</b></td>              <td vAlign="top" height="21">Arun Venkatesan</td>              <td vAlign="top" height="21"><b>Date Entered</b></td>              <td vAlign="top" height="21">01/08/2006&nbsp;</td>            </tr>            <tr bgColor="#ffffff">              <td vAlign="top" height="21"></td>              <td vAlign="top" height="21"></td>              <td vAlign="top" height="21"><b>Date Last Update</b></td>              <td vAlign="top" height="21">08/09/2006&nbsp;</td>            </tr>          </tbody>        </table>      </td>    </tr>  </tbody></table>
<p align="right"><a href="#TMT Generated Test Plans">Top</a></p><a name="Test Case ID: GeneConnectver 1.1ParsersBlast Source ParserGenBankProtein_01"><center><table bgColor="#CCCCFF" border="1" cellPadding="0" cellSpacing="0" width="700">  <tbody>    <tr>      <td align="middle" bgColor="#CCCCFF">        <p align="left"><font color="#660066"><b>Test Case ID : 1654</b></font></p>      </td>    </tr>    <tr>      <td bgcolor="#FFFFFF"><b>Short Title : </b>GenBankProtein_01</td>    </tr>    <tr>      <td bgcolor="#FFFFFF"><b>Title : </b>Genbank protein FASTA file</td>    </tr>  </tbody></table><br><table bgColor="#CCCCFF" border="1" cellPadding="0" cellSpacing="0" width="700">  <tbody>    <tr>      <td align="middle" bgColor="#CCCCFF">        <p align="left"><font color="#660066"><b>VITALS</b></font></p>      </td>    </tr>    <tr>      <td>        <table bgColor="#CCCCFF" border="1" cellPadding="2" width="100%">          <tbody>            <tr bgColor="#ffffff">              <td vAlign="top"><b>Product</b></td>              <td vAlign="top" colspan="3">GeneConnect</td>            <tr bgColor="#ffffff">              <td vAlign="top"><b>Component</b></td>              <td vAlign="top">Parsers<br>              </td>              <td vAlign="top"><b>Test Area</b></td>              <td vAlign="top">Blast Source Parser</td>            </tr>            <tr bgColor="#ffffff">              <td vAlign="top"><b>Reference</b></td>              <td vAlign="top">1</td>              <td vAlign="top"><b>Priority</b></td>              <td vAlign="top">P1</td>            </tr>          </tbody>        </table>    <tr>      <td align="middle" bgColor="#CCCCFF">        <p align="left"><font color="#660066"><b>DESCRIPTION</b></font></p>      </td>    </tr>    <tr>      <td>        <table bgColor="#CCCCFF" border="1" cellPadding="2" width="100%" height="113">          <tbody>            <tr bgColor="#ffffff"><td vAlign="top" height="32" colspan="4"><font class="itsmultlinefont"><pre>Purpose:
To ensure that the GenBank Protein Fasta file is parsed to get organism wise GenBank Protein fasta files (e.g.genbank_protein_Homo_sapiens.fasta).

Workflow details:
1.The GeneConnect Server downloads the data from the website, the Parsers parse the datafiles to create link files and  the uploader loads the data from the link files in to the respective base tables in the GeneConnect database. This test case is for testing  of the parser: GenbankProteinParser.pl 
2.This Parser is involved in the GeneConnect First Run ? Preparation run for preparing organism wise fasta files for BLAST.

Setup:
1. Verify following entries in UpdateLinux.bat:
b. The JVM parameter -DentityExpansionLimit = 5000000 has been specified in the java command.
c. The JVM parameter -Xmx512M has been specified in the java command. 
d. Check if JAVA_HOME, ORACLE_HOME, PATH AND CLASS_PATH are set properly.
e. Set the parameters of the java command:
java  -Xmx512M  -DentityExpansionLimit=5000000 -Djava.security.manager  -Djava.security.policy=java.policy com.dataminer.server.jobmanager.JobManager

2. Verify the path of the input data file (gbankProteinTestfile.fsa_aa, which resides in the main folder from which GeneConnect server is run) in the config file (/Config/CommandFile.xml). 

3.Check the server.properties file (in the 'Config' folder) and verify the database driver name and the dbURL and other database arguments including the TNS service name for the Oracle database, which can be any one of the following:
a) oracle.jdbc.driver.OracleDriver : for Driver name
b) jdbc:oracle:oci8:@&lt;database name&gt; : for database URL

4. Check the external GenbankmRNAParser.pl Perl Parser set up as follows:
a) Ensure that the external parser's .sh file (genbankprotein.sh) is present in the 'ParserScripts' folder.

5. Add the following entry in Config/CommandFile.xml. This denotes that 
a) Test file is present in '/GeneConnectServer' directory of local machine. 
b) GenbankmRNAParser.sh will call the external perl parser

&lt;DataSource name="GENBANK protein" type="" siteurl="opt1" 
	 	username="anonymous" password=" "&gt;
	 	&lt;ExternalParser writeToDB="false" commandfile="ParserScripts/GenbankProteinParser.sh"/&gt;
		&lt;BaseDirectory dir="/GeneConnectServer" islocal="true"&gt;
	 		&lt;File name="gbankProteinTestfile.fsa_aa"/&gt;
	 	&lt;/BaseDirectory&gt;		
&lt;/DataSource&gt; 

6. Verify that Perl, Bioperl and Blast (standalone) is installed and   paths of Perl and Bioperl are specified in the UpdateLinux.sh file.

7. Run the script UpdateLinux.sh. 

Input/Data:
The test input file provided with the deliverable i.e.gbankProteinTestfile.fsa_aa [The input files are taken organism wise i.e. formating is done for a single organism file at a time.]

Expected Output:
1. Check if the following files are formed: genbank_protein_Homo_sapiens.fasta
gebank_protein_Mus_musculus.fasta 
gebank_protein_Rattus_norvegicus.fasta


</pre></font></td>            </tr>            <tr bgColor="#ffffff">              <td vAlign="top" height="21"><b>Mode of Execution</b></td>              <td vAlign="top" height="1">Manual                </td>              <td vAlign="top" height="21"><b>Verification Logic</b></td>              <td vAlign="top" height="1">DATABASE<br>                </td>            </tr>            <tr bgColor="#ffffff">              <td vAlign="top" height="21"><b>Author</b></td>              <td vAlign="top" height="21">Arun Venkatesan</td>              <td vAlign="top" height="21"><b>Date Entered</b></td>              <td vAlign="top" height="21">01/08/2006&nbsp;</td>            </tr>            <tr bgColor="#ffffff">              <td vAlign="top" height="21"></td>              <td vAlign="top" height="21"></td>              <td vAlign="top" height="21"><b>Date Last Update</b></td>              <td vAlign="top" height="21">08/09/2006&nbsp;</td>            </tr>          </tbody>        </table>      </td>    </tr>  </tbody></table>
<p align="right"><a href="#TMT Generated Test Plans">Top</a></p><a name="Test Area Name: GeneConnectver 1.1ParsersEnsembl External Parser"><center><table bgColor="#CCCCFF" border="1" cellPadding="0" cellSpacing="0" width="700">  <tbody>    <tr>      <td align="middle" bgColor="#CCCCFF">        <p align="left"><font color="#660066"><b>Test Area Name: Ensembl External Parser</b></font></p> </td> </tr> <tr> <td>&nbsp;</td> </tr> </tbody></table><br><table bgColor="#CCCCFF" border="1" cellPadding="0" cellSpacing="0" width="700"> <tbody> <tr> <td align="middle" bgColor="#CCCCFF"> <p align="left"><font color="#660066"><b>VITALS</b></font></p> </td> </tr> <tr> <td> <table bgColor="#CCCCFF" border="1" cellPadding="2" width="696"> <tbody> <tr bgColor="#ffffff"> <td vAlign="top" width="131"><b>Reference</b></td> <td vAlign="top" width="529">&nbsp;<a href="" target="_blank"></a></td> </tr> </tbody> </table> <tr> <td align="middle" bgColor="#CCCCFF"> <p align="left"><font color="#660066"><b>DESCRIPTION</b></font></p> </td> </tr> <tr> <td> <table bgColor="#CCCCFF" border="1" cellPadding="2" width="100%" height="55"> <tbody> <tr bgColor="#ffffff"> <td vAlign="top" height="1"></td> </tr> </tbody> </table> </td> </tr> </tbody></table>
<p align="right"><a href="#TMT Generated Test Plans">Top</a></p><a name="Test Case ID: GeneConnectver 1.1ParsersEnsembl External ParserEnsembl Parser"><center><table bgColor="#CCCCFF" border="1" cellPadding="0" cellSpacing="0" width="700">  <tbody>    <tr>      <td align="middle" bgColor="#CCCCFF">        <p align="left"><font color="#660066"><b>Test Case ID : 1658</b></font></p>      </td>    </tr>    <tr>      <td bgcolor="#FFFFFF"><b>Short Title : </b>Ensembl Parser</td>    </tr>    <tr>      <td bgcolor="#FFFFFF"><b>Title : </b>To check whether the Ensembl mysql database is created and the Ensembl link files are formed</td>    </tr>  </tbody></table><br><table bgColor="#CCCCFF" border="1" cellPadding="0" cellSpacing="0" width="700">  <tbody>    <tr>      <td align="middle" bgColor="#CCCCFF">        <p align="left"><font color="#660066"><b>VITALS</b></font></p>      </td>    </tr>    <tr>      <td>        <table bgColor="#CCCCFF" border="1" cellPadding="2" width="100%">          <tbody>            <tr bgColor="#ffffff">              <td vAlign="top"><b>Product</b></td>              <td vAlign="top" colspan="3">GeneConnect</td>            <tr bgColor="#ffffff">              <td vAlign="top"><b>Component</b></td>              <td vAlign="top">Parsers<br>              </td>              <td vAlign="top"><b>Test Area</b></td>              <td vAlign="top">Ensembl External Parser</td>            </tr>            <tr bgColor="#ffffff">              <td vAlign="top"><b>Reference</b></td>              <td vAlign="top">1</td>              <td vAlign="top"><b>Priority</b></td>              <td vAlign="top">P2</td>            </tr>          </tbody>        </table>    <tr>      <td align="middle" bgColor="#CCCCFF">        <p align="left"><font color="#660066"><b>DESCRIPTION</b></font></p>      </td>    </tr>    <tr>      <td>        <table bgColor="#CCCCFF" border="1" cellPadding="2" width="100%" height="113">          <tbody>            <tr bgColor="#ffffff"><td vAlign="top" height="32" colspan="4"><font class="itsmultlinefont"><pre>Purpose:
To check whether the Ensembl mysql database is created and the Ensembl link files are formed with the correctly parsed data.This test case is written for a Linux machine. 

Workflow details:
1.The GeneConnect Server downloads the data from the website, the Parsers parse the datafiles to create link files and  the uploader loads the data from the link files in to the respective base tables in the GeneConnect database. This test case is for testing  of the parser :Ensembl_Parser.pl
2.This Parser is involved in the GeneConnect Second Run for Annotation Parsers.
3.This parser provides the link files for the following links :
[Figure in the round brackets indicate Link number from the graph for Annotation parsers .]
Graph provided at: http://cabigcvs.nci.nih.gov/viewcvs/viewcvs.cgi/geneconnect/Coding%20and%20Testing/Test%20Plans/workflow.doc?search=None&hideattic=0 

ENSEMBLGENE_ENSEMBLTRANS (6),
ENSEMBLTRANS_ENSEMBLPROT(13),
ENSEMBLGENE_ENTREZGENE(1),
ENSEMBLGENE_UNIGENE(4), 
ENSEMBLTRANS_REFSEQMRNA(10),
ENSEMBLPROT_REFSEQPROTEIN(15),
ENSEMBLPROT_UNIPROT(20),
ENTREZGENE_ENSEMBLGENE(2),
UNIGENE_ENSEMBLGENE(5),
REFSEQMRNA_ENSEMBLTRANS(12),
REFSEQPROTEIN_ENSEMBLPROT(17),
UNIPROT_ENSEMBLPROT(22).



Setup:
1. Verify following entries in UpdateLinux.sh:
b. The JVM parameter -DentityExpansionLimit = 5000000 has been specified in the java command.
c. The JVM parameter -Xmx512M has been specified in the java command. 
d. Check if JAVA_HOME, ORACLE_HOME, PATH AND CLASS_PATH are set properly.
e. Set the parameters of the java command:
java  -Xmx512M  -DentityExpansionLimit=5000000 -Djava.security.manager  -Djava.security.policy=java.policy com.dataminer.server.jobmanager.JobManager

2. Verify the path of the input data files ,(which resides in the main folder from which GeneConnect server is run) in the config file (/Config/CommandFile.xml). 
Also verify that the properties file, ens_properties.txt is present in the '/GeneConnectServer/ParserScripts' folder. It must have the following parameters (Set the paths and name of the database accordingly, following is an example):
organism=homo_sapiens
ensembl_release=39
genome_release=36a
path_to_mysql=/usr/bin
mysqluser=root
mysqlrootpasswd=""
hostname=ps3750.persistent.co.in
dbname=homo_sapiens_core_39_36a


3. Check the server.properties file (in the 'Config' folder) and verify the database driver name and the dbURL and other database arguments including the TNS service name for the Oracle database, which can be any one of the following:
a) oracle.jdbc.driver.OracleDriver : for Driver name
b) jdbc:oracle:oci8:@&lt;database name&gt; : for database URL

4. Ensure that the external perl parser's .sh file is present in the 'ParserScripts' folder.

5. Add the following entry in Config/CommandFile.xml.This denotes that 
a) Test file is present in '/GeneConnectServer' directory of local machine. 
b) ensembl.sh will call the external perl parser

&lt;DataSource name="ENSEMBL" type="" siteurl="" 
 		username="" password=""&gt;
 		&lt;ExternalParser writeToDB="false" commandfile="ParserScripts/ensembl.sh"/&gt;
 		&lt;BaseDirectory dir="/opt1/geneconnect/GeneConnectServer" islocal="true"&gt;
 			&lt;File name="CHECKSUMS.gz" /&gt;
 			&lt;File name="alt_allele.txt.table.gz" /&gt;
 			&lt;File name="analysis.txt.table.gz" /&gt;
 			&lt;File name="analysis_description.txt.table.gz" /&gt;
 			&lt;File name="assembly.txt.table.gz" /&gt;
 			&lt;File name="assembly_bak.txt.table.gz" /&gt;
 			&lt;File name="assembly_exception.txt.table.gz" /&gt;
 			&lt;File name="attrib_type.txt.table.gz" /&gt;
 			&lt;File name="chr2contig.txt.table.gz" /&gt;
 			&lt;File name="clone2contig.txt.table.gz" /&gt;
 			&lt;File name="coord_system.txt.table.gz" /&gt;
 			&lt;File name="coord_system_bak.txt.table.gz" /&gt;
 			&lt;File name="density_feature.txt.table.gz" /&gt;
 			&lt;File name="density_type.txt.table.gz" /&gt;
 			&lt;File name="dna.txt.table.gz" /&gt;
 			&lt;File name="dna_align_feature.txt.table.gz" /&gt;
 			&lt;File name="dnac.txt.table.gz" /&gt;
 			&lt;File name="exon.txt.table.gz" /&gt;
 			&lt;File name="exon_stable_id.txt.table.gz" /&gt;
 			&lt;File name="exon_transcript.txt.table.gz" /&gt;
 			&lt;File name="external_db.txt.table.gz" /&gt;
 			&lt;File name="external_synonym.txt.table.gz" /&gt;
 			&lt;File name="gene.txt.table.gz" /&gt;
 			&lt;File name="gene_archive.txt.table.gz" /&gt;
 			&lt;File name="gene_attrib.txt.table.gz" /&gt;
 			&lt;File name="gene_stable_id.txt.table.gz" /&gt;
 			&lt;File name="go_xref.txt.table.gz" /&gt;
 			&lt;File name="homo_sapiens_core_39_36a.mysql40_compatible.sql.gz" /&gt;
 			&lt;File name="homo_sapiens_core_39_36a.sql.gz" /&gt;
 			&lt;File name="identity_xref.txt.table.gz" /&gt;
 			&lt;File name="interpro.txt.table.gz" /&gt;
 			&lt;File name="karyotype.txt.table.gz" /&gt;
 			&lt;File name="map.txt.table.gz" /&gt;
 			&lt;File name="mapping_session.txt.table.gz" /&gt;
 			&lt;File name="marker.txt.table.gz" /&gt;
 			&lt;File name="marker_feature.txt.table.gz" /&gt;
 			&lt;File name="marker_map_location.txt.table.gz" /&gt;
 			&lt;File name="marker_synonym.txt.table.gz" /&gt;
 			&lt;File name="meta.txt.table.gz" /&gt;
 			&lt;File name="meta_bak.txt.table.gz" /&gt;
 			&lt;File name="meta_coord.txt.table.gz" /&gt;
 			&lt;File name="misc_attrib.txt.table.gz" /&gt;
 			&lt;File name="misc_feature.txt.table.gz" /&gt;
 			&lt;File name="misc_feature_misc_set.txt.table.gz" /&gt;
 			&lt;File name="misc_set.txt.table.gz" /&gt;
 			&lt;File name="object_xref.txt.table.gz" /&gt;
 			&lt;File name="oligo_array.txt.table.gz" /&gt;
 			&lt;File name="oligo_feature.txt.table.gz" /&gt;
 			&lt;File name="oligo_feature_backup.txt.table.gz" /&gt;
 			&lt;File name="oligo_probe.txt.table.gz" /&gt;
 			&lt;File name="peptide_archive.txt.table.gz" /&gt;
 			&lt;File name="prediction_exon.txt.table.gz" /&gt;
 			&lt;File name="prediction_transcript.txt.table.gz" /&gt;
 			&lt;File name="protein_align_feature.txt.table.gz" /&gt;
 			&lt;File name="protein_feature.txt.table.gz" /&gt;
 			&lt;File name="qtl.txt.table.gz" /&gt;
 			&lt;File name="qtl_feature.txt.table.gz" /&gt;
 			&lt;File name="qtl_synonym.txt.table.gz" /&gt;
 			&lt;File name="regulatory_factor.txt.table.gz" /&gt;
 			&lt;File name="regulatory_factor_coding.txt.table.gz" /&gt;
 			&lt;File name="regulatory_feature.txt.table.gz" /&gt;
 			&lt;File name="regulatory_feature_object.txt.table.gz" /&gt;
 			&lt;File name="regulatory_search_region.txt.table.gz" /&gt;
 			&lt;File name="repeat_consensus.txt.table.gz" /&gt;
 			&lt;File name="repeat_feature.txt.table.gz" /&gt;
 			&lt;File name="sc2contig.txt.table.gz" /&gt;
 			&lt;File name="seq_region.txt.table.gz" /&gt;
 			&lt;File name="seq_region_attrib.txt.table.gz" /&gt;
 			&lt;File name="seq_region_bak.txt.table.gz" /&gt;
 			&lt;File name="simple_feature.txt.table.gz" /&gt;
 			&lt;File name="stable_id_event.txt.table.gz" /&gt;
 			&lt;File name="supporting_feature.txt.table.gz" /&gt;
 			&lt;File name="transcript.txt.table.gz" /&gt;
 			&lt;File name="transcript_attrib.txt.table.gz" /&gt;
 			&lt;File name="transcript_stable_id.txt.table.gz" /&gt;
 			&lt;File name="transcript_supporting_feature.txt.table.gz" /&gt;
 			&lt;File name="translation.txt.table.gz" /&gt;
 			&lt;File name="translation_attrib.txt.table.gz" /&gt;
 			&lt;File name="translation_stable_id.txt.table.gz" /&gt;
 			&lt;File name="unmapped_object.txt.table.gz" /&gt;
 			&lt;File name="unmapped_reason.txt.table.gz" /&gt;
 			&lt;File name="xref.txt.table.gz" /&gt;
 		&lt;/BaseDirectory&gt;		
 	&lt;/DataSource&gt; 

5.Verify that Perl, Bioperl is installed and paths of Perl and Bioperl are specified in the UpdateLinux.sh file.DBI and DBD::mysql module,Mysql installation is required.

6. Run the script UpdateLinux.sh.
 
Input/Data:
The homo_sapiens_core_39_36a mysql dump files in a single directory.

Expected Output:
a) Check if the following seven link files are formed :
1.EnsemblGene_EnsemblTranscript.txt
2.EnsemblTranscript_EnsemblPeptide.txt
3.EnsemblGene_EntrezGene.txt
4.EnsemblGene_UniGene.txt
5.EnsemblTranscript_RefSeqmRNA.txt
6.EnsemblPeptide_RefSeqProtein.txt
7.EnsemblPeptide_UniProKBPrimaryAccession.txt
8.Entrezgene_Ensemblgene.txt
9.Unigene_Ensemblgene.txt
10.Refseqmrna_Ensembltranscript.txt
11.Refseqprotein_Ensemblpeptide.txt
12.UniprotKB_Ensemblpeptide.txt

b)The correct data should be loaded into the tables.

Verification Logic:

1.Check the correctness of the data loaded into tables by matching all the Ensembl Gene base tables with the expected output files as listed below:

TABLE EXPECTED-----------------------------------OUTPUT FILE

ENSEMBLGENE_ENSEMBLTRANS----&gt;EnsemblGene_EnsemblTranscript.txt
ENSEMBLTRANS_ENSEMBLPROT----&gt;EnsemblTranscript_EnsemblPeptide.txt
ENSEMBLGENE_ENTREZGENE------&gt;EnsemblGene_EntrezGene.txt
ENSEMBLGENE_UNIGENE---------&gt;EnsemblGene_UniGene.txt
ENSEMBLTRANS_REFSEQMRNA-----&gt;EnsemblTranscript_RefSeqmRNA.txt
ENSEMBLPROT_REFSEQPROTEIN---&gt;EnsemblPeptide_RefSeqProtein.txt
ENSEMBLPROT_UNIPROT---------&gt;EnsemblPeptide_UniProKBPrimaryAccession.txt
ENTREZGENE_ENSEMBLGENE------&gt;EntrezGene_EnsemblGene.txt
UNIGENE_ENSEMBLGENE---------&gt;UniGene_EnsemblGene.txt
REFSEQMRNA_ENSEMBLTRANS-----&gt;RefSeqmRNA_EnsemblTranscript.txt
REFSEQPROTEIN_ENSEMBLPROT---&gt;RefSeqProtein_EnsemblPeptide.txt
UNIPROT_ENSEMBLPROT---------&gt;UniProKBPrimaryAccession_EnsemblPeptide.txt

2. For every Ensembl gene, transcript and protein, the loaded data in the tables should match the information available on Ensembl website at http://www.ensembl.org/index.html for that gene.The method is as follows:

a)EnsemblGene--&gt;EnsemblTranscript:

i) In an Ensembl Gene page of Ensembl database you will find the Ensembl Transcript entry in the section called "Transcripts".This mapping and that in the base table should match.
ii)If there are more than one entry of Ensembl Transcript for an Ensembl Gene(e.g. ENSG00000206466),then they should be present in separate rows in the ensemblgene_ensembltranscript.txt file with the same Ensembl Gene ID.

--------------------------------------------------------------------------
b)EnsemblTranscript--&gt;EnsemblPeptide:

i)Select gene from the drop down menu of the search items and enter the Ensembl Transcript ID here.Click on the Ensembl Gene ,in the section called "Transcript",Ensembl Transcript ID is provided.Below this section 
in "Transcript information" ,click "Protein information".Here,Ensembl Protein entry for the Ensembl Transcript is provided.This mapping and that in the base table should match.


--------------------------------------------------------------------------
c)EnsemblGene--&gt;EntrezGene  and vice versa:

i) In an Ensembl Gene page of Ensembl database you will find the Entrez Gene entry in the section called "Similarity Matches".The EntrezGene entry here, and that in the base table should match.
ii)If there is no Entrez Gene annotation provided for an Ensembl Gene, 
there should be no entry of this Ensembl Gene in the ensemblgene_entrezgene.txt and entrezgene_ensemblgene.txt file and base tables.
iii) If there are more than one entry of Entrez Gene for an Ensembl Gene,
then they should be present in separate rows in the ensemblgene_entrezgene.txt and entrezgene_ensemblgene.txt file with the same Ensembl Gene ID.

--------------------------------------------------------------------------
d)EnsemblGene--&gt;UniGene  and vice versa:

i)Select gene from the drop down menu of the search items and enter the Ensembl Gene ID,click on the Ensembl Gene,in the section called "Similarity Matches",UniGene entry for the Ensembl Gene is provided.This mapping and that in the base table should match.
ii) If there is no UniGene annotation provided for an Ensembl Gene(e.g. ENSG00000206469),there should be no entry of this Ensembl Gene in the ensemblgene_unigene.txt and unigene_ensemblgene.txt file.
iii) If there are more than one entry of UniGene for an Ensembl Gene(e.g. ENSG00000197298),then they should be present in separate lines in the ensemblgene_unigene.txt and unigene_ensemblgene.txt file with the same Ensembl Gene ID.

--------------------------------------------------------------------------
e)EnsemblTranscript--&gt;RefSeqmRNA  and vice versa:

i)Select gene from the drop down menu of the search items and enter the  Ensembl Transcript ID, click on the Ensembl Gene, in the section "Transcript","Similarity Matches", the RefSeq mRNA entry for the Ensembl Transcript is provided.This mapping and that in the base tables should match.
ii)If there is no RefSeq mRNA annotation provided for an Ensembl Transcript(e.g. ),there should be no entry of this Ensembl Gene in the ensembltranscript_refseqmrna.txt and refseqmrna_ensembltranscript.txt file.
iii)If there are more than one entry of RefSeq mRNA for an Ensembl Transcript(e.g. ENST00000355960),then they should be present in separate lines in the ensembltranscript_refseqmrna.txt and refseqmrna_ensembltranscript.txt file with the same Ensembl Transcript ID.
iv)Only the Refseq mRNA accession number(e.g. NM_006398) should be taken and not the version number which is appended to the accssion number(e.g. NM_006398.2 )


--------------------------------------------------------------------------
f)EnsemblPeptide--&gt;RefSeqProtein  and vice versa:

i)Select gene from the drop down menu of the search items and enter the Ensembl Protein ID ,click on the Ensembl Gene,in the section called "Similarity Matches" ,the RefSeq Protein entry for the Ensembl Protein is provided here.This mapping and that in the base tables should match.
ii)If there is no RefSeq Protein annotation provided for an Ensembl Protein there should be no entry of this Ensembl Protein in the ensemblprotein_refseqprotein.txt and refseqprotein_ensemblprotein.txt file.
iii)If there are more than one entry of RefSeq Protein for an Ensembl Transcript(e.g. ENSP00000354632),then they should be present in separate lines in the ensemblprotein_refseqprotein.txt and refseqprotein_ensemblprotein.txt file with the same Ensembl Protein ID.
iv)Only the Refseq Protein accession number(e.g. NP_536848) should be taken and not the version number which is appended to the accssion number(e.g. NP_536848.1 )

--------------------------------------------------------------------------
g)EnsemblPeptide--&gt;UniProKBPrimaryAccession and vice versa:

i)Select gene from the drop down menu of the search items and enter the Ensembl Protein ID ,click on the Ensembl Gene,in the section called "Similarity Matches" , UniProt entry for the Ensembl Protein here.This mapping and that in the base tables should match.
ii)If there is no UniProt annotation provided for an Ensembl Protein there should be no entry of this Ensembl Protein in the ensemblprotein_uniprot.txt and uniprot_ensemblprotein.txt file.
iii)If there are more than one entry of UniProt for an Ensembl Transcript(e.g. ENSP00000373039),then they should be present in separate lines in the ensemblprotein_refseqprotein.txt and uniprot_ensemblprotein.txt file with the same Ensembl Protein ID.

--------------------------------------------------------------------------



 





</pre></font></td>            </tr>            <tr bgColor="#ffffff">              <td vAlign="top" height="21"><b>Mode of Execution</b></td>              <td vAlign="top" height="1">Manual                </td>              <td vAlign="top" height="21"><b>Verification Logic</b></td>              <td vAlign="top" height="1">UI<br>                </td>            </tr>            <tr bgColor="#ffffff">              <td vAlign="top" height="21"><b>Author</b></td>              <td vAlign="top" height="21">Shweta Chavan</td>              <td vAlign="top" height="21"><b>Date Entered</b></td>              <td vAlign="top" height="21">03/08/2006&nbsp;</td>            </tr>            <tr bgColor="#ffffff">              <td vAlign="top" height="21"></td>              <td vAlign="top" height="21"></td>              <td vAlign="top" height="21"><b>Date Last Update</b></td>              <td vAlign="top" height="21">08/09/2006&nbsp;</td>            </tr>          </tbody>        </table>      </td>    </tr>  </tbody></table>
<p align="right"><a href="#TMT Generated Test Plans">Top</a></p><a name="Test Area Name: GeneConnectver 1.1ParsersEntrez Gene parser"><center><table bgColor="#CCCCFF" border="1" cellPadding="0" cellSpacing="0" width="700">  <tbody>    <tr>      <td align="middle" bgColor="#CCCCFF">        <p align="left"><font color="#660066"><b>Test Area Name: Entrez Gene parser</b></font></p> </td> </tr> <tr> <td>&nbsp;</td> </tr> </tbody></table><br><table bgColor="#CCCCFF" border="1" cellPadding="0" cellSpacing="0" width="700"> <tbody> <tr> <td align="middle" bgColor="#CCCCFF"> <p align="left"><font color="#660066"><b>VITALS</b></font></p> </td> </tr> <tr> <td> <table bgColor="#CCCCFF" border="1" cellPadding="2" width="696"> <tbody> <tr bgColor="#ffffff"> <td vAlign="top" width="131"><b>Reference</b></td> <td vAlign="top" width="529">&nbsp;<a href="" target="_blank"></a></td> </tr> </tbody> </table> <tr> <td align="middle" bgColor="#CCCCFF"> <p align="left"><font color="#660066"><b>DESCRIPTION</b></font></p> </td> </tr> <tr> <td> <table bgColor="#CCCCFF" border="1" cellPadding="2" width="100%" height="55"> <tbody> <tr bgColor="#ffffff"> <td vAlign="top" height="1">Link between entrez and unigene</td> </tr> </tbody> </table> </td> </tr> </tbody></table>
<p align="right"><a href="#TMT Generated Test Plans">Top</a></p><a name="Test Case ID: GeneConnectver 1.1ParsersEntrez Gene parserEntrez-Unigene_01"><center><table bgColor="#CCCCFF" border="1" cellPadding="0" cellSpacing="0" width="700">  <tbody>    <tr>      <td align="middle" bgColor="#CCCCFF">        <p align="left"><font color="#660066"><b>Test Case ID : 1606</b></font></p>      </td>    </tr>    <tr>      <td bgcolor="#FFFFFF"><b>Short Title : </b>Entrez-Unigene_01</td>    </tr>    <tr>      <td bgcolor="#FFFFFF"><b>Title : </b>Entrez to Unigene parser</td>    </tr>  </tbody></table><br><table bgColor="#CCCCFF" border="1" cellPadding="0" cellSpacing="0" width="700">  <tbody>    <tr>      <td align="middle" bgColor="#CCCCFF">        <p align="left"><font color="#660066"><b>VITALS</b></font></p>      </td>    </tr>    <tr>      <td>        <table bgColor="#CCCCFF" border="1" cellPadding="2" width="100%">          <tbody>            <tr bgColor="#ffffff">              <td vAlign="top"><b>Product</b></td>              <td vAlign="top" colspan="3">GeneConnect</td>            <tr bgColor="#ffffff">              <td vAlign="top"><b>Component</b></td>              <td vAlign="top">Parsers<br>              </td>              <td vAlign="top"><b>Test Area</b></td>              <td vAlign="top">Entrez Gene parser</td>            </tr>            <tr bgColor="#ffffff">              <td vAlign="top"><b>Reference</b></td>              <td vAlign="top">1</td>              <td vAlign="top"><b>Priority</b></td>              <td vAlign="top">P2</td>            </tr>          </tbody>        </table>    <tr>      <td align="middle" bgColor="#CCCCFF">        <p align="left"><font color="#660066"><b>DESCRIPTION</b></font></p>      </td>    </tr>    <tr>      <td>        <table bgColor="#CCCCFF" border="1" cellPadding="2" width="100%" height="113">          <tbody>            <tr bgColor="#ffffff"><td vAlign="top" height="32" colspan="4"><font class="itsmultlinefont"><pre>Purpose:
To check whether the correct UniGene ID has been mapped to the corresponding Entrez Gene ID in the parsed data by parsing a test input data file (entrez_unigeneTestFile.xml) provided with the deliverable.

Workflow details:
1.The GeneConnect Server downloads the data from the website, the Parsers parse the datafiles to create link files and  the uploader loads the data from the link files in to the respective base tables in the GeneConnect database. This test case is for testing  of the Entrez_Unigene parser .
2.This Parser is involved in the GeneConnect Second Run for Annotation Parsers.
3.This parser provides the link files for the following links :
UNIGENE_ENTREZ (3).

[Figure in the round brackets indicate Link number from the graph for Annotation parsers ]
Graph provided at: http://cabigcvs.nci.nih.gov/viewcvs/viewcvs.cgi/geneconnect/Coding%20and%20Testing/Test%20Plans/workflow.doc?search=None&hideattic=0 

Setup:
1. Verify UpdateLinux.sh script file to check if it exists in the main folder from which GeneConnect server is run. Also verify the following in the UpdateLinux.sh:
a. The path of test config file is correctly set.
b. The JVM parameter -DentityExpansionLimit = 5000000 has been specified in the java command.
c. The JVM parameter -Xmx512M has been specified in the java command. 
d. Check if JAVA_HOME, ORACLE_HOME, PATH AND CLASS_PATH are set properly.
e. Set the parameters of the java command:
java  -Xmx512M  -DentityExpansionLimit=5000000 -Djava.security.manager  -Djava.security.policy=java.policy com.dataminer.server.jobmanager.JobManager

2. Verify the path of the input data file (entrez_unigeneTestFile.xml, which resides in the main folder from which GeneConnect server is run) in the config file CommandFile.xml (in the 'Config' folder).

3. Check the database to ensure that all the base tables required for entrez_unigene data population are present. The names of these tables can be found in the design document. 

4. Check the server.properties file (in the 'Config' folder) and verify the database driver name and the dbURL and other database arguments including the TNS service name for the Oracle database, which can be any one of the following:
a) oracle.jdbc.driver.OracleDriver : for Driver name
b) jdbc:oracle:oci8:@&lt;database name&gt; : for database URL

5.The entrez_unigene parser is available in Annotation parser library as java class.

6. Verify that Perl is installed and the path of Perl is specified in the UpdateLinux.sh file.

7. Add the following entry in Config/CommandFile.xml. This denotes that 
a) The test file is present in '/GeneConnectServer' directory of local machine.
 
&lt;DataSource name="ENTREZ_GENE" type="" siteurl="opt1" 
	 	username="anonymous" password=" "&gt;
	 	&lt;BaseDirectory dir="/GeneConnectServer" islocal="true"&gt;
	 		&lt;File name="entrez_unigeneTestFile.xml"/&gt;
	 	&lt;/BaseDirectory&gt;		
&lt;/DataSource&gt;

7. Run the script UpdateLinux.sh. 

Input/Data:
The test input dat file (entrez_refseqTestFile.xml) provided with the deliverable. 

Expected Output:
1. The correct data should be loaded into the tables.

Verification Logic:

1. Check the correctness of the data loaded into tables as follows:

a) Match all the mappings in the unigeneTestFile.xml file with UniGene base table (i.e.) UNIGENE_ENTREZ. The SQL query is as follows: SELECT * from UNIGENE_ENTREZ WHERE UNIGENE_GENEID in (list of Unigene IDs);

b) Match the data in the NCBI website with the database tables. Go to http://www.ncbi.nlm.nih.gov/ , enter a entrez gene ID, search on 'Gene' from the drop-down, get the corresponding Entrez gene ID by looking up the 'Additional Links' section. The UniGeneID-EntrezGeneID mapping should be the same in both the website and database table.

2. If there is no UniGene annotation provided for an Entrez Gene,there should be no entry of this Entrez Gene in the entrez_unigeneTestFile.xml file.

3. If there are more than one entry of UniGene for an Entrez Gene,then they should be present in separate lines with the same Entrez Gene ID in the entrez_unigeneTestFile.xml file and the base tables.</pre></font></td>            </tr>            <tr bgColor="#ffffff">              <td vAlign="top" height="21"><b>Mode of Execution</b></td>              <td vAlign="top" height="1">Manual                </td>              <td vAlign="top" height="21"><b>Verification Logic</b></td>              <td vAlign="top" height="1">DATABASE<br>                </td>            </tr>            <tr bgColor="#ffffff">              <td vAlign="top" height="21"><b>Author</b></td>              <td vAlign="top" height="21">Arun Venkatesan</td>              <td vAlign="top" height="21"><b>Date Entered</b></td>              <td vAlign="top" height="21">21/07/2006&nbsp;</td>            </tr>            <tr bgColor="#ffffff">              <td vAlign="top" height="21"></td>              <td vAlign="top" height="21"></td>              <td vAlign="top" height="21"><b>Date Last Update</b></td>              <td vAlign="top" height="21">08/09/2006&nbsp;</td>            </tr>          </tbody>        </table>      </td>    </tr>  </tbody></table>
<p align="right"><a href="#TMT Generated Test Plans">Top</a></p><a name="Test Area Name: GeneConnectver 1.1ParsersEntrez_Refseq External Parser"><center><table bgColor="#CCCCFF" border="1" cellPadding="0" cellSpacing="0" width="700">  <tbody>    <tr>      <td align="middle" bgColor="#CCCCFF">        <p align="left"><font color="#660066"><b>Test Area Name: Entrez_Refseq External Parser</b></font></p> </td> </tr> <tr> <td>&nbsp;</td> </tr> </tbody></table><br><table bgColor="#CCCCFF" border="1" cellPadding="0" cellSpacing="0" width="700"> <tbody> <tr> <td align="middle" bgColor="#CCCCFF"> <p align="left"><font color="#660066"><b>VITALS</b></font></p> </td> </tr> <tr> <td> <table bgColor="#CCCCFF" border="1" cellPadding="2" width="696"> <tbody> <tr bgColor="#ffffff"> <td vAlign="top" width="131"><b>Reference</b></td> <td vAlign="top" width="529">&nbsp;<a href="" target="_blank"></a></td> </tr> </tbody> </table> <tr> <td align="middle" bgColor="#CCCCFF"> <p align="left"><font color="#660066"><b>DESCRIPTION</b></font></p> </td> </tr> <tr> <td> <table bgColor="#CCCCFF" border="1" cellPadding="2" width="100%" height="55"> <tbody> <tr bgColor="#ffffff"> <td vAlign="top" height="1"></td> </tr> </tbody> </table> </td> </tr> </tbody></table>
<p align="right"><a href="#TMT Generated Test Plans">Top</a></p><a name="Test Case ID: GeneConnectver 1.1ParsersEntrez_Refseq External ParserEntrezGene_RefSeqmRNA_01"><center><table bgColor="#CCCCFF" border="1" cellPadding="0" cellSpacing="0" width="700">  <tbody>    <tr>      <td align="middle" bgColor="#CCCCFF">        <p align="left"><font color="#660066"><b>Test Case ID : 1580</b></font></p>      </td>    </tr>    <tr>      <td bgcolor="#FFFFFF"><b>Short Title : </b>EntrezGene_RefSeqmRNA_01</td>    </tr>    <tr>      <td bgcolor="#FFFFFF"><b>Title : </b>Check whether the correct  RefSeq mRNA ID is mapped to an Entrez Gene ID</td>    </tr>  </tbody></table><br><table bgColor="#CCCCFF" border="1" cellPadding="0" cellSpacing="0" width="700">  <tbody>    <tr>      <td align="middle" bgColor="#CCCCFF">        <p align="left"><font color="#660066"><b>VITALS</b></font></p>      </td>    </tr>    <tr>      <td>        <table bgColor="#CCCCFF" border="1" cellPadding="2" width="100%">          <tbody>            <tr bgColor="#ffffff">              <td vAlign="top"><b>Product</b></td>              <td vAlign="top" colspan="3">GeneConnect</td>            <tr bgColor="#ffffff">              <td vAlign="top"><b>Component</b></td>              <td vAlign="top">Parsers<br>              </td>              <td vAlign="top"><b>Test Area</b></td>              <td vAlign="top">Entrez_Refseq External Parser</td>            </tr>            <tr bgColor="#ffffff">              <td vAlign="top"><b>Reference</b></td>              <td vAlign="top">1</td>              <td vAlign="top"><b>Priority</b></td>              <td vAlign="top">P3</td>            </tr>          </tbody>        </table>    <tr>      <td align="middle" bgColor="#CCCCFF">        <p align="left"><font color="#660066"><b>DESCRIPTION</b></font></p>      </td>    </tr>    <tr>      <td>        <table bgColor="#CCCCFF" border="1" cellPadding="2" width="100%" height="113">          <tbody>            <tr bgColor="#ffffff"><td vAlign="top" height="32" colspan="4"><font class="itsmultlinefont"><pre>Purpose:
To check whether the correct Entrez Gene ID has been mapped to the corresponding RefSeq mRNA ID in the parsed data by parsing a test input data file provided with the deliverable.

Workflow details:
1.The GeneConnect Server downloads the data from the website, the Parsers parse the datafiles to create link files and  the uploader loads the data from the link files in to the respective base tables in the GeneConnect database. This test case is for testing  of the parser :Entrez_Refseq.pl .
2.This Parser is involved in the GeneConnect Second Run for Annotation Parsers.
3.This parser provides the link files for the following links :
[Figure in the round brackets indicate Link number from the graph for Annotation parsers.]
Graph provided at: http://cabigcvs.nci.nih.gov/viewcvs/viewcvs.cgi/geneconnect/Coding%20and%20Testing/Test%20Plans/workflow.doc?search=None&hideattic=0

ENTREZGENE_REFSEQMRNA (7)

Setup:
1. Verify UpdateLinux.sh script file to check if it exists in the main folder from which GeneConnect server is run. Also verify the following in the UpdateLinux.sh :
a. The path of test config file is correctly set.
b. The JVM parameter -DentityExpansionLimit = 5000000 has been specified in the java command.
c. The JVM parameter -Xmx512M has been specified in the java command. 
d. Check if JAVA_HOME, ORACLE_HOME, PATH AND CLASS_PATH are set properly.
e. Set the parameters of the java command :
java  -Xmx512M  -DentityExpansionLimit=5000000 -Djava.security.manager  -Djava.security.policy=java.policy com.dataminer.server.jobmanager.JobManager

2. Verify the path of the input data file (entrez_refseqTestFile.dat, which resides in the main folder from which GeneConnect server is run) in the config file CommandFile.xml (in the 'Config' folder).

3. Check the database to ensure that all the base tables required for entrez_refseq data population are present. The names of these tables can be found in the design document.

4. Check the server.properties file (in the 'Config' folder) and verify the database driver name and the dbURL and other database arguments including the TNS service name for the Oracle database, which can be any one of the following:
a) oracle.jdbc.driver.OracleDriver : for Driver name
b) jdbc:oracle:oci8:@&lt;database name&gt; : for database URL

5. Check the external entrez_refseq's Perl Parser set up as follows:
a) Ensure that the external parser's .sh file (entrez_refseqparser.sh) is present in the 'ParserScripts' folder.

6. Check whether Perl is installed on the machine and are present in the path of the directory from where testing is being done. Specify this path in the external perl parser code, whenever the Perl path (e.g. /usr/bin/perl) or library path (e.g. Bioperl path) is specified.

7. Add the following entry in Config/CommandFile.xml. This denotes that 
a) the test file is present in '/GeneConnectServer' directory of local machine. 
b) entrez_refseqparser.sh will call the external perl parser

&lt;DataSource name="ENTREZGENE_REFSEQMRNA" type="" siteurl="opt1" 
	 	username="anonymous" password=" "&gt;
	 	&lt;ExternalParser writeToDB="false" commandfile="ParserScripts/entrez_refseqparser.sh"/&gt;
		&lt;BaseDirectory dir="/GeneConnectServer" islocal="true"&gt;
	 		&lt;File name="entrez_refseqTestFile.dat"/&gt;
	 	&lt;/BaseDirectory&gt;		
&lt;/DataSource&gt;

8. Run the script UpdateLinux.sh. 

Input/Data:
The test input dat file provided with the deliverable. 

Expected Output:
1. The output file entrezgene_refseqmrna.txt should be formed.
2. The correct data should be loaded into the tables.

Verification Logic:
1. Check the correctness of the data loaded into tables as follows:

a) Match all the mappings in the entrez_refseqTestFile.dat file with entrez_refseq base table (i.e.) ENTREZGENE_REFSEQMRNA. The SQL query is as follows: "SELECT * from ENTREZGENE_REFSEQMRNA;". 

b) Match the data in the NCBI website with that in the output file and in the database tables. Go to http://www.ncbi.nlm.nih.gov/ , enter a entrez gene ID, search on 'Gene' from the drop-down, and look for 'mRNA Sequence' in the 'NCBI Reference Sequences' section. The EntrezGeneID-RefseqmRNAID mapping should be the same in both the website and table.


</pre></font></td>            </tr>            <tr bgColor="#ffffff">              <td vAlign="top" height="21"><b>Mode of Execution</b></td>              <td vAlign="top" height="1">Manual                </td>              <td vAlign="top" height="21"><b>Verification Logic</b></td>              <td vAlign="top" height="1">UI<br>                </td>            </tr>            <tr bgColor="#ffffff">              <td vAlign="top" height="21"><b>Author</b></td>              <td vAlign="top" height="21">Madhurima</td>              <td vAlign="top" height="21"><b>Date Entered</b></td>              <td vAlign="top" height="21">28/06/2006&nbsp;</td>            </tr>            <tr bgColor="#ffffff">              <td vAlign="top" height="21"></td>              <td vAlign="top" height="21"></td>              <td vAlign="top" height="21"><b>Date Last Update</b></td>              <td vAlign="top" height="21">08/09/2006&nbsp;</td>            </tr>          </tbody>        </table>      </td>    </tr>  </tbody></table>
<p align="right"><a href="#TMT Generated Test Plans">Top</a></p><a name="Test Case ID: GeneConnectver 1.1ParsersEntrez_Refseq External ParserRefseqmRNA_ RefseqProtein_01"><center><table bgColor="#CCCCFF" border="1" cellPadding="0" cellSpacing="0" width="700">  <tbody>    <tr>      <td align="middle" bgColor="#CCCCFF">        <p align="left"><font color="#660066"><b>Test Case ID : 1605</b></font></p>      </td>    </tr>    <tr>      <td bgcolor="#FFFFFF"><b>Short Title : </b>RefseqmRNA_ RefseqProtein_01</td>    </tr>    <tr>      <td bgcolor="#FFFFFF"><b>Title : </b>Refseq mRNA to Refseq protein link</td>    </tr>  </tbody></table><br><table bgColor="#CCCCFF" border="1" cellPadding="0" cellSpacing="0" width="700">  <tbody>    <tr>      <td align="middle" bgColor="#CCCCFF">        <p align="left"><font color="#660066"><b>VITALS</b></font></p>      </td>    </tr>    <tr>      <td>        <table bgColor="#CCCCFF" border="1" cellPadding="2" width="100%">          <tbody>            <tr bgColor="#ffffff">              <td vAlign="top"><b>Product</b></td>              <td vAlign="top" colspan="3">GeneConnect</td>            <tr bgColor="#ffffff">              <td vAlign="top"><b>Component</b></td>              <td vAlign="top">Parsers<br>              </td>              <td vAlign="top"><b>Test Area</b></td>              <td vAlign="top">Entrez_Refseq External Parser</td>            </tr>            <tr bgColor="#ffffff">              <td vAlign="top"><b>Reference</b></td>              <td vAlign="top">1</td>              <td vAlign="top"><b>Priority</b></td>              <td vAlign="top">P2</td>            </tr>          </tbody>        </table>    <tr>      <td align="middle" bgColor="#CCCCFF">        <p align="left"><font color="#660066"><b>DESCRIPTION</b></font></p>      </td>    </tr>    <tr>      <td>        <table bgColor="#CCCCFF" border="1" cellPadding="2" width="100%" height="113">          <tbody>            <tr bgColor="#ffffff"><td vAlign="top" height="32" colspan="4"><font class="itsmultlinefont"><pre>Purpose:
To check whether the correct RefSeq mRNA ID has been mapped to the corresponding Refseq Protein ID in the parsed data by parsing a test input data file provided with the deliverable.

Workflow details:
1.The GeneConnect Server downloads the data from the website, the Parsers parse the datafiles to create link files and  the uploader loads the data from the link files in to the respective base tables in the GeneConnect database. This test case is for testing  of the parser :Entrez_Refseq.pl .
2.This Parser is involved in the GeneConnect Second Run for Annotation Parsers.
3.This parser provides the link files for the following links :
[Figure in the round brackets indicate Link number from the graph for Annotation parsers ]
Graph provided at: http://cabigcvs.nci.nih.gov/viewcvs/viewcvs.cgi/geneconnect/Coding%20and%20Testing/Test%20Plans/workflow.doc?search=None&hideattic=0

REFSEQMRNA_REFSEQPROTEIN (14)

Setup:
1. Verify UpdateLinux.sh script file to check if it exists in the main folder from which GeneConnect server is run. Also verify the following in the UpdateLinux.sh:
a. The path of test config file is correctly set.
b. The JVM parameter -DentityExpansionLimit = 5000000 has been specified in the java command.
c. The JVM parameter -Xmx512M has been specified in the java command. 
d. Check if JAVA_HOME, ORACLE_HOME, PATH AND CLASS_PATH are set properly.
e. Set the parameters of the java command :
java  -Xmx512M  -DentityExpansionLimit=5000000 -Djava.security.manager  -Djava.security.policy=java.policy com.dataminer.server.jobmanager.JobManager

2. Verify the path of the input data file (entrez_refseqTestFile.dat, which resides in the main folder from which GeneConnect server is run) in the config file CommandFile.xml (in the 'Config' folder).

3. Check the database to ensure that all the base tables required for entrez_refseq data population are present. The names of these tables can be found in the design document.

4. Check the server.properties file (in the 'Config' folder) and verify the database driver name and the dbURL and other database arguments including the TNS service name for the Oracle database, which can be any one of the following:
a) oracle.jdbc.driver.OracleDriver : for Driver name
b) jdbc:oracle:oci8:@&lt;database name&gt; : for database URL

5. Check the external entrez_refseq's Perl Parser set up as follows:
a) Ensure that the external parser's .sh file (entrez_refseqparser.sh) is present in the 'ParserScripts' folder.

6. Check whether Perl is installed on the machine and are present in the path of the directory from where testing is being done. Specify this path in the external perl parser code, whenever the Perl path (e.g. /usr/bin/perl) or library path (e.g. Bioperl path) is specified.

7. Add the following entry in Config/CommandFile.xml. This denotes that
a) The test file is present in '/GeneConnectServer' directory of local machine. 
b) entrez_refseqparser.sh will call the external perl parser

&lt;DataSource name="REFSEQMRNA_REFSEQPROTEIN" type="" siteurl="opt1" 
	 	username="anonymous" password=" "&gt;
	 	&lt;ExternalParser writeToDB="false" commandfile="ParserScripts/entrez_refseqparser.sh"/&gt;
		&lt;BaseDirectory dir="/GeneConnectServer" islocal="true"&gt;
	 		&lt;File name="entrez_refseqTestFile.dat"/&gt;
	 	&lt;/BaseDirectory&gt;		
&lt;/DataSource&gt; 

8. Run the script UpdateLinux.sh. 

Input/Data:
The test input dat file (entrez_refseqTestFile.dat) provided with the deliverable. 

Expected Output:
1. The output file refseqmrna_refseqprotein.txt should be formed.
2. The correct data should be loaded into the tables.

Verification Logic:
1. Check the correctness of the data loaded into tables as follows:

a) Match all the mappings in the entrez_refseqTestFile.dat file with  refseqmrna_refseqprotein base table (i.e.) REFSEQMRNA_REFSEQPROTEIN. The SQL query is as follows: "SELECT * from REFSEQMRNA_REFSEQPROTEIN;" 

b) Match the data in the NCBI website with that in the output files and in the database tables. Go to http://www.ncbi.nlm.nih.gov/RefSeq/, enter a Refseq mrna ID, search on 'Nucleotide' from the drop-down, Click on the 'RefSeq mRNA'.Look for a primary tag called 'CDS',under this tag click the 'protein_id'. The 'Locus' here is the RefSeq Protein ID. The RefseqmRNAID-RefseqproteinID mapping should be the same in both the website and table.



</pre></font></td>            </tr>            <tr bgColor="#ffffff">              <td vAlign="top" height="21"><b>Mode of Execution</b></td>              <td vAlign="top" height="1">Manual                </td>              <td vAlign="top" height="21"><b>Verification Logic</b></td>              <td vAlign="top" height="1">DATABASE<br>                </td>            </tr>            <tr bgColor="#ffffff">              <td vAlign="top" height="21"><b>Author</b></td>              <td vAlign="top" height="21">Madhurima</td>              <td vAlign="top" height="21"><b>Date Entered</b></td>              <td vAlign="top" height="21">21/07/2006&nbsp;</td>            </tr>            <tr bgColor="#ffffff">              <td vAlign="top" height="21"></td>              <td vAlign="top" height="21"></td>              <td vAlign="top" height="21"><b>Date Last Update</b></td>              <td vAlign="top" height="21">08/09/2006&nbsp;</td>            </tr>          </tbody>        </table>      </td>    </tr>  </tbody></table>
<p align="right"><a href="#TMT Generated Test Plans">Top</a></p><a name="Test Area Name: GeneConnectver 1.1ParsersUnigene Parser"><center><table bgColor="#CCCCFF" border="1" cellPadding="0" cellSpacing="0" width="700">  <tbody>    <tr>      <td align="middle" bgColor="#CCCCFF">        <p align="left"><font color="#660066"><b>Test Area Name: Unigene Parser</b></font></p> </td> </tr> <tr> <td>&nbsp;</td> </tr> </tbody></table><br><table bgColor="#CCCCFF" border="1" cellPadding="0" cellSpacing="0" width="700"> <tbody> <tr> <td align="middle" bgColor="#CCCCFF"> <p align="left"><font color="#660066"><b>VITALS</b></font></p> </td> </tr> <tr> <td> <table bgColor="#CCCCFF" border="1" cellPadding="2" width="696"> <tbody> <tr bgColor="#ffffff"> <td vAlign="top" width="131"><b>Reference</b></td> <td vAlign="top" width="529">&nbsp;<a href="" target="_blank"></a></td> </tr> </tbody> </table> <tr> <td align="middle" bgColor="#CCCCFF"> <p align="left"><font color="#660066"><b>DESCRIPTION</b></font></p> </td> </tr> <tr> <td> <table bgColor="#CCCCFF" border="1" cellPadding="2" width="100%" height="55"> <tbody> <tr bgColor="#ffffff"> <td vAlign="top" height="1"></td> </tr> </tbody> </table> </td> </tr> </tbody></table>
<p align="right"><a href="#TMT Generated Test Plans">Top</a></p><a name="Test Case ID: GeneConnectver 1.1ParsersUnigene ParserUnigene Parser"><center><table bgColor="#CCCCFF" border="1" cellPadding="0" cellSpacing="0" width="700">  <tbody>    <tr>      <td align="middle" bgColor="#CCCCFF">        <p align="left"><font color="#660066"><b>Test Case ID : 1660</b></font></p>      </td>    </tr>    <tr>      <td bgcolor="#FFFFFF"><b>Short Title : </b>Unigene Parser</td>    </tr>    <tr>      <td bgcolor="#FFFFFF"><b>Title : </b>Unigene Parser</td>    </tr>  </tbody></table><br><table bgColor="#CCCCFF" border="1" cellPadding="0" cellSpacing="0" width="700">  <tbody>    <tr>      <td align="middle" bgColor="#CCCCFF">        <p align="left"><font color="#660066"><b>VITALS</b></font></p>      </td>    </tr>    <tr>      <td>        <table bgColor="#CCCCFF" border="1" cellPadding="2" width="100%">          <tbody>            <tr bgColor="#ffffff">              <td vAlign="top"><b>Product</b></td>              <td vAlign="top" colspan="3">GeneConnect</td>            <tr bgColor="#ffffff">              <td vAlign="top"><b>Component</b></td>              <td vAlign="top">Parsers<br>              </td>              <td vAlign="top"><b>Test Area</b></td>              <td vAlign="top">Unigene Parser</td>            </tr>            <tr bgColor="#ffffff">              <td vAlign="top"><b>Reference</b></td>              <td vAlign="top">1</td>              <td vAlign="top"><b>Priority</b></td>              <td vAlign="top">P2</td>            </tr>          </tbody>        </table>    <tr>      <td align="middle" bgColor="#CCCCFF">        <p align="left"><font color="#660066"><b>DESCRIPTION</b></font></p>      </td>    </tr>    <tr>      <td>        <table bgColor="#CCCCFF" border="1" cellPadding="2" width="100%" height="113">          <tbody>            <tr bgColor="#ffffff"><td vAlign="top" height="32" colspan="4"><font class="itsmultlinefont"><pre>Purpose:
To ensure the correctness of the parsed data by parsing a test input data file  provided with the deliverable.

Workflow details:
1.The GeneConnect Server downloads the data from the website, the Parsers parse the datafiles to create link files and  the uploader loads the data from the link files in to the respective base tables in the GeneConnect database. This test case is for testing  of the Unigene parser .
2.This Parser is involved in the GeneConnect Second Run for Annotation Parsers.

Setup:
1. Verify UpdateLinux.sh script file to check if to check if it exists in the main folder from which GeneConnect server is run. Also verify the following in the UpdateLinux.sh:
a. The path of test config file is correctly set.
b. The JVM parameter -DentityExpansionLimit = 5000000 has been specified in the java command.
c. The JVM parameter -Xmx512M has been specified in the java command. 
d. Check if JAVA_HOME, ORACLE_HOME, PATH AND CLASS_PATH are set properly.
e. Set the parameters of the java command in unigeneTestRun.sh script:
java  -Xmx512M  -DentityExpansionLimit=5000000 -Djava.security.manager  -Djava.security.policy=java.policy com.dataminer.server.jobmanager.JobManager

2. Verify the path of the input data file (unigeneTestFile.dat.gz, which resides in the main folder from which GeneConnect server is run) in the config file (CommandFile.xml). 

3. Check the database to ensure that all the base tables required for Unigene data population are present. The names of these tables can be
found in the design document.

4. Check the server.properties file (in the 'Config' folder) and verify the database driver name and the dbURL and other database arguments including the TNS service name for the Oracle database, which can be any one of the following:
a) oracle.jdbc.driver.OracleDriver : for Driver name
b) jdbc:oracle:oci8:@&lt;database name&gt; : for database URL

5. Check the external unigene Java Parser set up as follows:
a) Ensure that the external parser's .sh file (unigeneparser.sh) is present in the 'ParserScripts' folder.
b) Ensure that the java parameters are as follows: java  -Xmx512M  -DentityExpansionLimit=5000000 -Djava.security.manager  -Djava.security.policy=java.policy edu.wustl.geneconnect.parser.UniGeneParser %1 %2 %3 %4 %5 %6
c) Ensure that the classpath, JAVA_HOME, ORACLE_HOME are set properly in the UnigeneParser.sh file.

6. Add the following entry in Config/CommandFile.xml. This denotes that 
a) The test file is present in '/GeneConnectServer' directory of local machine. 
b) UnigeneParser.sh calls the external Java parser

&lt;DataSource name="UNIGENE" type="" siteurl="opt1" 
	 	username="anonymous" password=" "&gt;
	 	&lt;ExternalParser writeToDB="false" commandfile="ParserScripts/UnigeneParser.sh"/&gt;
		&lt;BaseDirectory dir="/GeneConnectServer" islocal="true"&gt;
	 		&lt;File name="unigeneTestFile.dat.gz"/&gt;
	 	&lt;/BaseDirectory&gt;		
&lt;/DataSource&gt;

6. Verify that Perl is installed and the path of Perl is specified in the UpdateLinux.sh file.

7. Run the script UpdateLinux.sh. 

Input/Data:
The test input dat file provided with the deliverable. 

Expected Output:
1. Correct data should loaded in the base table.

Verification Logic:
1. Check the correctness of the data loaded into tables as follows:

a) Match all the mappings in the unigeneTestFile.dat.gz file with UniGene base table (i.e.) UNIGENE. The SQL query is as follows: SELECT * from UNIGENE ;(GENEID column in this table is the list of all Unigene ID's).

b) Match the data in the NCBI website with the database tables. Go to  http://www.ncbi.nlm.nih.gov/entrez/query.fcgi?db=unigene , enter a Unigene ID, click the 'Links' at the top right, then click 'Gene'. This will provide a Entrez gene name in 'Summary' section with tag 'Gene aliases:'.The UniGeneID-EntrezGene name mapping should be the same in both the website and database table.
Also if the Entrez Gene name is absent in the database table for a particular Unigene ID, then the 'Links' to 'Gene' should be absent for that Unigene ID at the Unigene website .



</pre></font></td>            </tr>            <tr bgColor="#ffffff">              <td vAlign="top" height="21"><b>Mode of Execution</b></td>              <td vAlign="top" height="1">Manual                </td>              <td vAlign="top" height="21"><b>Verification Logic</b></td>              <td vAlign="top" height="1">UI<br>                </td>            </tr>            <tr bgColor="#ffffff">              <td vAlign="top" height="21"><b>Author</b></td>              <td vAlign="top" height="21">Arun Venkatesan</td>              <td vAlign="top" height="21"><b>Date Entered</b></td>              <td vAlign="top" height="21">17/08/2006&nbsp;</td>            </tr>            <tr bgColor="#ffffff">              <td vAlign="top" height="21"></td>              <td vAlign="top" height="21"></td>              <td vAlign="top" height="21"><b>Date Last Update</b></td>              <td vAlign="top" height="21">08/09/2006&nbsp;</td>            </tr>          </tbody>        </table>      </td>    </tr>  </tbody></table>
<p align="right"><a href="#TMT Generated Test Plans">Top</a></p><a name="Test Area Name: GeneConnectver 1.1ParsersUniprot External Parser"><center><table bgColor="#CCCCFF" border="1" cellPadding="0" cellSpacing="0" width="700">  <tbody>    <tr>      <td align="middle" bgColor="#CCCCFF">        <p align="left"><font color="#660066"><b>Test Area Name: Uniprot External Parser</b></font></p> </td> </tr> <tr> <td>&nbsp;</td> </tr> </tbody></table><br><table bgColor="#CCCCFF" border="1" cellPadding="0" cellSpacing="0" width="700"> <tbody> <tr> <td align="middle" bgColor="#CCCCFF"> <p align="left"><font color="#660066"><b>VITALS</b></font></p> </td> </tr> <tr> <td> <table bgColor="#CCCCFF" border="1" cellPadding="2" width="696"> <tbody> <tr bgColor="#ffffff"> <td vAlign="top" width="131"><b>Reference</b></td> <td vAlign="top" width="529">&nbsp;<a href="" target="_blank"></a></td> </tr> </tbody> </table> <tr> <td align="middle" bgColor="#CCCCFF"> <p align="left"><font color="#660066"><b>DESCRIPTION</b></font></p> </td> </tr> <tr> <td> <table bgColor="#CCCCFF" border="1" cellPadding="2" width="100%" height="55"> <tbody> <tr bgColor="#ffffff"> <td vAlign="top" height="1"></td> </tr> </tbody> </table> </td> </tr> </tbody></table>
<p align="right"><a href="#TMT Generated Test Plans">Top</a></p><a name="Test Case ID: GeneConnectver 1.1ParsersUniprot External ParserUniprot to Refseq"><center><table bgColor="#CCCCFF" border="1" cellPadding="0" cellSpacing="0" width="700">  <tbody>    <tr>      <td align="middle" bgColor="#CCCCFF">        <p align="left"><font color="#660066"><b>Test Case ID : 1648</b></font></p>      </td>    </tr>    <tr>      <td bgcolor="#FFFFFF"><b>Short Title : </b>Uniprot to Refseq</td>    </tr>    <tr>      <td bgcolor="#FFFFFF"><b>Title : </b>Check whether the correct Uniprot accession is mapped to RefSeq protein ID</td>    </tr>  </tbody></table><br><table bgColor="#CCCCFF" border="1" cellPadding="0" cellSpacing="0" width="700">  <tbody>    <tr>      <td align="middle" bgColor="#CCCCFF">        <p align="left"><font color="#660066"><b>VITALS</b></font></p>      </td>    </tr>    <tr>      <td>        <table bgColor="#CCCCFF" border="1" cellPadding="2" width="100%">          <tbody>            <tr bgColor="#ffffff">              <td vAlign="top"><b>Product</b></td>              <td vAlign="top" colspan="3">GeneConnect</td>            <tr bgColor="#ffffff">              <td vAlign="top"><b>Component</b></td>              <td vAlign="top">Parsers<br>              </td>              <td vAlign="top"><b>Test Area</b></td>              <td vAlign="top">Uniprot External Parser</td>            </tr>            <tr bgColor="#ffffff">              <td vAlign="top"><b>Reference</b></td>              <td vAlign="top">1</td>              <td vAlign="top"><b>Priority</b></td>              <td vAlign="top">P3</td>            </tr>          </tbody>        </table>    <tr>      <td align="middle" bgColor="#CCCCFF">        <p align="left"><font color="#660066"><b>DESCRIPTION</b></font></p>      </td>    </tr>    <tr>      <td>        <table bgColor="#CCCCFF" border="1" cellPadding="2" width="100%" height="113">          <tbody>            <tr bgColor="#ffffff"><td vAlign="top" height="32" colspan="4"><font class="itsmultlinefont"><pre>Purpose:
To ensure the correctness of the parsed data by parsing a test input data file (uniprot_refseqTestFile.dat) provided with the deliverable. This test case is written for a Linux machine.

Workflow details:
1.The GeneConnect Server downloads the data from the website, the Parsers parse the datafiles to create link files and  the uploader loads the data from the link files in to the respective base tables in the GeneConnect database. This test case is for testing  of the parser : Uniprot_RefSeq.pl 
2.This Parser is involved in the GeneConnect Second Run for Annotation Parsers.
3.This parser provides the link files for the following links :
[Figure in the round brackets indicate Link number from the graph for Annotation parsers ]
Graph provided at: http://cabigcvs.nci.nih.gov/viewcvs/viewcvs.cgi/geneconnect/Coding%20and%20Testing/Test%20Plans/workflow.doc?search=None&hideattic=0

UNIPROT_REFSEQPROTEIN(23),
REFSEQPROTEIN_UNIPROT(25)
 

Setup:
1. Verify UpdateLinux.sh script file to check if it exists in the main folder from which GeneConnect server is run. Also verify the following in the UpdateLinux.sh:
a. The path of test config file is correctly set.
b. The JVM parameter -DentityExpansionLimit = 5000000 has been specified in the java command.
c. The JVM parameter -Xmx512M has been specified in the java command. 
d. Check if JAVA_HOME, ORACLE_HOME, PATH AND CLASS_PATH are set properly.
e. Set the parameters of the java command in uniprot_refseqTestRun.sh script:
java  -Xmx512M  -DentityExpansionLimit=5000000 -Djava.security.manager  -Djava.security.policy=java.policy com.dataminer.server.jobmanager.JobManager

2. Verify the path of the input data file (uniprot_refseqTestFile.dat, which resides in the main folder from which GeneConnect server is run) in the config file CommandFile.xml (in the 'Config' folder).

3. Check the database to ensure that all the base tables required for uniprot_refseq data population are present. The names of these tables can be found in the design document.

4. Check the server.properties file (in the 'Config' folder) and verify the database driver name and the dbURL and other database arguments including the TNS service name for the Oracle database, which can be any one of the following:
a) oracle.jdbc.driver.OracleDriver : for Driver name
b) jdbc:oracle:oci8:@&lt;database name&gt; : for database URL

5. Check the external uniprot_refseq's Perl Parser set up as follows:
a) Ensure that the external parser's .sh file (uniprot_refseq.sh) is present in the 'ParserScripts' folder.

6. Check whether Perl is installed on the machine and are present in the path of the directory from where testing is being done. Specify this path in the external perl parser code, whenever the Perl path (e.g. /usr/bin/perl) or library path (e.g. Bioperl path) is specified.

7. Add the following entry in Config/CommandFile.xml. This denotes that 
a) the test file is present in '/GeneConnectServer' directory of local machine. 
b) uniprot_refseq.sh will call the external perl parser

&lt;DataSource name="UNIPROT_REFSEQ" type="" siteurl="opt1" 
	 	username="anonymous" password=" "&gt;
	 	&lt;ExternalParser writeToDB="false" commandfile="ParserScripts/uniprot_refseq.sh"/&gt;
		&lt;BaseDirectory dir="/GeneConnectServer" islocal="true"&gt;
	 		&lt;File name="uniprot_refseqTestFile.dat"/&gt;
	 	&lt;/BaseDirectory&gt;		
&lt;/DataSource&gt;

8. Run the script UpdateLinux.sh. 

Input/Data:
The test input dat file (uniprot_refseqTestFile.dat) provided with the deliverable.

Expected Output:
1. The output file uniprot_refseq.txt should be formed.
2. The correct data should be loaded into the tables.

Verification Logic:
1. Check the correctness of the data loaded into tables as follows:

a) Match all the mappings in the uniprot_refseqTestFile.dat file with that in the output file and with base tables (i.e.) UNIPROT_REFSEQPROTEIN and REFSEQPROTEIN_UNIPROT.The SQL query is as follows: " SELECT * from UNIPROT_REFSEQPROTEIN;" and " SELECT * from REFSEQPROTEIN_UNIPROT;" respectively.









</pre></font></td>            </tr>            <tr bgColor="#ffffff">              <td vAlign="top" height="21"><b>Mode of Execution</b></td>              <td vAlign="top" height="1">Manual                </td>              <td vAlign="top" height="21"><b>Verification Logic</b></td>              <td vAlign="top" height="1">UI<br>                </td>            </tr>            <tr bgColor="#ffffff">              <td vAlign="top" height="21"><b>Author</b></td>              <td vAlign="top" height="21">Shweta Chavan</td>              <td vAlign="top" height="21"><b>Date Entered</b></td>              <td vAlign="top" height="21">28/07/2006&nbsp;</td>            </tr>            <tr bgColor="#ffffff">              <td vAlign="top" height="21"></td>              <td vAlign="top" height="21"></td>              <td vAlign="top" height="21"><b>Date Last Update</b></td>              <td vAlign="top" height="21">08/09/2006&nbsp;</td>            </tr>          </tbody>        </table>      </td>    </tr>  </tbody></table>
<p align="right"><a href="#TMT Generated Test Plans">Top</a></p><a name="Test Case ID: GeneConnectver 1.1ParsersUniprot External ParserUniprot to Genbank"><center><table bgColor="#CCCCFF" border="1" cellPadding="0" cellSpacing="0" width="700">  <tbody>    <tr>      <td align="middle" bgColor="#CCCCFF">        <p align="left"><font color="#660066"><b>Test Case ID : 1649</b></font></p>      </td>    </tr>    <tr>      <td bgcolor="#FFFFFF"><b>Short Title : </b>Uniprot to Genbank</td>    </tr>    <tr>      <td bgcolor="#FFFFFF"><b>Title : </b>Check whether the correct Uniprot accession is mapped to Genbank protein ID</td>    </tr>  </tbody></table><br><table bgColor="#CCCCFF" border="1" cellPadding="0" cellSpacing="0" width="700">  <tbody>    <tr>      <td align="middle" bgColor="#CCCCFF">        <p align="left"><font color="#660066"><b>VITALS</b></font></p>      </td>    </tr>    <tr>      <td>        <table bgColor="#CCCCFF" border="1" cellPadding="2" width="100%">          <tbody>            <tr bgColor="#ffffff">              <td vAlign="top"><b>Product</b></td>              <td vAlign="top" colspan="3">GeneConnect</td>            <tr bgColor="#ffffff">              <td vAlign="top"><b>Component</b></td>              <td vAlign="top">Parsers<br>              </td>              <td vAlign="top"><b>Test Area</b></td>              <td vAlign="top">Uniprot External Parser</td>            </tr>            <tr bgColor="#ffffff">              <td vAlign="top"><b>Reference</b></td>              <td vAlign="top">1</td>              <td vAlign="top"><b>Priority</b></td>              <td vAlign="top">P3</td>            </tr>          </tbody>        </table>    <tr>      <td align="middle" bgColor="#CCCCFF">        <p align="left"><font color="#660066"><b>DESCRIPTION</b></font></p>      </td>    </tr>    <tr>      <td>        <table bgColor="#CCCCFF" border="1" cellPadding="2" width="100%" height="113">          <tbody>            <tr bgColor="#ffffff"><td vAlign="top" height="32" colspan="4"><font class="itsmultlinefont"><pre>Purpose:
To ensure the correctness of the parsed data by parsing a test input data file (uniprot_genbankTestFile.dat) provided with the deliverable. This test case is written for a Linux machine.

Workflow details:
1.The GeneConnect Server downloads the data from the website, the Parsers parse the datafiles to create link files and  the uploader loads the data from the link files in to the respective base tables in the GeneConnect database. This test case is for testing  of the parser : Uniprot_GenBank.pl 
2.This Parser is involved in the GeneConnect Second Run for Annotation Parsers.
3.This parser provides the link files for the following links :
[Figure in the round brackets indicate Link number from the graph for Annotation parsers ]
Graph provided at: http://cabigcvs.nci.nih.gov/viewcvs/viewcvs.cgi/geneconnect/Coding%20and%20Testing/Test%20Plans/workflow.doc?search=None&hideattic=0

UNIPROT_GENBANKPROTEIN (28), 
GENBANKPROT_UNIPROT (26).

Setup:
1. Verify UpdateLinux.sh script file to check if it exists in the main folder from which GeneConnect server is run. Also verify the following in the UpdateLinux.sh:
a. The path of test config file is correctly set.
b. The JVM parameter -DentityExpansionLimit = 5000000 has been specified in the java command.
c. The JVM parameter -Xmx512M has been specified in the java command. 
d. Check if JAVA_HOME, ORACLE_HOME, PATH AND CLASS_PATH are set properly.
e. Set the parameters of the java command in uniprot_genbankTestRun.sh script:
java  -Xmx512M  -DentityExpansionLimit=5000000 -Djava.security.manager  -Djava.security.policy=java.policy com.dataminer.server.jobmanager.JobManager

2. Verify the path of the input data file (uniprot_genbankTestFile.dat, which resides in the main folder from which GeneConnect server is run) in the config file CommandFile.xml(in the 'Config' folder).

3. Check the database to ensure that all the base tables required for uniprot_genbank data population are present. The names of these tables can be found in the design document.

4. Check the server.properties file (in the 'Config' folder) and verify the database driver name and the dbURL and other database arguments including the TNS service name for the Oracle database, which can be any one of the following:
a) oracle.jdbc.driver.OracleDriver : for Driver name
b) jdbc:oracle:oci8:@&lt;database name&gt; : for database URL

5. Check the external uniprot_genbank's Perl Parser set up as follows:
a) Ensure that the external parser's .sh file (uniprot_genbank.sh) is present in the 'ParserScripts' folder.

6. Check whether Perl is installed on the machine and are present in the path of the directory from where testing is being done. Specify this path in the external perl parser code, whenever the Perl path (e.g. /usr/bin/perl) or library path (e.g. Bioperl path) is specified.

7. Add the following entry in Config/CommandFile.xml. This denotes that 
a) the test file is present in '/GeneConnectServer' directory of local machine. 
b) uniprot_genbank.sh will call the external perl parser

&lt;DataSource name="UNIPROT_GENBANK" type="" siteurl="opt1" 
	 	username="anonymous" password=" "&gt;
	 	&lt;ExternalParser writeToDB="false" commandfile="ParserScripts/uniprot_genbank.sh"/&gt;
		&lt;BaseDirectory dir="/GeneConnectServer" islocal="true"&gt;
	 		&lt;File name="uniprot_genbankTestFile.dat"/&gt;
	 	&lt;/BaseDirectory&gt;		
&lt;/DataSource&gt;

8. Run the script UpdateLinux.sh. 

Input/Data:
The test input dat file provided with the deliverable.

Expected Output:
1.The output file uniprot_refseq.txt should be formed.
2.The correct data should be loaded into the tables.

Verification Logic:
1. Check the correctness of the data loaded into tables as follows:

a) Match all the mappings in the uniprot_genbankTestFile.dat file with uniprot_genbank base tables (i.e.) UNIPROT_GENBANKPROTEIN and GENBANKPROTEIN_UNIPROT. The SQL query is as follows : "SELECT * from UNIPROT_GENBANKPROTEIN ;" and "SELECT * from GENBANKPROTEIN_UNIPROT ;" respectively.

b)Match the data in the Uniprot website with that in the output file and the database tables. Go to http://www.ebi.uniprot.org/index.shtml , enter a Uniprot accession number in the 'Text Search UniProt Knowledgebase' area, look for a tag 'Cross-references', click on Genbank . Look for a primary tag called 'CDS',under this tag the 'protein_id' is the Genbank Protein ID.  The Uniprot accession-Genbank proteinID mapping should be the same in both the website and table.



</pre></font></td>            </tr>            <tr bgColor="#ffffff">              <td vAlign="top" height="21"><b>Mode of Execution</b></td>              <td vAlign="top" height="1">Manual                </td>              <td vAlign="top" height="21"><b>Verification Logic</b></td>              <td vAlign="top" height="1">UI<br>                </td>            </tr>            <tr bgColor="#ffffff">              <td vAlign="top" height="21"><b>Author</b></td>              <td vAlign="top" height="21">Shweta Chavan</td>              <td vAlign="top" height="21"><b>Date Entered</b></td>              <td vAlign="top" height="21">28/07/2006&nbsp;</td>            </tr>            <tr bgColor="#ffffff">              <td vAlign="top" height="21"></td>              <td vAlign="top" height="21"></td>              <td vAlign="top" height="21"><b>Date Last Update</b></td>              <td vAlign="top" height="21">08/09/2006&nbsp;</td>            </tr>          </tbody>        </table>      </td>    </tr>  </tbody></table>
<p align="right"><a href="#TMT Generated Test Plans">Top</a></p></body></html>
