<head>
<META content="Microsoft FrontPage 4.0" name=GENERATOR>
<STYLE type=text/css>BODY {
	BACKGROUND-COLOR: #ffffff; COLOR: #000000; FONT-FAMILY: "Times New Roman"; FONT-SIZE: 12pt; TEXT-ALIGN: left
}
 UL.zzul {
	DISPLAY: block; LIST-STYLE: none}
SPAN.zzspace {
	LEFT: 11px; POSITION: relative
}
 A {
	COLOR: #0000ff
}
 A:visited {
	COLOR: #0000ff
}
 A:hover {
	COLOR: #ff0000
}
TD {
	COLOR: #000000; FONT-FAMILY: "Times New Roman"; FONT-SIZE: 12pt; TEXT-ALIGN: left
}
 TABLE {
	BORDER-BOTTOM: medium none; BORDER-LEFT: medium none; BORDER-RIGHT: medium none; BORDER-TOP: medium none; PADDING-BOTTOM: 3px; PADDING-LEFT: 3px; PADDING-RIGHT: 3px; PADDING-TOP: 3px; WIDTH: 100%
 }
 </STYLE>
 <SCRIPT type=text/javascript>
 <!--
 function exp_coll(ind)
 {
 s = document.getElementById("sp_" + ind);
 i = document.getElementById("im_" + ind);
 i2 = document.getElementById("im2_" + ind);
 if (s.style.display == 'none')
 {
  s.style.display = 'block';   
 }
 else if (s.style.display == 'block')
 {
   s.style.display = 'none';   
 }
}
 function coll_all(i)
 {  
 for(ind=1;ind<=i;ind++){
  x = document.getElementById("sp_" + ind);
  x.style.display = 'none';
  if (x.style.display == 'none')
  {    
 }
}
}
 function exp_all(i)
 {
  for(ind=1;ind<=i;ind++){
  x = document.getElementById("sp_" + ind);
  x.style.display = 'block';
  if (x.style.display == 'block')
  {    
 }
  }  
}
 -->
</SCRIPT>
</HEAD>
<BODY>
<center><a name="TMT Generated Test Plans"><h1>Test Plan</h1></a></center><hr><a name="xys"><table><tr><td>This is a TMT (Test Management Tool) generated Test Plan. To read a short note on TMT on the caBIG Best Practices SIG website, click <a href="http://cabig-bpsigwiki.bioinformatics.northwestern.edu/bpsigwiki/index.php/Manual_Testing_-_Test_Management_Tool">here</a>.<br> For more details or if you find any issue with this report, please send an email to <a href="mailto:tmt@persistent.co.in">tmt@persistent.co.in</a>.</td></tr></table></a><hr><br><br>
<A href="javascript:exp_all(4);">Expand All</A> &nbsp; <A href="javascript:coll_all(4);">Collapse All</A> 

<UL class=zzul>
  <LI>
<LI><A href="javascript:exp_coll(1);">+</A> <A 
	  href="javascript:exp_coll(1);"></A> &nbsp; <A  href="#Product Name: GeneConnect">GeneConnect</A> 
<UL class=zzul id=sp_1 style="DISPLAY: block">
<SCRIPT type=text/javascript>
 ind = 1; 
 x = document.getElementById("sp_" + ind);
 x.style.display = 'none'; 
 if (x.style.display == 'none'){ }
</SCRIPT>
<LI><A href="javascript:exp_coll(2);">+</A> <A 
	  href="javascript:exp_coll(2);"></A> &nbsp; <A  href="#Version Name: GeneConnectver 1.1">ver 1.1</A> 
<UL class=zzul id=sp_2 style="DISPLAY: block">
<SCRIPT type=text/javascript>
 ind = 2; 
 x = document.getElementById("sp_" + ind);
 x.style.display = 'none'; 
 if (x.style.display == 'none'){ }
</SCRIPT>
<LI><A href="javascript:exp_coll(3);">+</A> <A 
	  href="javascript:exp_coll(3);"></A> &nbsp; <A  href="#Component Name: GeneConnectver 1.1caCore API">caCore API</A> 
<UL class=zzul id=sp_3 style="DISPLAY: block">
<SCRIPT type=text/javascript>
 ind = 3; 
 x = document.getElementById("sp_" + ind);
 x.style.display = 'none'; 
 if (x.style.display == 'none'){ }
</SCRIPT>
<LI><A href="javascript:exp_coll(4);">+</A> <A 
	  href="javascript:exp_coll(4);"></A> &nbsp; <A  href="#Test Area Name: GeneConnectver 1.1caCore APISearch">Search</A> 
<UL class=zzul id=sp_4 style="DISPLAY: block">
<SCRIPT type=text/javascript>
 ind = 4; 
 x = document.getElementById("sp_" + ind);
 x.style.display = 'none'; 
 if (x.style.display == 'none'){ }
</SCRIPT>
<LI><SPAN class=zzspace>&nbsp;&nbsp;
<A href="#Test Case ID: GeneConnectver 1.1caCore APISearchBasic Genomic ID Search">Basic Genomic ID Search</A> </SPAN>
<LI><SPAN class=zzspace>&nbsp;&nbsp;
<A href="#Test Case ID: GeneConnectver 1.1caCore APISearchQuery Based on Confidence Score">Query Based on Confidence Score</A> </SPAN>
<LI><SPAN class=zzspace>&nbsp;&nbsp;
<A href="#Test Case ID: GeneConnectver 1.1caCore APISearchQuery By ONT">Query By ONT</A> </SPAN>
<LI><SPAN class=zzspace>&nbsp;&nbsp;
<A href="#Test Case ID: GeneConnectver 1.1caCore APISearchQuery By Limiting ID Frequency">Query By Limiting ID Frequency</A> </SPAN>
</LI></UL>

</LI></UL>

</LI></UL>
<br><br><a><table><tr><td width="50%"><table bgColor="#CCCCFF" border="1" cellPadding="2" width="100"><tr><td colspan="10"><font color="#660066"><b>Test case count: Component verses Priority</b></font></td></tr><tr><td  align="center" bgcolor="#ccccff"><font color="#660066"><b>Component Names</b></font></td><td align="center" bgcolor="#ccccff"> <font color="#660066"><b>P1</b></font></td><td align="center" bgcolor="#ccccff"> <font color="#660066"><b>P2</b></font></td><td align="center" bgcolor="#ccccff"> <font color="#660066"><b>P3</b></font></td><td align="center" bgcolor="#ccccff"> <font color="#660066"><b>P4</b></font></td><td align="center" bgcolor="#ccccff"> <font color="#660066"><b>Total</b></td></tr><tr><td bgcolor="#ffffff">ver 1.1</td><td  bgcolor="#ffffff">14</td><td  bgcolor="#ffffff">1</td><td  bgcolor="#ffffff">0</td><td  bgcolor="#ffffff">0</td><td bgcolor="#ffffff">15</td></tr><tr><td bgcolor="#ffffff">caCore API</td><td  bgcolor="#ffffff">4</td><td  bgcolor="#ffffff">0</td><td  bgcolor="#ffffff">0</td><td  bgcolor="#ffffff">0</td><td bgcolor="#ffffff">4</td></tr><tr><td bgcolor="#ccccff"><b>Column Total</b></td><td  bgcolor="#ccccff"><b>18<b></td><td  bgcolor="#ccccff"><b>1<b></td><td  bgcolor="#ccccff"><b>0<b></td><td  bgcolor="#ccccff"><b>0<b></td><td  bgcolor="#ccccff"><b>19</b></td></table></a></td><td width="50%">&nbsp;</td></tr></table><br><br><a name="Product Name: GeneConnect"><center><table bgColor="#CCCCFF" border="1" cellPadding="0" cellSpacing="0" width="700">  <tbody>    <tr>      <td align="middle" bgColor="#CCCCFF">        <p align="left"><font color="#660066"><b>Product Name: GeneConnect</b></font></p> </td> </tr> <tr> <td>&nbsp;</td> </tr> </tbody></table><br><table bgColor="#CCCCFF" border="1" cellPadding="0" cellSpacing="0" width="700"> <tbody> <tr> <td align="middle" bgColor="#CCCCFF"> <p align="left"><font color="#660066"><b>VITALS</b></font></p> </td> </tr> <tr> <td> <table bgColor="#CCCCFF" border="1" cellPadding="2" width="696"> <tbody> <tr bgColor="#ffffff"> <td vAlign="top" width="131"><b>Reference</b></td> <td vAlign="top" width="529">&nbsp;<a href="" target="_blank"></a></td> </tr> </tbody> </table> <tr> <td align="middle" bgColor="#CCCCFF"> <p align="left"><font color="#660066"><b>DESCRIPTION</b></font></p> </td> </tr> <tr> <td> <table bgColor="#CCCCFF" border="1" cellPadding="2" width="100%" height="55"> <tbody> <tr bgColor="#ffffff"> <td vAlign="top" height="1"></td> </tr> </tbody> </table> </td> </tr> </tbody></table>
<p align="right"><a href="#TMT Generated Test Plans">Top</a></p><a name="Version Name: GeneConnectver 1.1"><center><table bgColor="#CCCCFF" border="1" cellPadding="0" cellSpacing="0" width="700">  <tbody>    <tr>      <td align="middle" bgColor="#CCCCFF">        <p align="left"><font color="#660066"><b>Version Name: ver 1.1</b></font></p> </td> </tr> <tr> <td>&nbsp;</td> </tr> </tbody></table><br><table bgColor="#CCCCFF" border="1" cellPadding="0" cellSpacing="0" width="700"> <tbody> <tr> <td align="middle" bgColor="#CCCCFF"> <p align="left"><font color="#660066"><b>VITALS</b></font></p> </td> </tr> <tr> <td> <table bgColor="#CCCCFF" border="1" cellPadding="2" width="696"> <tbody> <tr bgColor="#ffffff"> <td vAlign="top" width="131"><b>Reference</b></td> <td vAlign="top" width="529">&nbsp;<a href="" target="_blank"></a></td> </tr> </tbody> </table> <tr> <td align="middle" bgColor="#CCCCFF"> <p align="left"><font color="#660066"><b>DESCRIPTION</b></font></p> </td> </tr> <tr> <td> <table bgColor="#CCCCFF" border="1" cellPadding="2" width="100%" height="55"> <tbody> <tr bgColor="#ffffff"> <td vAlign="top" height="1">First Version</td> </tr> </tbody> </table> </td> </tr> </tbody></table>
<p align="right"><a href="#TMT Generated Test Plans">Top</a></p><a name="Component Name: GeneConnectver 1.1caCore API"><center><table bgColor="#CCCCFF" border="1" cellPadding="0" cellSpacing="0" width="700">  <tbody>    <tr>      <td align="middle" bgColor="#CCCCFF">        <p align="left"><font color="#660066"><b>Component Name: caCore API</b></font></p> </td> </tr> <tr> <td>&nbsp;</td> </tr> </tbody></table><br><table bgColor="#CCCCFF" border="1" cellPadding="0" cellSpacing="0" width="700"> <tbody> <tr> <td align="middle" bgColor="#CCCCFF"> <p align="left"><font color="#660066"><b>VITALS</b></font></p> </td> </tr> <tr> <td> <table bgColor="#CCCCFF" border="1" cellPadding="2" width="696"> <tbody> <tr bgColor="#ffffff"> <td vAlign="top" width="131"><b>Reference</b></td> <td vAlign="top" width="529">&nbsp;<a href="" target="_blank"></a></td> </tr> </tbody> </table> <tr> <td align="middle" bgColor="#CCCCFF"> <p align="left"><font color="#660066"><b>DESCRIPTION</b></font></p> </td> </tr> <tr> <td> <table bgColor="#CCCCFF" border="1" cellPadding="2" width="100%" height="55"> <tbody> <tr bgColor="#ffffff"> <td vAlign="top" height="1">Gene Connect provides caCORE Java APIs to enable programmers to query the system.
The user will populate the DetachedCriteria object of hibernate specifying teh query to execute and call the caCORE like API?s ApplicationService.query() method, also specifying the return type object. </td> </tr> </tbody> </table> </td> </tr> </tbody></table>
<p align="right"><a href="#TMT Generated Test Plans">Top</a></p><a name="Test Area Name: GeneConnectver 1.1caCore APISearch"><center><table bgColor="#CCCCFF" border="1" cellPadding="0" cellSpacing="0" width="700">  <tbody>    <tr>      <td align="middle" bgColor="#CCCCFF">        <p align="left"><font color="#660066"><b>Test Area Name: Search</b></font></p> </td> </tr> <tr> <td>&nbsp;</td> </tr> </tbody></table><br><table bgColor="#CCCCFF" border="1" cellPadding="0" cellSpacing="0" width="700"> <tbody> <tr> <td align="middle" bgColor="#CCCCFF"> <p align="left"><font color="#660066"><b>VITALS</b></font></p> </td> </tr> <tr> <td> <table bgColor="#CCCCFF" border="1" cellPadding="2" width="696"> <tbody> <tr bgColor="#ffffff"> <td vAlign="top" width="131"><b>Reference</b></td> <td vAlign="top" width="529">&nbsp;<a href="" target="_blank"></a></td> </tr> </tbody> </table> <tr> <td align="middle" bgColor="#CCCCFF"> <p align="left"><font color="#660066"><b>DESCRIPTION</b></font></p> </td> </tr> <tr> <td> <table bgColor="#CCCCFF" border="1" cellPadding="2" width="100%" height="55"> <tbody> <tr bgColor="#ffffff"> <td vAlign="top" height="1"></td> </tr> </tbody> </table> </td> </tr> </tbody></table>
<p align="right"><a href="#TMT Generated Test Plans">Top</a></p><a name="Test Case ID: GeneConnectver 1.1caCore APISearchBasic Genomic ID Search"><center><table bgColor="#CCCCFF" border="1" cellPadding="0" cellSpacing="0" width="700">  <tbody>    <tr>      <td align="middle" bgColor="#CCCCFF">        <p align="left"><font color="#660066"><b>Test Case ID : 1870</b></font></p>      </td>    </tr>    <tr>      <td bgcolor="#FFFFFF"><b>Short Title : </b>Basic Genomic ID Search</td>    </tr>    <tr>      <td bgcolor="#FFFFFF"><b>Title : </b>Basic Genomic ID Search</td>    </tr>  </tbody></table><br><table bgColor="#CCCCFF" border="1" cellPadding="0" cellSpacing="0" width="700">  <tbody>    <tr>      <td align="middle" bgColor="#CCCCFF">        <p align="left"><font color="#660066"><b>VITALS</b></font></p>      </td>    </tr>    <tr>      <td>        <table bgColor="#CCCCFF" border="1" cellPadding="2" width="100%">          <tbody>            <tr bgColor="#ffffff">              <td vAlign="top"><b>Product</b></td>              <td vAlign="top" colspan="3">GeneConnect</td>            <tr bgColor="#ffffff">              <td vAlign="top"><b>Component</b></td>              <td vAlign="top">caCore API<br>              </td>              <td vAlign="top"><b>Test Area</b></td>              <td vAlign="top">Search</td>            </tr>            <tr bgColor="#ffffff">              <td vAlign="top"><b>Reference</b></td>              <td vAlign="top">1</td>              <td vAlign="top"><b>Priority</b></td>              <td vAlign="top">P1</td>            </tr>          </tbody>        </table>    <tr>      <td align="middle" bgColor="#CCCCFF">        <p align="left"><font color="#660066"><b>DESCRIPTION</b></font></p>      </td>    </tr>    <tr>      <td>        <table bgColor="#CCCCFF" border="1" cellPadding="2" width="100%" height="113">          <tbody>            <tr bgColor="#ffffff"><td vAlign="top" height="32" colspan="4"><font class="itsmultlinefont"><pre>Purpose:

To use caCore Like API query interface to retrive a set of associated genomic identifiers with a given data source as output and genomic identifer as input.

Setup:

1. caCore-Like system has been deployed
2. Application Server (where caCore-Like system is deployed) has been   
   started.
3. For test client refer to:
   http://cabigcvs.nci.nih.gov/viewcvs/viewcvs.cgi/geneconnect/Coding% 
   20and%20Testing/GeneConnectWeb/GeneConnect_caCore_Client/
4. For this test case refer to code 
   GeneConnect_caCore_Client/edu/wustl/geneconnect/testCase  
   /SimpleSearchTest.java

Procedure:
1. To start with create client application.

2. Create a domain object of GenomicIdentifierSet class.

3. Create a domain object of Gene class.

4. Set ensemblGeneId attribute of Gene to 'ENSG00000120738'.

5. Set unigeneAsOutput attribute of Gene to 'true'.

6. Set gene attribute of GenomicIdentifierSet to above creted Gene.

7. Create a domain object of Protein class.

8. Set ensemblPeptideAsOutput attribute of Protein to 'true'.

9. Set protein attribute of GenomicIdentifierSet to above creted Protein.

10. Create a domain object of MessengerRNA class.

10. Set ensemblTranscriptAsOutput attribute of MessengerRNA to 'true'.

11. Set messengerRNA attribute of GenomicIdentifierSet to above creted MessengerRNA.

12. Call a ApplicationService.search(targetClassName,GenomicIdentifierSetObject) of caCore API.

6. Execute the client application.


Input/Data:

The  query object must specify the search value of one or more genomic identifiers and ouyput data sources(attribute of Gene,MessengerRNA and /or Protein class).

Expected Output:

1. A list of objects which satisfy the search condition(s) will be returned. OR An empty list will be returned if no object matches search condition(s).
2. For data list please refer to  caCORE_API_Results.xml file 
   http://cabigcvs.nci.nih.gov/viewcvs/viewcvs.cgi/geneconnect/Coding%
   20and%20Testing/Test%20Activity%20Logs/caCORE_API_Results.xml



Verification Logic:

1. Verify by browsing mannualy through database.  

2. The correctness of the Frequency value for a particular genomic identifier can be verified as follows:
   Frequency is calculated as the ratio of number of occurrences of a ?single genomic identifier? in each combination of genomic identifiers to the total number of traversable paths for the given input.  

3. The correctness of the Confidence Score for a set of genomic identifiers can be verified as follows:
   Confidence score is calculated as the ratio of number of occurrences of a unique ?combination of genomic identifiers (includes only selected input and output data sources)? in a query, to the total number of traversable paths for the given input.</pre></font></td>            </tr>            <tr bgColor="#ffffff">              <td vAlign="top" height="21"><b>Mode of Execution</b></td>              <td vAlign="top" height="1">Manual                </td>              <td vAlign="top" height="21"><b>Verification Logic</b></td>              <td vAlign="top" height="1">DATABASE<br>                </td>            </tr>            <tr bgColor="#ffffff">              <td vAlign="top" height="21"><b>Author</b></td>              <td vAlign="top" height="21">Sachin Lale</td>              <td vAlign="top" height="21"><b>Date Entered</b></td>              <td vAlign="top" height="21">18/09/2006&nbsp;</td>            </tr>            <tr bgColor="#ffffff">              <td vAlign="top" height="21"></td>              <td vAlign="top" height="21"></td>              <td vAlign="top" height="21"><b>Date Last Update</b></td>              <td vAlign="top" height="21">24/01/2007&nbsp;</td>            </tr>          </tbody>        </table>      </td>    </tr>  </tbody></table>
<p align="right"><a href="#TMT Generated Test Plans">Top</a></p><a name="Test Case ID: GeneConnectver 1.1caCore APISearchQuery Based on Confidence Score"><center><table bgColor="#CCCCFF" border="1" cellPadding="0" cellSpacing="0" width="700">  <tbody>    <tr>      <td align="middle" bgColor="#CCCCFF">        <p align="left"><font color="#660066"><b>Test Case ID : 1871</b></font></p>      </td>    </tr>    <tr>      <td bgcolor="#FFFFFF"><b>Short Title : </b>Query Based on Confidence Score</td>    </tr>    <tr>      <td bgcolor="#FFFFFF"><b>Title : </b>Query Based on Confidence Score</td>    </tr>  </tbody></table><br><table bgColor="#CCCCFF" border="1" cellPadding="0" cellSpacing="0" width="700">  <tbody>    <tr>      <td align="middle" bgColor="#CCCCFF">        <p align="left"><font color="#660066"><b>VITALS</b></font></p>      </td>    </tr>    <tr>      <td>        <table bgColor="#CCCCFF" border="1" cellPadding="2" width="100%">          <tbody>            <tr bgColor="#ffffff">              <td vAlign="top"><b>Product</b></td>              <td vAlign="top" colspan="3">GeneConnect</td>            <tr bgColor="#ffffff">              <td vAlign="top"><b>Component</b></td>              <td vAlign="top">caCore API<br>              </td>              <td vAlign="top"><b>Test Area</b></td>              <td vAlign="top">Search</td>            </tr>            <tr bgColor="#ffffff">              <td vAlign="top"><b>Reference</b></td>              <td vAlign="top">1</td>              <td vAlign="top"><b>Priority</b></td>              <td vAlign="top">P1</td>            </tr>          </tbody>        </table>    <tr>      <td align="middle" bgColor="#CCCCFF">        <p align="left"><font color="#660066"><b>DESCRIPTION</b></font></p>      </td>    </tr>    <tr>      <td>        <table bgColor="#CCCCFF" border="1" cellPadding="2" width="100%" height="113">          <tbody>            <tr bgColor="#ffffff"><td vAlign="top" height="32" colspan="4"><font class="itsmultlinefont"><pre>Purpose:

To use caCore Like API query interface to retrive a set of associated genomic identifiers with a given data source as output and genomic identifer as input and criteria on confidence score.


Setup:

1. caCore-Like system has been deployed
2. Application Server (where caCore-Like system is deployed) has been   
   started.
3. For test client refer to:
   http://cabigcvs.nci.nih.gov/viewcvs/viewcvs.cgi/geneconnect/Coding% 
   20and%20Testing/GeneConnectWeb/GeneConnect_caCore_Client/
4. For this test case refer to code 
   GeneConnect_caCore_Client/edu/wustl/geneconnect/testCase/
   SearchWithConfidenceTest.java

Procedure:
1. To start with create client application.

2. Create a domain object of GenomicIdentifierSet class.

3. Create a domain object of Gene class.

4. Set ensemblGeneId attribute of Gene to 'ENSG00000120738'.

5. Set unigeneAsOutput attribute of Gene to 'true'.

6. Set gene attribute of GenomicIdentifierSet to above creted Gene.

7. Create a domain object of Protein class.

8. Set ensemblPeptideAsOutput attribute of Protein to 'true'.

9. Set protein attribute of GenomicIdentifierSet to above creted Protein.

10. Create a domain object of MessengerRNA class.

10. Set ensemblTranscriptAsOutput attribute of MessengerRNA to 'true'.

11. Set messengerRNA attribute of GenomicIdentifierSet to above creted MessengerRNA.

12. Set confidenceScore attribute of GenomicIdentifierSet to '0.3'.

13. Call a ApplicationService.search(targetClassName,GenomicIdentifierSetObject) of caCore API.

6. Execute the client application.


Input/Data:

The  query object must specify the search value of one or more genomic identifiers and ouyput data sources(attribute of Gene,MessengerRNA and /or Protein class).

Expected Output:

1. A list of objects which satisfy the search condition(s) will be returned. OR An empty list will be returned if no object matches search condition(s).

2.  For data list please refer to  caCORE_API_Results.xml file 
   http://cabigcvs.nci.nih.gov/viewcvs/viewcvs.cgi/geneconnect/Coding%
   20and%20Testing/Test%20Activity%20Logs/caCORE_API_Results.xml

Verification Logic:

1. Verify by browsing mannualy through database.  

2. The correctness of the Frequency value for a particular genomic identifier can be verified as follows:
   Frequency is calculated as the ratio of number of occurrences of a ?single genomic identifier? in each combination of genomic identifiers to the total number of traversable paths for the given input.  

3. The correctness of the Confidence Score for a set of genomic identifiers can be verified as follows:
   Confidence score is calculated as the ratio of number of occurrences of a unique ?combination of genomic identifiers (includes only selected input and output data sources)? in a query, to the total number of traversable paths for the given input.</pre></font></td>            </tr>            <tr bgColor="#ffffff">              <td vAlign="top" height="21"><b>Mode of Execution</b></td>              <td vAlign="top" height="1">Manual                </td>              <td vAlign="top" height="21"><b>Verification Logic</b></td>              <td vAlign="top" height="1">DATABASE<br>                </td>            </tr>            <tr bgColor="#ffffff">              <td vAlign="top" height="21"><b>Author</b></td>              <td vAlign="top" height="21">Sachin Lale</td>              <td vAlign="top" height="21"><b>Date Entered</b></td>              <td vAlign="top" height="21">18/09/2006&nbsp;</td>            </tr>            <tr bgColor="#ffffff">              <td vAlign="top" height="21"></td>              <td vAlign="top" height="21"></td>              <td vAlign="top" height="21"><b>Date Last Update</b></td>              <td vAlign="top" height="21">24/01/2007&nbsp;</td>            </tr>          </tbody>        </table>      </td>    </tr>  </tbody></table>
<p align="right"><a href="#TMT Generated Test Plans">Top</a></p><a name="Test Case ID: GeneConnectver 1.1caCore APISearchQuery By ONT"><center><table bgColor="#CCCCFF" border="1" cellPadding="0" cellSpacing="0" width="700">  <tbody>    <tr>      <td align="middle" bgColor="#CCCCFF">        <p align="left"><font color="#660066"><b>Test Case ID : 1872</b></font></p>      </td>    </tr>    <tr>      <td bgcolor="#FFFFFF"><b>Short Title : </b>Query By ONT</td>    </tr>    <tr>      <td bgcolor="#FFFFFF"><b>Title : </b>Query By Order of Node Traversal</td>    </tr>  </tbody></table><br><table bgColor="#CCCCFF" border="1" cellPadding="0" cellSpacing="0" width="700">  <tbody>    <tr>      <td align="middle" bgColor="#CCCCFF">        <p align="left"><font color="#660066"><b>VITALS</b></font></p>      </td>    </tr>    <tr>      <td>        <table bgColor="#CCCCFF" border="1" cellPadding="2" width="100%">          <tbody>            <tr bgColor="#ffffff">              <td vAlign="top"><b>Product</b></td>              <td vAlign="top" colspan="3">GeneConnect</td>            <tr bgColor="#ffffff">              <td vAlign="top"><b>Component</b></td>              <td vAlign="top">caCore API<br>              </td>              <td vAlign="top"><b>Test Area</b></td>              <td vAlign="top">Search</td>            </tr>            <tr bgColor="#ffffff">              <td vAlign="top"><b>Reference</b></td>              <td vAlign="top">1</td>              <td vAlign="top"><b>Priority</b></td>              <td vAlign="top">P1</td>            </tr>          </tbody>        </table>    <tr>      <td align="middle" bgColor="#CCCCFF">        <p align="left"><font color="#660066"><b>DESCRIPTION</b></font></p>      </td>    </tr>    <tr>      <td>        <table bgColor="#CCCCFF" border="1" cellPadding="2" width="100%" height="113">          <tbody>            <tr bgColor="#ffffff"><td vAlign="top" height="32" colspan="4"><font class="itsmultlinefont"><pre>Purpose:

To use caCore Like API query interface to retrive a set of associated genomic identifiers with a given data source as output and genomic identifer as input and criteria on Order of node traversal.

Setup:

Setup:

1. caCore-Like system has been deployed
2. Application Server (where caCore-Like system is deployed) has been   
   started.
3. For test client refer to:
   http://cabigcvs.nci.nih.gov/viewcvs/viewcvs.cgi/geneconnect/Coding% 
   20and%20Testing/GeneConnectWeb/GeneConnect_caCore_Client/
4. For this test case refer to code 
   GeneConnect_caCore_Client/edu/wustl/geneconnect/testCase/
   SearchWithOntTest.java

Procedure:
1. To start with create client application.

2. Create a domain object of GenomicIdentifierSet class.

3. Create a domain object of Gene class.

4. Set ensemblGeneId attribute of Gene to 'ENSG00000120738'.

5. Set unigeneAsOutput attribute of Gene to 'true'.

6. Set gene attribute of GenomicIdentifierSet to above creted Gene.

7. Create a domain object of Protein class.

8. Set ensemblPeptideAsOutput attribute of Protein to 'true'.

9. Set protein attribute of GenomicIdentifierSet to above creted Protein.

10. Create a domain object of MessengerRNA class.

10. Set ensemblTranscriptAsOutput attribute of MessengerRNA to 'true'.

11. Set messengerRNA attribute of GenomicIdentifierSet to above creted MessengerRNA.

12. Create instance of OrderOfNodeTraversal.

13. Set the data source and link type attribute. 

14. Set the path as UniGene {INFERRED} Ensembl Gene {DIRECT} Ensembl Transcript {DIRECT} Ensembl Protein

15. Add OrderOfNodeTraversal object to a collection and set this collection to orderOfNodeTraversalCollection attribute of GenomicIdentifierSet.

16. Call a ApplicationService.search(targetClassName,GenomicIdentifierSetObject) of caCore API.

17. Execute the client application.


Input/Data:

The  query object must specify the search value of one or more genomic identifiers and ouyput data sources(attribute of Gene,MessengerRNA and /or Protein class).

Expected Output:

1. A list of objects which satisfy the search condition(s) will be returned. OR An empty list will be returned if no object matches search condition(s).
2. For data list please refer to  caCORE_API_Results.xml file 
   http://cabigcvs.nci.nih.gov/viewcvs/viewcvs.cgi/geneconnect/Coding%
   20and%20Testing/Test%20Activity%20Logs/caCORE_API_Results.xml


Verification Logic:

1. Verify by browsing mannualy through database.  

2. The correctness of the Frequency value for a particular genomic identifier can be verified as follows:
   Frequency is calculated as the ratio of number of occurrences of a ?single genomic identifier? in each combination of genomic identifiers to the total number of traversable paths for the given input.  

3. The correctness of the Confidence Score for a set of genomic identifiers can be verified as follows:
   Confidence score is calculated as the ratio of number of occurrences of a unique ?combination of genomic identifiers (includes only selected input and output data sources)? in a query, to the total number of traversable paths for the given input.</pre></font></td>            </tr>            <tr bgColor="#ffffff">              <td vAlign="top" height="21"><b>Mode of Execution</b></td>              <td vAlign="top" height="1">Manual                </td>              <td vAlign="top" height="21"><b>Verification Logic</b></td>              <td vAlign="top" height="1">DATABASE<br>                </td>            </tr>            <tr bgColor="#ffffff">              <td vAlign="top" height="21"><b>Author</b></td>              <td vAlign="top" height="21">Sachin Lale</td>              <td vAlign="top" height="21"><b>Date Entered</b></td>              <td vAlign="top" height="21">18/09/2006&nbsp;</td>            </tr>            <tr bgColor="#ffffff">              <td vAlign="top" height="21"></td>              <td vAlign="top" height="21"></td>              <td vAlign="top" height="21"><b>Date Last Update</b></td>              <td vAlign="top" height="21">24/01/2007&nbsp;</td>            </tr>          </tbody>        </table>      </td>    </tr>  </tbody></table>
<p align="right"><a href="#TMT Generated Test Plans">Top</a></p><a name="Test Case ID: GeneConnectver 1.1caCore APISearchQuery By Limiting ID Frequency"><center><table bgColor="#CCCCFF" border="1" cellPadding="0" cellSpacing="0" width="700">  <tbody>    <tr>      <td align="middle" bgColor="#CCCCFF">        <p align="left"><font color="#660066"><b>Test Case ID : 1873</b></font></p>      </td>    </tr>    <tr>      <td bgcolor="#FFFFFF"><b>Short Title : </b>Query By Limiting ID Frequency</td>    </tr>    <tr>      <td bgcolor="#FFFFFF"><b>Title : </b>Query By Limiting ID Frequency</td>    </tr>  </tbody></table><br><table bgColor="#CCCCFF" border="1" cellPadding="0" cellSpacing="0" width="700">  <tbody>    <tr>      <td align="middle" bgColor="#CCCCFF">        <p align="left"><font color="#660066"><b>VITALS</b></font></p>      </td>    </tr>    <tr>      <td>        <table bgColor="#CCCCFF" border="1" cellPadding="2" width="100%">          <tbody>            <tr bgColor="#ffffff">              <td vAlign="top"><b>Product</b></td>              <td vAlign="top" colspan="3">GeneConnect</td>            <tr bgColor="#ffffff">              <td vAlign="top"><b>Component</b></td>              <td vAlign="top">caCore API<br>              </td>              <td vAlign="top"><b>Test Area</b></td>              <td vAlign="top">Search</td>            </tr>            <tr bgColor="#ffffff">              <td vAlign="top"><b>Reference</b></td>              <td vAlign="top">1</td>              <td vAlign="top"><b>Priority</b></td>              <td vAlign="top">P1</td>            </tr>          </tbody>        </table>    <tr>      <td align="middle" bgColor="#CCCCFF">        <p align="left"><font color="#660066"><b>DESCRIPTION</b></font></p>      </td>    </tr>    <tr>      <td>        <table bgColor="#CCCCFF" border="1" cellPadding="2" width="100%" height="113">          <tbody>            <tr bgColor="#ffffff"><td vAlign="top" height="32" colspan="4"><font class="itsmultlinefont"><pre>Purpose:

To use caCore Like API query interface to retrive a set of associated genomic identifiers with a given data source as output and genomic identifer as input and criteria on frequency.


Setup:

1. caCore-Like system has been deployed
2. Application Server (where caCore-Like system is deployed) has been   
   started.
3. For test client refer to:
   http://cabigcvs.nci.nih.gov/viewcvs/viewcvs.cgi/geneconnect/Coding% 
   20and%20Testing/GeneConnectWeb/GeneConnect_caCore_Client/
4. For this test case refer to code 
   GeneConnect_caCore_Client/edu/wustl/geneconnect/testCase/
   SearchWithFrequencyTest.java

Procedure:
1. To start with create client application.

2. Create a domain object of GenomicIdentifierSet class.

3. Create a domain object of Gene class.

4. Set ensemblGeneId attribute of Gene to 'ENSG00000120738'.

5. Set unigeneAsOutput attribute of Gene to 'true'.

6. Set gene attribute of GenomicIdentifierSet to above creted Gene.

7. Create a domain object of Protein class.

8. Set ensemblPeptideAsOutput attribute of Protein to 'true'.

9. Set protein attribute of GenomicIdentifierSet to above creted Protein.

10. Create a domain object of MessengerRNA class.

10. Set ensemblTranscriptAsOutput attribute of MessengerRNA to 'true'.

11. Set messengerRNA attribute of GenomicIdentifierSet to above creted MessengerRNA.

12. Create instance of ConsensusIdentifierData.

13. Create Instance of edu.wustl.geneconnect.domain.UniGene

14. Set genomicIdentifier attribute of ConsensusIdentifierData to above creted UniGene object.

15. Set frequecny attribute of ConsensusIdentifierData to '0.3'.

16. Add ConsensusIdentifierData object to a collection and set this collection to consensusIdentifierDataCollection attribute of 
GenomicIdentifierSet.

17. Call a ApplicationService.search(targetClassName,GenomicIdentifierSetObject) of caCore API.

18. Execute the client application.


Input/Data:

The  query object must specify the search value of one or more genomic identifiers and ouyput data sources(attribute of Gene,MessengerRNA and /or Protein class).

Expected Output:

1. A list of objects which satisfy the search condition(s) will be returned. OR An empty list will be returned if no object matches search condition(s).
2. For data list please refer to  caCORE_API_Results.xml file 
   http://cabigcvs.nci.nih.gov/viewcvs/viewcvs.cgi/geneconnect/Coding%
   20and%20Testing/Test%20Activity%20Logs/caCORE_API_Results.xml

Verification Logic:

1. Verify by browsing mannualy through database.  

2. The correctness of the Frequency value for a particular genomic identifier can be verified as follows:
   Frequency is calculated as the ratio of number of occurrences of a ?single genomic identifier? in each combination of genomic identifiers to the total number of traversable paths for the given input.  

3. The correctness of the Confidence Score for a set of genomic identifiers can be verified as follows:
   Confidence score is calculated as the ratio of number of occurrences of a unique ?combination of genomic identifiers (includes only selected input and output data sources)? in a query, to the total number of traversable paths for the given input.</pre></font></td>            </tr>            <tr bgColor="#ffffff">              <td vAlign="top" height="21"><b>Mode of Execution</b></td>              <td vAlign="top" height="1">Manual                </td>              <td vAlign="top" height="21"><b>Verification Logic</b></td>              <td vAlign="top" height="1">DATABASE<br>                </td>            </tr>            <tr bgColor="#ffffff">              <td vAlign="top" height="21"><b>Author</b></td>              <td vAlign="top" height="21">Sachin Lale</td>              <td vAlign="top" height="21"><b>Date Entered</b></td>              <td vAlign="top" height="21">18/09/2006&nbsp;</td>            </tr>            <tr bgColor="#ffffff">              <td vAlign="top" height="21"></td>              <td vAlign="top" height="21"></td>              <td vAlign="top" height="21"><b>Date Last Update</b></td>              <td vAlign="top" height="21">24/01/2007&nbsp;</td>            </tr>          </tbody>        </table>      </td>    </tr>  </tbody></table>
<p align="right"><a href="#TMT Generated Test Plans">Top</a></p></body></html>
